/**
 * Autogenerated by Avro
 * 
 * DO NOT EDIT DIRECTLY
 */
package org.opencb.biodata.models.variant.avro;  
@SuppressWarnings("all")
@org.apache.avro.specific.AvroGenerated
public class VariantAvro extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  public static final org.apache.avro.Schema SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\",\"name\":\"VariantAvro\",\"namespace\":\"org.opencb.biodata.models.variant.avro\",\"fields\":[{\"name\":\"chromosome\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"* Chromosome where the genomic variation occurred.\"},{\"name\":\"start\",\"type\":\"int\",\"doc\":\"* Normalized position where the genomic variation starts.\\n         * <ul>\\n         * <li>SNVs have the same start and end position</li>\\n         * <li>Insertions start in the last present position: if the first nucleotide\\n         * is inserted in position 6, the start is position 5</li>\\n         * <li>Deletions start in the first previously present position: if the first\\n         * deleted nucleotide is in position 6, the start is position 6</li>\\n         * </ul>\"},{\"name\":\"end\",\"type\":\"int\",\"doc\":\"* Normalized position where the genomic variation ends.\\n         * <ul>\\n         * <li>SNVs have the same start and end positions</li>\\n         * <li>Insertions end in the first present position: if the last nucleotide\\n         * is inserted in position 9, the end is position 10</li>\\n         * <li>Deletions ends in the last previously present position: if the last\\n         * deleted nucleotide is in position 9, the end is position 9</li>\\n         * </ul>\"},{\"name\":\"reference\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"* Reference allele.\"},{\"name\":\"alternate\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"* Alternate allele.\"},{\"name\":\"strand\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"* Reference strand for this variant\"},{\"name\":\"ids\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},\"doc\":\"* Set of identifiers used for this genomic variation.\",\"default\":[]},{\"name\":\"length\",\"type\":\"int\",\"doc\":\"* Length of the genomic variation, which depends on the variation type.\\n         * <ul>\\n         * <li>SNVs have a length of 1 nucleotide</li>\\n         * <li>Indels have the length of the largest allele</li>\\n         * </ul>\"},{\"name\":\"type\",\"type\":{\"type\":\"enum\",\"name\":\"VariantType\",\"doc\":\"* Type of variation, which depends mostly on its length.\\n     * <ul>\\n     * <li>SNVs involve a single nucleotide, without changes in length</li>\\n     * <li>MNVs involve multiple nucleotides, without changes in length</li>\\n     * <li>Indels are insertions or deletions of less than SV_THRESHOLD (50) nucleotides</li>\\n     * <li>Structural variations are large changes of more than SV_THRESHOLD nucleotides</li>\\n     * <li>Copy-number variations alter the number of copies of a region</li>\\n     * </ul>\",\"symbols\":[\"SNV\",\"SNP\",\"MNV\",\"MNP\",\"INDEL\",\"SV\",\"INSERTION\",\"DELETION\",\"TRANSLOCATION\",\"INVERSION\",\"CNV\",\"NO_VARIATION\",\"SYMBOLIC\",\"MIXED\"]},\"doc\":\"* Type of variation: single nucleotide, indel or structural variation.\"},{\"name\":\"hgvs\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},\"avro.java.string\":\"String\"}],\"doc\":\"* Unique identifier following the HGVS nomenclature.\"},{\"name\":\"studies\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"StudyEntry\",\"fields\":[{\"name\":\"studyId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"* Unique identifier of the study.\"},{\"name\":\"files\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"FileEntry\",\"fields\":[{\"name\":\"fileId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"* Unique identifier of the source file.\"},{\"name\":\"call\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"* Original call position for the variant, if the file was normalized.\\n         *\\n         * {position}:{reference}:{alternate}(,{other_alternate})*:{allele_index}\"},{\"name\":\"attributes\",\"type\":{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"},\"doc\":\"* Optional attributes that probably depend on the format of the file the\\n         * variant was initially read from.\"}]}},\"doc\":\"* List of files from the study where the variant was present.\",\"default\":[]},{\"name\":\"secondaryAlternates\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AlternateCoordinate\",\"fields\":[{\"name\":\"chromosome\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"start\",\"type\":[\"null\",\"int\"]},{\"name\":\"end\",\"type\":[\"null\",\"int\"]},{\"name\":\"reference\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"alternate\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"type\",\"type\":\"VariantType\"}]}}],\"doc\":\"* Alternate alleles that appear along with a variant alternate.\",\"default\":null},{\"name\":\"format\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},\"doc\":\"* Fields stored for each sample.\"},{\"name\":\"samplesData\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},\"doc\":\"* Genotypes and other sample-related information. Each position is related\\n         * with one sample. The content are lists of values in the same order than the\\n         * format array. The length of this lists must be the same as the format field.\"},{\"name\":\"stats\",\"type\":{\"type\":\"map\",\"values\":{\"type\":\"record\",\"name\":\"VariantStats\",\"fields\":[{\"name\":\"refAllele\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"altAllele\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"refAlleleCount\",\"type\":[\"null\",\"int\"]},{\"name\":\"altAlleleCount\",\"type\":[\"null\",\"int\"]},{\"name\":\"genotypesCount\",\"type\":{\"type\":\"map\",\"values\":\"int\",\"avro.java.string\":\"String\",\"java-key-class\":\"org.opencb.biodata.models.feature.Genotype\"}},{\"name\":\"genotypesFreq\",\"type\":{\"type\":\"map\",\"values\":\"float\",\"avro.java.string\":\"String\",\"java-key-class\":\"org.opencb.biodata.models.feature.Genotype\"}},{\"name\":\"missingAlleles\",\"type\":[\"null\",\"int\"]},{\"name\":\"missingGenotypes\",\"type\":[\"null\",\"int\"]},{\"name\":\"refAlleleFreq\",\"type\":[\"null\",\"float\"]},{\"name\":\"altAlleleFreq\",\"type\":[\"null\",\"float\"]},{\"name\":\"maf\",\"type\":[\"null\",\"float\"]},{\"name\":\"mgf\",\"type\":[\"null\",\"float\"]},{\"name\":\"mafAllele\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"mgfGenotype\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"passedFilters\",\"type\":[\"null\",\"boolean\"]},{\"name\":\"mendelianErrors\",\"type\":[\"null\",\"int\"]},{\"name\":\"casesPercentDominant\",\"type\":[\"null\",\"float\"]},{\"name\":\"controlsPercentDominant\",\"type\":[\"null\",\"float\"]},{\"name\":\"casesPercentRecessive\",\"type\":[\"null\",\"float\"]},{\"name\":\"controlsPercentRecessive\",\"type\":[\"null\",\"float\"]},{\"name\":\"quality\",\"type\":[\"null\",\"float\"]},{\"name\":\"numSamples\",\"type\":[\"null\",\"int\"]},{\"name\":\"variantType\",\"type\":\"VariantType\"},{\"name\":\"hw\",\"type\":{\"type\":\"record\",\"name\":\"VariantHardyWeinbergStats\",\"fields\":[{\"name\":\"chi2\",\"type\":[\"null\",\"float\"]},{\"name\":\"pValue\",\"type\":[\"null\",\"float\"]},{\"name\":\"n\",\"type\":[\"null\",\"int\"]},{\"name\":\"n_AA_11\",\"type\":[\"null\",\"int\"]},{\"name\":\"n_Aa_10\",\"type\":[\"null\",\"int\"]},{\"name\":\"n_aa_00\",\"type\":[\"null\",\"int\"]},{\"name\":\"e_AA_11\",\"type\":[\"null\",\"float\"]},{\"name\":\"e_Aa_10\",\"type\":[\"null\",\"float\"]},{\"name\":\"e_aa_00\",\"type\":[\"null\",\"float\"]},{\"name\":\"p\",\"type\":[\"null\",\"float\"]},{\"name\":\"q\",\"type\":[\"null\",\"float\"]}]}}]},\"avro.java.string\":\"String\"},\"doc\":\"* Statistics of the genomic variation, such as its alleles/genotypes count\\n         * or its minimum allele frequency, grouped by cohort name.\"}]}},\"doc\":\"* Information specific to each study the variant was read from, such as\\n         * samples or statistics.\"},{\"name\":\"annotation\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VariantAnnotation\",\"fields\":[{\"name\":\"chromosome\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"start\",\"type\":\"int\"},{\"name\":\"end\",\"type\":\"int\"},{\"name\":\"reference\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"alternate\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"id\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"xrefs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Xref\",\"fields\":[{\"name\":\"id\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"source\",\"typ","e\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]},{\"name\":\"hgvs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"consequenceTypes\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ConsequenceType\",\"fields\":[{\"name\":\"geneName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"ensemblGeneId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"ensemblTranscriptId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"strand\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"biotype\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"transcriptAnnotationFlags\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"cdnaPosition\",\"type\":[\"null\",\"int\"]},{\"name\":\"cdsPosition\",\"type\":[\"null\",\"int\"]},{\"name\":\"codon\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"proteinVariantAnnotation\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"ProteinVariantAnnotation\",\"fields\":[{\"name\":\"uniprotAccession\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"uniprotName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"position\",\"type\":\"int\"},{\"name\":\"reference\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"alternate\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"uniprotVariantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"functionalDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"substitutionScores\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Score\",\"fields\":[{\"name\":\"score\",\"type\":\"double\"},{\"name\":\"source\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]},{\"name\":\"keywords\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"features\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ProteinFeature\",\"fields\":[{\"name\":\"id\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"start\",\"type\":\"int\"},{\"name\":\"end\",\"type\":\"int\"},{\"name\":\"type\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]}]}]},{\"name\":\"sequenceOntologyTerms\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"SequenceOntologyTerm\",\"fields\":[{\"name\":\"accession\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"name\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}}]}},\"default\":[]},{\"name\":\"populationFrequencies\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"PopulationFrequency\",\"fields\":[{\"name\":\"study\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"population\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"superPopulation\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"refAllele\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"altAllele\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"refAlleleFreq\",\"type\":\"float\"},{\"name\":\"altAlleleFreq\",\"type\":\"float\"},{\"name\":\"refHomGenotypeFreq\",\"type\":[\"null\",\"float\"]},{\"name\":\"hetGenotypeFreq\",\"type\":[\"null\",\"float\"]},{\"name\":\"altHomGenotypeFreq\",\"type\":[\"null\",\"float\"]}]}}]},{\"name\":\"conservation\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"Score\"}]},{\"name\":\"geneExpression\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Expression\",\"fields\":[{\"name\":\"geneName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"transcriptId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"experimentalFactor\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"factorValue\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"experimentId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"technologyPlatform\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"expression\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"ExpressionCall\",\"symbols\":[\"UP\",\"DOWN\"]}]},{\"name\":\"pvalue\",\"type\":[\"null\",\"float\"]}]}}]},{\"name\":\"geneTraitAssociation\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"GeneTraitAssociation\",\"fields\":[{\"name\":\"id\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"name\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"hpo\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"score\",\"type\":[\"null\",\"float\"]},{\"name\":\"numberOfPubmeds\",\"type\":[\"null\",\"int\"]},{\"name\":\"associationTypes\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"sources\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"source\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]},{\"name\":\"geneDrugInteraction\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"GeneDrugInteraction\",\"fields\":[{\"name\":\"geneName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"drugName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"studyType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"type\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}]},{\"name\":\"variantTraitAssociation\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VariantTraitAssociation\",\"fields\":[{\"name\":\"clinvar\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ClinVar\",\"fields\":[{\"name\":\"accession\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"clinicalSignificance\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"traits\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},{\"name\":\"geneNames\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},{\"name\":\"reviewStatus\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]},{\"name\":\"gwas\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Gwas\",\"fields\":[{\"name\":\"snpIdCurrent\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"traits\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},{\"name\":\"riskAlleleFrequency\",\"type\":\"double\"},{\"name\":\"reportedGenes\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]},{\"name\":\"cosmic\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Cosmic\",\"fields\":[{\"name\":\"mutationId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"primarySite\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"siteSubtype\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"primaryHistology\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"histologySubtype\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"sampleSource\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"tumourOrigin\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"geneName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"mutationSomaticStatus\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}]}]}]},{\"name\":\"functionalScore\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"Score\"}]},{\"name\":\"additionalAttributes\",\"type\":[\"null\",{\"type\":\"map\",\"values\":[\"int\",\"float\",\"double\",{\"type\":\"string\",\"avro.java.string\":\"String\"},{\"type\":\"map\",\"values\":[\"int\",\"float\",\"double\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"avro.java.string\":\"String\"}],\"avro.java.string\":\"String\"}]}]}],\"doc\":\"* Annotations of the genomic variation.\"}]}");
  public static org.apache.avro.Schema getClassSchema() { return SCHEMA$; }
  /** * Chromosome where the genomic variation occurred. */
   private java.lang.String chromosome;
  /** * Normalized position where the genomic variation starts.
         * <ul>
         * <li>SNVs have the same start and end position</li>
         * <li>Insertions start in the last present position: if the first nucleotide
         * is inserted in position 6, the start is position 5</li>
         * <li>Deletions start in the first previously present position: if the first
         * deleted nucleotide is in position 6, the start is position 6</li>
         * </ul> */
   private int start;
  /** * Normalized position where the genomic variation ends.
         * <ul>
         * <li>SNVs have the same start and end positions</li>
         * <li>Insertions end in the first present position: if the last nucleotide
         * is inserted in position 9, the end is position 10</li>
         * <li>Deletions ends in the last previously present position: if the last
         * deleted nucleotide is in position 9, the end is position 9</li>
         * </ul> */
   private int end;
  /** * Reference allele. */
   private java.lang.String reference;
  /** * Alternate allele. */
   private java.lang.String alternate;
  /** * Reference strand for this variant */
   private java.lang.String strand;
  /** * Set of identifiers used for this genomic variation. */
   private java.util.List<java.lang.String> ids;
  /** * Length of the genomic variation, which depends on the variation type.
         * <ul>
         * <li>SNVs have a length of 1 nucleotide</li>
         * <li>Indels have the length of the largest allele</li>
         * </ul> */
   private int length;
  /** * Type of variation: single nucleotide, indel or structural variation. */
   private org.opencb.biodata.models.variant.avro.VariantType type;
  /** * Unique identifier following the HGVS nomenclature. */
   private java.util.Map<java.lang.String,java.util.List<java.lang.String>> hgvs;
  /** * Information specific to each study the variant was read from, such as
         * samples or statistics. */
   private java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry> studies;
  /** * Annotations of the genomic variation. */
   private org.opencb.biodata.models.variant.avro.VariantAnnotation annotation;

  /**
   * Default constructor.  Note that this does not initialize fields
   * to their default values from the schema.  If that is desired then
   * one should use <code>newBuilder()</code>. 
   */
  public VariantAvro() {}

  /**
   * All-args constructor.
   */
  public VariantAvro(java.lang.String chromosome, java.lang.Integer start, java.lang.Integer end, java.lang.String reference, java.lang.String alternate, java.lang.String strand, java.util.List<java.lang.String> ids, java.lang.Integer length, org.opencb.biodata.models.variant.avro.VariantType type, java.util.Map<java.lang.String,java.util.List<java.lang.String>> hgvs, java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry> studies, org.opencb.biodata.models.variant.avro.VariantAnnotation annotation) {
    this.chromosome = chromosome;
    this.start = start;
    this.end = end;
    this.reference = reference;
    this.alternate = alternate;
    this.strand = strand;
    this.ids = ids;
    this.length = length;
    this.type = type;
    this.hgvs = hgvs;
    this.studies = studies;
    this.annotation = annotation;
  }

  public org.apache.avro.Schema getSchema() { return SCHEMA$; }
  // Used by DatumWriter.  Applications should not call. 
  public java.lang.Object get(int field$) {
    switch (field$) {
    case 0: return chromosome;
    case 1: return start;
    case 2: return end;
    case 3: return reference;
    case 4: return alternate;
    case 5: return strand;
    case 6: return ids;
    case 7: return length;
    case 8: return type;
    case 9: return hgvs;
    case 10: return studies;
    case 11: return annotation;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }
  // Used by DatumReader.  Applications should not call. 
  @SuppressWarnings(value="unchecked")
  public void put(int field$, java.lang.Object value$) {
    switch (field$) {
    case 0: chromosome = (java.lang.String)value$; break;
    case 1: start = (java.lang.Integer)value$; break;
    case 2: end = (java.lang.Integer)value$; break;
    case 3: reference = (java.lang.String)value$; break;
    case 4: alternate = (java.lang.String)value$; break;
    case 5: strand = (java.lang.String)value$; break;
    case 6: ids = (java.util.List<java.lang.String>)value$; break;
    case 7: length = (java.lang.Integer)value$; break;
    case 8: type = (org.opencb.biodata.models.variant.avro.VariantType)value$; break;
    case 9: hgvs = (java.util.Map<java.lang.String,java.util.List<java.lang.String>>)value$; break;
    case 10: studies = (java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry>)value$; break;
    case 11: annotation = (org.opencb.biodata.models.variant.avro.VariantAnnotation)value$; break;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }

  /**
   * Gets the value of the 'chromosome' field.
   * * Chromosome where the genomic variation occurred.   */
  public java.lang.String getChromosome() {
    return chromosome;
  }

  /**
   * Sets the value of the 'chromosome' field.
   * * Chromosome where the genomic variation occurred.   * @param value the value to set.
   */
  public void setChromosome(java.lang.String value) {
    this.chromosome = value;
  }

  /**
   * Gets the value of the 'start' field.
   * * Normalized position where the genomic variation starts.
         * <ul>
         * <li>SNVs have the same start and end position</li>
         * <li>Insertions start in the last present position: if the first nucleotide
         * is inserted in position 6, the start is position 5</li>
         * <li>Deletions start in the first previously present position: if the first
         * deleted nucleotide is in position 6, the start is position 6</li>
         * </ul>   */
  public java.lang.Integer getStart() {
    return start;
  }

  /**
   * Sets the value of the 'start' field.
   * * Normalized position where the genomic variation starts.
         * <ul>
         * <li>SNVs have the same start and end position</li>
         * <li>Insertions start in the last present position: if the first nucleotide
         * is inserted in position 6, the start is position 5</li>
         * <li>Deletions start in the first previously present position: if the first
         * deleted nucleotide is in position 6, the start is position 6</li>
         * </ul>   * @param value the value to set.
   */
  public void setStart(java.lang.Integer value) {
    this.start = value;
  }

  /**
   * Gets the value of the 'end' field.
   * * Normalized position where the genomic variation ends.
         * <ul>
         * <li>SNVs have the same start and end positions</li>
         * <li>Insertions end in the first present position: if the last nucleotide
         * is inserted in position 9, the end is position 10</li>
         * <li>Deletions ends in the last previously present position: if the last
         * deleted nucleotide is in position 9, the end is position 9</li>
         * </ul>   */
  public java.lang.Integer getEnd() {
    return end;
  }

  /**
   * Sets the value of the 'end' field.
   * * Normalized position where the genomic variation ends.
         * <ul>
         * <li>SNVs have the same start and end positions</li>
         * <li>Insertions end in the first present position: if the last nucleotide
         * is inserted in position 9, the end is position 10</li>
         * <li>Deletions ends in the last previously present position: if the last
         * deleted nucleotide is in position 9, the end is position 9</li>
         * </ul>   * @param value the value to set.
   */
  public void setEnd(java.lang.Integer value) {
    this.end = value;
  }

  /**
   * Gets the value of the 'reference' field.
   * * Reference allele.   */
  public java.lang.String getReference() {
    return reference;
  }

  /**
   * Sets the value of the 'reference' field.
   * * Reference allele.   * @param value the value to set.
   */
  public void setReference(java.lang.String value) {
    this.reference = value;
  }

  /**
   * Gets the value of the 'alternate' field.
   * * Alternate allele.   */
  public java.lang.String getAlternate() {
    return alternate;
  }

  /**
   * Sets the value of the 'alternate' field.
   * * Alternate allele.   * @param value the value to set.
   */
  public void setAlternate(java.lang.String value) {
    this.alternate = value;
  }

  /**
   * Gets the value of the 'strand' field.
   * * Reference strand for this variant   */
  public java.lang.String getStrand() {
    return strand;
  }

  /**
   * Sets the value of the 'strand' field.
   * * Reference strand for this variant   * @param value the value to set.
   */
  public void setStrand(java.lang.String value) {
    this.strand = value;
  }

  /**
   * Gets the value of the 'ids' field.
   * * Set of identifiers used for this genomic variation.   */
  public java.util.List<java.lang.String> getIds() {
    return ids;
  }

  /**
   * Sets the value of the 'ids' field.
   * * Set of identifiers used for this genomic variation.   * @param value the value to set.
   */
  public void setIds(java.util.List<java.lang.String> value) {
    this.ids = value;
  }

  /**
   * Gets the value of the 'length' field.
   * * Length of the genomic variation, which depends on the variation type.
         * <ul>
         * <li>SNVs have a length of 1 nucleotide</li>
         * <li>Indels have the length of the largest allele</li>
         * </ul>   */
  public java.lang.Integer getLength() {
    return length;
  }

  /**
   * Sets the value of the 'length' field.
   * * Length of the genomic variation, which depends on the variation type.
         * <ul>
         * <li>SNVs have a length of 1 nucleotide</li>
         * <li>Indels have the length of the largest allele</li>
         * </ul>   * @param value the value to set.
   */
  public void setLength(java.lang.Integer value) {
    this.length = value;
  }

  /**
   * Gets the value of the 'type' field.
   * * Type of variation: single nucleotide, indel or structural variation.   */
  public org.opencb.biodata.models.variant.avro.VariantType getType() {
    return type;
  }

  /**
   * Sets the value of the 'type' field.
   * * Type of variation: single nucleotide, indel or structural variation.   * @param value the value to set.
   */
  public void setType(org.opencb.biodata.models.variant.avro.VariantType value) {
    this.type = value;
  }

  /**
   * Gets the value of the 'hgvs' field.
   * * Unique identifier following the HGVS nomenclature.   */
  public java.util.Map<java.lang.String,java.util.List<java.lang.String>> getHgvs() {
    return hgvs;
  }

  /**
   * Sets the value of the 'hgvs' field.
   * * Unique identifier following the HGVS nomenclature.   * @param value the value to set.
   */
  public void setHgvs(java.util.Map<java.lang.String,java.util.List<java.lang.String>> value) {
    this.hgvs = value;
  }

  /**
   * Gets the value of the 'studies' field.
   * * Information specific to each study the variant was read from, such as
         * samples or statistics.   */
  public java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry> getStudies() {
    return studies;
  }

  /**
   * Sets the value of the 'studies' field.
   * * Information specific to each study the variant was read from, such as
         * samples or statistics.   * @param value the value to set.
   */
  public void setStudies(java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry> value) {
    this.studies = value;
  }

  /**
   * Gets the value of the 'annotation' field.
   * * Annotations of the genomic variation.   */
  public org.opencb.biodata.models.variant.avro.VariantAnnotation getAnnotation() {
    return annotation;
  }

  /**
   * Sets the value of the 'annotation' field.
   * * Annotations of the genomic variation.   * @param value the value to set.
   */
  public void setAnnotation(org.opencb.biodata.models.variant.avro.VariantAnnotation value) {
    this.annotation = value;
  }

  /** Creates a new VariantAvro RecordBuilder */
  public static org.opencb.biodata.models.variant.avro.VariantAvro.Builder newBuilder() {
    return new org.opencb.biodata.models.variant.avro.VariantAvro.Builder();
  }
  
  /** Creates a new VariantAvro RecordBuilder by copying an existing Builder */
  public static org.opencb.biodata.models.variant.avro.VariantAvro.Builder newBuilder(org.opencb.biodata.models.variant.avro.VariantAvro.Builder other) {
    return new org.opencb.biodata.models.variant.avro.VariantAvro.Builder(other);
  }
  
  /** Creates a new VariantAvro RecordBuilder by copying an existing VariantAvro instance */
  public static org.opencb.biodata.models.variant.avro.VariantAvro.Builder newBuilder(org.opencb.biodata.models.variant.avro.VariantAvro other) {
    return new org.opencb.biodata.models.variant.avro.VariantAvro.Builder(other);
  }
  
  /**
   * RecordBuilder for VariantAvro instances.
   */
  public static class Builder extends org.apache.avro.specific.SpecificRecordBuilderBase<VariantAvro>
    implements org.apache.avro.data.RecordBuilder<VariantAvro> {

    private java.lang.String chromosome;
    private int start;
    private int end;
    private java.lang.String reference;
    private java.lang.String alternate;
    private java.lang.String strand;
    private java.util.List<java.lang.String> ids;
    private int length;
    private org.opencb.biodata.models.variant.avro.VariantType type;
    private java.util.Map<java.lang.String,java.util.List<java.lang.String>> hgvs;
    private java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry> studies;
    private org.opencb.biodata.models.variant.avro.VariantAnnotation annotation;

    /** Creates a new Builder */
    private Builder() {
      super(org.opencb.biodata.models.variant.avro.VariantAvro.SCHEMA$);
    }
    
    /** Creates a Builder by copying an existing Builder */
    private Builder(org.opencb.biodata.models.variant.avro.VariantAvro.Builder other) {
      super(other);
      if (isValidValue(fields()[0], other.chromosome)) {
        this.chromosome = data().deepCopy(fields()[0].schema(), other.chromosome);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.start)) {
        this.start = data().deepCopy(fields()[1].schema(), other.start);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.end)) {
        this.end = data().deepCopy(fields()[2].schema(), other.end);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.reference)) {
        this.reference = data().deepCopy(fields()[3].schema(), other.reference);
        fieldSetFlags()[3] = true;
      }
      if (isValidValue(fields()[4], other.alternate)) {
        this.alternate = data().deepCopy(fields()[4].schema(), other.alternate);
        fieldSetFlags()[4] = true;
      }
      if (isValidValue(fields()[5], other.strand)) {
        this.strand = data().deepCopy(fields()[5].schema(), other.strand);
        fieldSetFlags()[5] = true;
      }
      if (isValidValue(fields()[6], other.ids)) {
        this.ids = data().deepCopy(fields()[6].schema(), other.ids);
        fieldSetFlags()[6] = true;
      }
      if (isValidValue(fields()[7], other.length)) {
        this.length = data().deepCopy(fields()[7].schema(), other.length);
        fieldSetFlags()[7] = true;
      }
      if (isValidValue(fields()[8], other.type)) {
        this.type = data().deepCopy(fields()[8].schema(), other.type);
        fieldSetFlags()[8] = true;
      }
      if (isValidValue(fields()[9], other.hgvs)) {
        this.hgvs = data().deepCopy(fields()[9].schema(), other.hgvs);
        fieldSetFlags()[9] = true;
      }
      if (isValidValue(fields()[10], other.studies)) {
        this.studies = data().deepCopy(fields()[10].schema(), other.studies);
        fieldSetFlags()[10] = true;
      }
      if (isValidValue(fields()[11], other.annotation)) {
        this.annotation = data().deepCopy(fields()[11].schema(), other.annotation);
        fieldSetFlags()[11] = true;
      }
    }
    
    /** Creates a Builder by copying an existing VariantAvro instance */
    private Builder(org.opencb.biodata.models.variant.avro.VariantAvro other) {
            super(org.opencb.biodata.models.variant.avro.VariantAvro.SCHEMA$);
      if (isValidValue(fields()[0], other.chromosome)) {
        this.chromosome = data().deepCopy(fields()[0].schema(), other.chromosome);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.start)) {
        this.start = data().deepCopy(fields()[1].schema(), other.start);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.end)) {
        this.end = data().deepCopy(fields()[2].schema(), other.end);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.reference)) {
        this.reference = data().deepCopy(fields()[3].schema(), other.reference);
        fieldSetFlags()[3] = true;
      }
      if (isValidValue(fields()[4], other.alternate)) {
        this.alternate = data().deepCopy(fields()[4].schema(), other.alternate);
        fieldSetFlags()[4] = true;
      }
      if (isValidValue(fields()[5], other.strand)) {
        this.strand = data().deepCopy(fields()[5].schema(), other.strand);
        fieldSetFlags()[5] = true;
      }
      if (isValidValue(fields()[6], other.ids)) {
        this.ids = data().deepCopy(fields()[6].schema(), other.ids);
        fieldSetFlags()[6] = true;
      }
      if (isValidValue(fields()[7], other.length)) {
        this.length = data().deepCopy(fields()[7].schema(), other.length);
        fieldSetFlags()[7] = true;
      }
      if (isValidValue(fields()[8], other.type)) {
        this.type = data().deepCopy(fields()[8].schema(), other.type);
        fieldSetFlags()[8] = true;
      }
      if (isValidValue(fields()[9], other.hgvs)) {
        this.hgvs = data().deepCopy(fields()[9].schema(), other.hgvs);
        fieldSetFlags()[9] = true;
      }
      if (isValidValue(fields()[10], other.studies)) {
        this.studies = data().deepCopy(fields()[10].schema(), other.studies);
        fieldSetFlags()[10] = true;
      }
      if (isValidValue(fields()[11], other.annotation)) {
        this.annotation = data().deepCopy(fields()[11].schema(), other.annotation);
        fieldSetFlags()[11] = true;
      }
    }

    /** Gets the value of the 'chromosome' field */
    public java.lang.String getChromosome() {
      return chromosome;
    }
    
    /** Sets the value of the 'chromosome' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setChromosome(java.lang.String value) {
      validate(fields()[0], value);
      this.chromosome = value;
      fieldSetFlags()[0] = true;
      return this; 
    }
    
    /** Checks whether the 'chromosome' field has been set */
    public boolean hasChromosome() {
      return fieldSetFlags()[0];
    }
    
    /** Clears the value of the 'chromosome' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearChromosome() {
      chromosome = null;
      fieldSetFlags()[0] = false;
      return this;
    }

    /** Gets the value of the 'start' field */
    public java.lang.Integer getStart() {
      return start;
    }
    
    /** Sets the value of the 'start' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setStart(int value) {
      validate(fields()[1], value);
      this.start = value;
      fieldSetFlags()[1] = true;
      return this; 
    }
    
    /** Checks whether the 'start' field has been set */
    public boolean hasStart() {
      return fieldSetFlags()[1];
    }
    
    /** Clears the value of the 'start' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearStart() {
      fieldSetFlags()[1] = false;
      return this;
    }

    /** Gets the value of the 'end' field */
    public java.lang.Integer getEnd() {
      return end;
    }
    
    /** Sets the value of the 'end' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setEnd(int value) {
      validate(fields()[2], value);
      this.end = value;
      fieldSetFlags()[2] = true;
      return this; 
    }
    
    /** Checks whether the 'end' field has been set */
    public boolean hasEnd() {
      return fieldSetFlags()[2];
    }
    
    /** Clears the value of the 'end' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearEnd() {
      fieldSetFlags()[2] = false;
      return this;
    }

    /** Gets the value of the 'reference' field */
    public java.lang.String getReference() {
      return reference;
    }
    
    /** Sets the value of the 'reference' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setReference(java.lang.String value) {
      validate(fields()[3], value);
      this.reference = value;
      fieldSetFlags()[3] = true;
      return this; 
    }
    
    /** Checks whether the 'reference' field has been set */
    public boolean hasReference() {
      return fieldSetFlags()[3];
    }
    
    /** Clears the value of the 'reference' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearReference() {
      reference = null;
      fieldSetFlags()[3] = false;
      return this;
    }

    /** Gets the value of the 'alternate' field */
    public java.lang.String getAlternate() {
      return alternate;
    }
    
    /** Sets the value of the 'alternate' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setAlternate(java.lang.String value) {
      validate(fields()[4], value);
      this.alternate = value;
      fieldSetFlags()[4] = true;
      return this; 
    }
    
    /** Checks whether the 'alternate' field has been set */
    public boolean hasAlternate() {
      return fieldSetFlags()[4];
    }
    
    /** Clears the value of the 'alternate' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearAlternate() {
      alternate = null;
      fieldSetFlags()[4] = false;
      return this;
    }

    /** Gets the value of the 'strand' field */
    public java.lang.String getStrand() {
      return strand;
    }
    
    /** Sets the value of the 'strand' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setStrand(java.lang.String value) {
      validate(fields()[5], value);
      this.strand = value;
      fieldSetFlags()[5] = true;
      return this; 
    }
    
    /** Checks whether the 'strand' field has been set */
    public boolean hasStrand() {
      return fieldSetFlags()[5];
    }
    
    /** Clears the value of the 'strand' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearStrand() {
      strand = null;
      fieldSetFlags()[5] = false;
      return this;
    }

    /** Gets the value of the 'ids' field */
    public java.util.List<java.lang.String> getIds() {
      return ids;
    }
    
    /** Sets the value of the 'ids' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setIds(java.util.List<java.lang.String> value) {
      validate(fields()[6], value);
      this.ids = value;
      fieldSetFlags()[6] = true;
      return this; 
    }
    
    /** Checks whether the 'ids' field has been set */
    public boolean hasIds() {
      return fieldSetFlags()[6];
    }
    
    /** Clears the value of the 'ids' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearIds() {
      ids = null;
      fieldSetFlags()[6] = false;
      return this;
    }

    /** Gets the value of the 'length' field */
    public java.lang.Integer getLength() {
      return length;
    }
    
    /** Sets the value of the 'length' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setLength(int value) {
      validate(fields()[7], value);
      this.length = value;
      fieldSetFlags()[7] = true;
      return this; 
    }
    
    /** Checks whether the 'length' field has been set */
    public boolean hasLength() {
      return fieldSetFlags()[7];
    }
    
    /** Clears the value of the 'length' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearLength() {
      fieldSetFlags()[7] = false;
      return this;
    }

    /** Gets the value of the 'type' field */
    public org.opencb.biodata.models.variant.avro.VariantType getType() {
      return type;
    }
    
    /** Sets the value of the 'type' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setType(org.opencb.biodata.models.variant.avro.VariantType value) {
      validate(fields()[8], value);
      this.type = value;
      fieldSetFlags()[8] = true;
      return this; 
    }
    
    /** Checks whether the 'type' field has been set */
    public boolean hasType() {
      return fieldSetFlags()[8];
    }
    
    /** Clears the value of the 'type' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearType() {
      type = null;
      fieldSetFlags()[8] = false;
      return this;
    }

    /** Gets the value of the 'hgvs' field */
    public java.util.Map<java.lang.String,java.util.List<java.lang.String>> getHgvs() {
      return hgvs;
    }
    
    /** Sets the value of the 'hgvs' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setHgvs(java.util.Map<java.lang.String,java.util.List<java.lang.String>> value) {
      validate(fields()[9], value);
      this.hgvs = value;
      fieldSetFlags()[9] = true;
      return this; 
    }
    
    /** Checks whether the 'hgvs' field has been set */
    public boolean hasHgvs() {
      return fieldSetFlags()[9];
    }
    
    /** Clears the value of the 'hgvs' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearHgvs() {
      hgvs = null;
      fieldSetFlags()[9] = false;
      return this;
    }

    /** Gets the value of the 'studies' field */
    public java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry> getStudies() {
      return studies;
    }
    
    /** Sets the value of the 'studies' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setStudies(java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry> value) {
      validate(fields()[10], value);
      this.studies = value;
      fieldSetFlags()[10] = true;
      return this; 
    }
    
    /** Checks whether the 'studies' field has been set */
    public boolean hasStudies() {
      return fieldSetFlags()[10];
    }
    
    /** Clears the value of the 'studies' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearStudies() {
      studies = null;
      fieldSetFlags()[10] = false;
      return this;
    }

    /** Gets the value of the 'annotation' field */
    public org.opencb.biodata.models.variant.avro.VariantAnnotation getAnnotation() {
      return annotation;
    }
    
    /** Sets the value of the 'annotation' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder setAnnotation(org.opencb.biodata.models.variant.avro.VariantAnnotation value) {
      validate(fields()[11], value);
      this.annotation = value;
      fieldSetFlags()[11] = true;
      return this; 
    }
    
    /** Checks whether the 'annotation' field has been set */
    public boolean hasAnnotation() {
      return fieldSetFlags()[11];
    }
    
    /** Clears the value of the 'annotation' field */
    public org.opencb.biodata.models.variant.avro.VariantAvro.Builder clearAnnotation() {
      annotation = null;
      fieldSetFlags()[11] = false;
      return this;
    }

    @Override
    public VariantAvro build() {
      try {
        VariantAvro record = new VariantAvro();
        record.chromosome = fieldSetFlags()[0] ? this.chromosome : (java.lang.String) defaultValue(fields()[0]);
        record.start = fieldSetFlags()[1] ? this.start : (java.lang.Integer) defaultValue(fields()[1]);
        record.end = fieldSetFlags()[2] ? this.end : (java.lang.Integer) defaultValue(fields()[2]);
        record.reference = fieldSetFlags()[3] ? this.reference : (java.lang.String) defaultValue(fields()[3]);
        record.alternate = fieldSetFlags()[4] ? this.alternate : (java.lang.String) defaultValue(fields()[4]);
        record.strand = fieldSetFlags()[5] ? this.strand : (java.lang.String) defaultValue(fields()[5]);
        record.ids = fieldSetFlags()[6] ? this.ids : (java.util.List<java.lang.String>) defaultValue(fields()[6]);
        record.length = fieldSetFlags()[7] ? this.length : (java.lang.Integer) defaultValue(fields()[7]);
        record.type = fieldSetFlags()[8] ? this.type : (org.opencb.biodata.models.variant.avro.VariantType) defaultValue(fields()[8]);
        record.hgvs = fieldSetFlags()[9] ? this.hgvs : (java.util.Map<java.lang.String,java.util.List<java.lang.String>>) defaultValue(fields()[9]);
        record.studies = fieldSetFlags()[10] ? this.studies : (java.util.List<org.opencb.biodata.models.variant.avro.StudyEntry>) defaultValue(fields()[10]);
        record.annotation = fieldSetFlags()[11] ? this.annotation : (org.opencb.biodata.models.variant.avro.VariantAnnotation) defaultValue(fields()[11]);
        return record;
      } catch (Exception e) {
        throw new org.apache.avro.AvroRuntimeException(e);
      }
    }
  }
}
