// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: variant.proto

package org.opencb.biodata.models.variant.protobuf;

public final class VariantProto {
  private VariantProto() {}
  public static void registerAllExtensions(
      com.google.protobuf.ExtensionRegistry registry) {
  }
  /**
   * Protobuf enum {@code org.opencb.biodata.models.variant.protobuf.VariantType}
   *
   * <pre>
   **
   * Type of variation, which depends mostly on its length.
   * &lt;ul&gt;
   * &lt;li&gt;SNVs involve a single nucleotide, without changes in length&lt;/li&gt;
   * &lt;li&gt;MNVs involve multiple nucleotides, without changes in length&lt;/li&gt;
   * &lt;li&gt;Indels are insertions or deletions of less than SV_THRESHOLD (50) nucleotides&lt;/li&gt;
   * &lt;li&gt;Structural variations are large changes of more than SV_THRESHOLD nucleotides&lt;/li&gt;
   * &lt;li&gt;Copy-number variations alter the number of copies of a region&lt;/li&gt;
   * &lt;/ul&gt;
   * </pre>
   */
  public enum VariantType
      implements com.google.protobuf.ProtocolMessageEnum {
    /**
     * <code>SNP = 0;</code>
     *
     * <pre>
     * SO:0000694
     * </pre>
     */
    SNP(0, 0),
    /**
     * <code>SNV = 1;</code>
     *
     * <pre>
     * SO:0001483
     * </pre>
     */
    SNV(1, 1),
    /**
     * <code>MNP = 2;</code>
     *
     * <pre>
     * SO:0001013
     * </pre>
     */
    MNP(2, 2),
    /**
     * <code>MNV = 3;</code>
     *
     * <pre>
     * SO:0002007
     * </pre>
     */
    MNV(3, 3),
    /**
     * <code>INDEL = 4;</code>
     *
     * <pre>
     * SO:1000032
     * </pre>
     */
    INDEL(4, 4),
    /**
     * <code>SV = 5;</code>
     *
     * <pre>
     * SO:0001537
     * </pre>
     */
    SV(5, 5),
    /**
     * <code>CNV = 6;</code>
     *
     * <pre>
     * SO:0001019
     * </pre>
     */
    CNV(6, 6),
    /**
     * <code>NO_VARIATION = 7;</code>
     *
     * <pre>
     * Defined in HTSJDK
     * </pre>
     */
    NO_VARIATION(7, 7),
    /**
     * <code>SYMBOLIC = 8;</code>
     *
     * <pre>
     * Defined in HTSJDK
     * </pre>
     */
    SYMBOLIC(8, 8),
    /**
     * <code>MIXED = 9;</code>
     *
     * <pre>
     * Defined in HTSJDK
     * </pre>
     */
    MIXED(9, 9),
    /**
     * <code>INSERTION = 10;</code>
     *
     * <pre>
     * SO:0000667
     * </pre>
     */
    INSERTION(10, 10),
    /**
     * <code>DELETION = 11;</code>
     *
     * <pre>
     * SO:0000159
     * </pre>
     */
    DELETION(11, 11),
    /**
     * <code>TRANSLOCATION = 12;</code>
     *
     * <pre>
     * SO:0000199
     * </pre>
     */
    TRANSLOCATION(12, 12),
    /**
     * <code>INVERSION = 13;</code>
     *
     * <pre>
     * SO:1000036
     * </pre>
     */
    INVERSION(13, 13),
    UNRECOGNIZED(-1, -1),
    ;

    /**
     * <code>SNP = 0;</code>
     *
     * <pre>
     * SO:0000694
     * </pre>
     */
    public static final int SNP_VALUE = 0;
    /**
     * <code>SNV = 1;</code>
     *
     * <pre>
     * SO:0001483
     * </pre>
     */
    public static final int SNV_VALUE = 1;
    /**
     * <code>MNP = 2;</code>
     *
     * <pre>
     * SO:0001013
     * </pre>
     */
    public static final int MNP_VALUE = 2;
    /**
     * <code>MNV = 3;</code>
     *
     * <pre>
     * SO:0002007
     * </pre>
     */
    public static final int MNV_VALUE = 3;
    /**
     * <code>INDEL = 4;</code>
     *
     * <pre>
     * SO:1000032
     * </pre>
     */
    public static final int INDEL_VALUE = 4;
    /**
     * <code>SV = 5;</code>
     *
     * <pre>
     * SO:0001537
     * </pre>
     */
    public static final int SV_VALUE = 5;
    /**
     * <code>CNV = 6;</code>
     *
     * <pre>
     * SO:0001019
     * </pre>
     */
    public static final int CNV_VALUE = 6;
    /**
     * <code>NO_VARIATION = 7;</code>
     *
     * <pre>
     * Defined in HTSJDK
     * </pre>
     */
    public static final int NO_VARIATION_VALUE = 7;
    /**
     * <code>SYMBOLIC = 8;</code>
     *
     * <pre>
     * Defined in HTSJDK
     * </pre>
     */
    public static final int SYMBOLIC_VALUE = 8;
    /**
     * <code>MIXED = 9;</code>
     *
     * <pre>
     * Defined in HTSJDK
     * </pre>
     */
    public static final int MIXED_VALUE = 9;
    /**
     * <code>INSERTION = 10;</code>
     *
     * <pre>
     * SO:0000667
     * </pre>
     */
    public static final int INSERTION_VALUE = 10;
    /**
     * <code>DELETION = 11;</code>
     *
     * <pre>
     * SO:0000159
     * </pre>
     */
    public static final int DELETION_VALUE = 11;
    /**
     * <code>TRANSLOCATION = 12;</code>
     *
     * <pre>
     * SO:0000199
     * </pre>
     */
    public static final int TRANSLOCATION_VALUE = 12;
    /**
     * <code>INVERSION = 13;</code>
     *
     * <pre>
     * SO:1000036
     * </pre>
     */
    public static final int INVERSION_VALUE = 13;


    public final int getNumber() {
      if (index == -1) {
        throw new java.lang.IllegalArgumentException(
            "Can't get the number of an unknown enum value.");
      }
      return value;
    }

    public static VariantType valueOf(int value) {
      switch (value) {
        case 0: return SNP;
        case 1: return SNV;
        case 2: return MNP;
        case 3: return MNV;
        case 4: return INDEL;
        case 5: return SV;
        case 6: return CNV;
        case 7: return NO_VARIATION;
        case 8: return SYMBOLIC;
        case 9: return MIXED;
        case 10: return INSERTION;
        case 11: return DELETION;
        case 12: return TRANSLOCATION;
        case 13: return INVERSION;
        default: return null;
      }
    }

    public static com.google.protobuf.Internal.EnumLiteMap<VariantType>
        internalGetValueMap() {
      return internalValueMap;
    }
    private static final com.google.protobuf.Internal.EnumLiteMap<
        VariantType> internalValueMap =
          new com.google.protobuf.Internal.EnumLiteMap<VariantType>() {
            public VariantType findValueByNumber(int number) {
              return VariantType.valueOf(number);
            }
          };

    public final com.google.protobuf.Descriptors.EnumValueDescriptor
        getValueDescriptor() {
      return getDescriptor().getValues().get(index);
    }
    public final com.google.protobuf.Descriptors.EnumDescriptor
        getDescriptorForType() {
      return getDescriptor();
    }
    public static final com.google.protobuf.Descriptors.EnumDescriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.getDescriptor().getEnumTypes().get(0);
    }

    private static final VariantType[] VALUES = values();

    public static VariantType valueOf(
        com.google.protobuf.Descriptors.EnumValueDescriptor desc) {
      if (desc.getType() != getDescriptor()) {
        throw new java.lang.IllegalArgumentException(
          "EnumValueDescriptor is not for this type.");
      }
      if (desc.getIndex() == -1) {
        return UNRECOGNIZED;
      }
      return VALUES[desc.getIndex()];
    }

    private final int index;
    private final int value;

    private VariantType(int index, int value) {
      this.index = index;
      this.value = value;
    }

    // @@protoc_insertion_point(enum_scope:org.opencb.biodata.models.variant.protobuf.VariantType)
  }

  /**
   * Protobuf enum {@code org.opencb.biodata.models.variant.protobuf.AllelesCode}
   */
  public enum AllelesCode
      implements com.google.protobuf.ProtocolMessageEnum {
    /**
     * <code>ALLELES_OK = 0;</code>
     */
    ALLELES_OK(0, 0),
    /**
     * <code>ALLELES_MISSING = 1;</code>
     */
    ALLELES_MISSING(1, 1),
    /**
     * <code>MULTIPLE_ALTERNATES = 2;</code>
     */
    MULTIPLE_ALTERNATES(2, 2),
    /**
     * <code>HAPLOID = 3;</code>
     */
    HAPLOID(3, 3),
    UNRECOGNIZED(-1, -1),
    ;

    /**
     * <code>ALLELES_OK = 0;</code>
     */
    public static final int ALLELES_OK_VALUE = 0;
    /**
     * <code>ALLELES_MISSING = 1;</code>
     */
    public static final int ALLELES_MISSING_VALUE = 1;
    /**
     * <code>MULTIPLE_ALTERNATES = 2;</code>
     */
    public static final int MULTIPLE_ALTERNATES_VALUE = 2;
    /**
     * <code>HAPLOID = 3;</code>
     */
    public static final int HAPLOID_VALUE = 3;


    public final int getNumber() {
      if (index == -1) {
        throw new java.lang.IllegalArgumentException(
            "Can't get the number of an unknown enum value.");
      }
      return value;
    }

    public static AllelesCode valueOf(int value) {
      switch (value) {
        case 0: return ALLELES_OK;
        case 1: return ALLELES_MISSING;
        case 2: return MULTIPLE_ALTERNATES;
        case 3: return HAPLOID;
        default: return null;
      }
    }

    public static com.google.protobuf.Internal.EnumLiteMap<AllelesCode>
        internalGetValueMap() {
      return internalValueMap;
    }
    private static final com.google.protobuf.Internal.EnumLiteMap<
        AllelesCode> internalValueMap =
          new com.google.protobuf.Internal.EnumLiteMap<AllelesCode>() {
            public AllelesCode findValueByNumber(int number) {
              return AllelesCode.valueOf(number);
            }
          };

    public final com.google.protobuf.Descriptors.EnumValueDescriptor
        getValueDescriptor() {
      return getDescriptor().getValues().get(index);
    }
    public final com.google.protobuf.Descriptors.EnumDescriptor
        getDescriptorForType() {
      return getDescriptor();
    }
    public static final com.google.protobuf.Descriptors.EnumDescriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.getDescriptor().getEnumTypes().get(1);
    }

    private static final AllelesCode[] VALUES = values();

    public static AllelesCode valueOf(
        com.google.protobuf.Descriptors.EnumValueDescriptor desc) {
      if (desc.getType() != getDescriptor()) {
        throw new java.lang.IllegalArgumentException(
          "EnumValueDescriptor is not for this type.");
      }
      if (desc.getIndex() == -1) {
        return UNRECOGNIZED;
      }
      return VALUES[desc.getIndex()];
    }

    private final int index;
    private final int value;

    private AllelesCode(int index, int value) {
      this.index = index;
      this.value = value;
    }

    // @@protoc_insertion_point(enum_scope:org.opencb.biodata.models.variant.protobuf.AllelesCode)
  }

  public interface VariantHardyWeinbergStatsOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional float chi2 = 1;</code>
     */
    float getChi2();

    /**
     * <code>optional float pValue = 2;</code>
     */
    float getPValue();

    /**
     * <code>optional int32 n = 3;</code>
     */
    int getN();

    /**
     * <code>optional int32 n_AA_11 = 4;</code>
     */
    int getNAA11();

    /**
     * <code>optional int32 n_Aa_10 = 5;</code>
     */
    int getNAa10();

    /**
     * <code>optional int32 n_aa_00 = 6;</code>
     */
    int getNAa00();

    /**
     * <code>optional float e_AA_11 = 7;</code>
     */
    float getEAA11();

    /**
     * <code>optional float e_Aa_10 = 8;</code>
     */
    float getEAa10();

    /**
     * <code>optional float e_aa_00 = 9;</code>
     */
    float getEAa00();

    /**
     * <code>optional float p = 10;</code>
     */
    float getP();

    /**
     * <code>optional float q = 11;</code>
     */
    float getQ();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats}
   */
  public  static final class VariantHardyWeinbergStats extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats)
      VariantHardyWeinbergStatsOrBuilder {
    // Use VariantHardyWeinbergStats.newBuilder() to construct.
    private VariantHardyWeinbergStats(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private VariantHardyWeinbergStats() {
      chi2_ = 0F;
      pValue_ = 0F;
      n_ = 0;
      nAA11_ = 0;
      nAa10_ = 0;
      nAa00_ = 0;
      eAA11_ = 0F;
      eAa10_ = 0F;
      eAa00_ = 0F;
      p_ = 0F;
      q_ = 0F;
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private VariantHardyWeinbergStats(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 13: {

              chi2_ = input.readFloat();
              break;
            }
            case 21: {

              pValue_ = input.readFloat();
              break;
            }
            case 24: {

              n_ = input.readInt32();
              break;
            }
            case 32: {

              nAA11_ = input.readInt32();
              break;
            }
            case 40: {

              nAa10_ = input.readInt32();
              break;
            }
            case 48: {

              nAa00_ = input.readInt32();
              break;
            }
            case 61: {

              eAA11_ = input.readFloat();
              break;
            }
            case 69: {

              eAa10_ = input.readFloat();
              break;
            }
            case 77: {

              eAa00_ = input.readFloat();
              break;
            }
            case 85: {

              p_ = input.readFloat();
              break;
            }
            case 93: {

              q_ = input.readFloat();
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder.class);
    }

    public static final int CHI2_FIELD_NUMBER = 1;
    private float chi2_;
    /**
     * <code>optional float chi2 = 1;</code>
     */
    public float getChi2() {
      return chi2_;
    }

    public static final int PVALUE_FIELD_NUMBER = 2;
    private float pValue_;
    /**
     * <code>optional float pValue = 2;</code>
     */
    public float getPValue() {
      return pValue_;
    }

    public static final int N_FIELD_NUMBER = 3;
    private int n_;
    /**
     * <code>optional int32 n = 3;</code>
     */
    public int getN() {
      return n_;
    }

    public static final int N_AA_11_FIELD_NUMBER = 4;
    private int nAA11_;
    /**
     * <code>optional int32 n_AA_11 = 4;</code>
     */
    public int getNAA11() {
      return nAA11_;
    }

    public static final int N_AA_10_FIELD_NUMBER = 5;
    private int nAa10_;
    /**
     * <code>optional int32 n_Aa_10 = 5;</code>
     */
    public int getNAa10() {
      return nAa10_;
    }

    public static final int N_AA_00_FIELD_NUMBER = 6;
    private int nAa00_;
    /**
     * <code>optional int32 n_aa_00 = 6;</code>
     */
    public int getNAa00() {
      return nAa00_;
    }

    public static final int E_AA_11_FIELD_NUMBER = 7;
    private float eAA11_;
    /**
     * <code>optional float e_AA_11 = 7;</code>
     */
    public float getEAA11() {
      return eAA11_;
    }

    public static final int E_AA_10_FIELD_NUMBER = 8;
    private float eAa10_;
    /**
     * <code>optional float e_Aa_10 = 8;</code>
     */
    public float getEAa10() {
      return eAa10_;
    }

    public static final int E_AA_00_FIELD_NUMBER = 9;
    private float eAa00_;
    /**
     * <code>optional float e_aa_00 = 9;</code>
     */
    public float getEAa00() {
      return eAa00_;
    }

    public static final int P_FIELD_NUMBER = 10;
    private float p_;
    /**
     * <code>optional float p = 10;</code>
     */
    public float getP() {
      return p_;
    }

    public static final int Q_FIELD_NUMBER = 11;
    private float q_;
    /**
     * <code>optional float q = 11;</code>
     */
    public float getQ() {
      return q_;
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (chi2_ != 0F) {
        output.writeFloat(1, chi2_);
      }
      if (pValue_ != 0F) {
        output.writeFloat(2, pValue_);
      }
      if (n_ != 0) {
        output.writeInt32(3, n_);
      }
      if (nAA11_ != 0) {
        output.writeInt32(4, nAA11_);
      }
      if (nAa10_ != 0) {
        output.writeInt32(5, nAa10_);
      }
      if (nAa00_ != 0) {
        output.writeInt32(6, nAa00_);
      }
      if (eAA11_ != 0F) {
        output.writeFloat(7, eAA11_);
      }
      if (eAa10_ != 0F) {
        output.writeFloat(8, eAa10_);
      }
      if (eAa00_ != 0F) {
        output.writeFloat(9, eAa00_);
      }
      if (p_ != 0F) {
        output.writeFloat(10, p_);
      }
      if (q_ != 0F) {
        output.writeFloat(11, q_);
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (chi2_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(1, chi2_);
      }
      if (pValue_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(2, pValue_);
      }
      if (n_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(3, n_);
      }
      if (nAA11_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(4, nAA11_);
      }
      if (nAa10_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(5, nAa10_);
      }
      if (nAa00_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(6, nAa00_);
      }
      if (eAA11_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(7, eAA11_);
      }
      if (eAa10_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(8, eAa10_);
      }
      if (eAa00_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(9, eAa00_);
      }
      if (p_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(10, p_);
      }
      if (q_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(11, q_);
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats)
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStatsOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        chi2_ = 0F;

        pValue_ = 0F;

        n_ = 0;

        nAA11_ = 0;

        nAa10_ = 0;

        nAa00_ = 0;

        eAA11_ = 0F;

        eAa10_ = 0F;

        eAa00_ = 0F;

        p_ = 0F;

        q_ = 0F;

        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats build() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats result = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats(this);
        result.chi2_ = chi2_;
        result.pValue_ = pValue_;
        result.n_ = n_;
        result.nAA11_ = nAA11_;
        result.nAa10_ = nAa10_;
        result.nAa00_ = nAa00_;
        result.eAA11_ = eAA11_;
        result.eAa10_ = eAa10_;
        result.eAa00_ = eAa00_;
        result.p_ = p_;
        result.q_ = q_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.getDefaultInstance()) return this;
        if (other.getChi2() != 0F) {
          setChi2(other.getChi2());
        }
        if (other.getPValue() != 0F) {
          setPValue(other.getPValue());
        }
        if (other.getN() != 0) {
          setN(other.getN());
        }
        if (other.getNAA11() != 0) {
          setNAA11(other.getNAA11());
        }
        if (other.getNAa10() != 0) {
          setNAa10(other.getNAa10());
        }
        if (other.getNAa00() != 0) {
          setNAa00(other.getNAa00());
        }
        if (other.getEAA11() != 0F) {
          setEAA11(other.getEAA11());
        }
        if (other.getEAa10() != 0F) {
          setEAa10(other.getEAa10());
        }
        if (other.getEAa00() != 0F) {
          setEAa00(other.getEAa00());
        }
        if (other.getP() != 0F) {
          setP(other.getP());
        }
        if (other.getQ() != 0F) {
          setQ(other.getQ());
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }

      private float chi2_ ;
      /**
       * <code>optional float chi2 = 1;</code>
       */
      public float getChi2() {
        return chi2_;
      }
      /**
       * <code>optional float chi2 = 1;</code>
       */
      public Builder setChi2(float value) {
        
        chi2_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float chi2 = 1;</code>
       */
      public Builder clearChi2() {
        
        chi2_ = 0F;
        onChanged();
        return this;
      }

      private float pValue_ ;
      /**
       * <code>optional float pValue = 2;</code>
       */
      public float getPValue() {
        return pValue_;
      }
      /**
       * <code>optional float pValue = 2;</code>
       */
      public Builder setPValue(float value) {
        
        pValue_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float pValue = 2;</code>
       */
      public Builder clearPValue() {
        
        pValue_ = 0F;
        onChanged();
        return this;
      }

      private int n_ ;
      /**
       * <code>optional int32 n = 3;</code>
       */
      public int getN() {
        return n_;
      }
      /**
       * <code>optional int32 n = 3;</code>
       */
      public Builder setN(int value) {
        
        n_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 n = 3;</code>
       */
      public Builder clearN() {
        
        n_ = 0;
        onChanged();
        return this;
      }

      private int nAA11_ ;
      /**
       * <code>optional int32 n_AA_11 = 4;</code>
       */
      public int getNAA11() {
        return nAA11_;
      }
      /**
       * <code>optional int32 n_AA_11 = 4;</code>
       */
      public Builder setNAA11(int value) {
        
        nAA11_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 n_AA_11 = 4;</code>
       */
      public Builder clearNAA11() {
        
        nAA11_ = 0;
        onChanged();
        return this;
      }

      private int nAa10_ ;
      /**
       * <code>optional int32 n_Aa_10 = 5;</code>
       */
      public int getNAa10() {
        return nAa10_;
      }
      /**
       * <code>optional int32 n_Aa_10 = 5;</code>
       */
      public Builder setNAa10(int value) {
        
        nAa10_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 n_Aa_10 = 5;</code>
       */
      public Builder clearNAa10() {
        
        nAa10_ = 0;
        onChanged();
        return this;
      }

      private int nAa00_ ;
      /**
       * <code>optional int32 n_aa_00 = 6;</code>
       */
      public int getNAa00() {
        return nAa00_;
      }
      /**
       * <code>optional int32 n_aa_00 = 6;</code>
       */
      public Builder setNAa00(int value) {
        
        nAa00_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 n_aa_00 = 6;</code>
       */
      public Builder clearNAa00() {
        
        nAa00_ = 0;
        onChanged();
        return this;
      }

      private float eAA11_ ;
      /**
       * <code>optional float e_AA_11 = 7;</code>
       */
      public float getEAA11() {
        return eAA11_;
      }
      /**
       * <code>optional float e_AA_11 = 7;</code>
       */
      public Builder setEAA11(float value) {
        
        eAA11_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float e_AA_11 = 7;</code>
       */
      public Builder clearEAA11() {
        
        eAA11_ = 0F;
        onChanged();
        return this;
      }

      private float eAa10_ ;
      /**
       * <code>optional float e_Aa_10 = 8;</code>
       */
      public float getEAa10() {
        return eAa10_;
      }
      /**
       * <code>optional float e_Aa_10 = 8;</code>
       */
      public Builder setEAa10(float value) {
        
        eAa10_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float e_Aa_10 = 8;</code>
       */
      public Builder clearEAa10() {
        
        eAa10_ = 0F;
        onChanged();
        return this;
      }

      private float eAa00_ ;
      /**
       * <code>optional float e_aa_00 = 9;</code>
       */
      public float getEAa00() {
        return eAa00_;
      }
      /**
       * <code>optional float e_aa_00 = 9;</code>
       */
      public Builder setEAa00(float value) {
        
        eAa00_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float e_aa_00 = 9;</code>
       */
      public Builder clearEAa00() {
        
        eAa00_ = 0F;
        onChanged();
        return this;
      }

      private float p_ ;
      /**
       * <code>optional float p = 10;</code>
       */
      public float getP() {
        return p_;
      }
      /**
       * <code>optional float p = 10;</code>
       */
      public Builder setP(float value) {
        
        p_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float p = 10;</code>
       */
      public Builder clearP() {
        
        p_ = 0F;
        onChanged();
        return this;
      }

      private float q_ ;
      /**
       * <code>optional float q = 11;</code>
       */
      public float getQ() {
        return q_;
      }
      /**
       * <code>optional float q = 11;</code>
       */
      public Builder setQ(float value) {
        
        q_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float q = 11;</code>
       */
      public Builder clearQ() {
        
        q_ = 0F;
        onChanged();
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats)
    private static final org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats();
    }

    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<VariantHardyWeinbergStats>
        PARSER = new com.google.protobuf.AbstractParser<VariantHardyWeinbergStats>() {
      public VariantHardyWeinbergStats parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new VariantHardyWeinbergStats(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<VariantHardyWeinbergStats> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<VariantHardyWeinbergStats> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface VariantStatsOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VariantStats)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional string refAllele = 1;</code>
     */
    java.lang.String getRefAllele();
    /**
     * <code>optional string refAllele = 1;</code>
     */
    com.google.protobuf.ByteString
        getRefAlleleBytes();

    /**
     * <code>optional string altAllele = 2;</code>
     */
    java.lang.String getAltAllele();
    /**
     * <code>optional string altAllele = 2;</code>
     */
    com.google.protobuf.ByteString
        getAltAlleleBytes();

    /**
     * <code>optional int32 refAlleleCount = 3;</code>
     */
    int getRefAlleleCount();

    /**
     * <code>optional int32 altAlleleCount = 4;</code>
     */
    int getAltAlleleCount();

    /**
     * <code>map&lt;string, int32&gt; genotypesCount = 5;</code>
     */
    java.util.Map<java.lang.String, java.lang.Integer>
    getGenotypesCount();

    /**
     * <code>map&lt;string, float&gt; genotypesFreq = 6;</code>
     */
    java.util.Map<java.lang.String, java.lang.Float>
    getGenotypesFreq();

    /**
     * <code>optional int32 missingAlleles = 7;</code>
     */
    int getMissingAlleles();

    /**
     * <code>optional int32 missingGenotypes = 8;</code>
     */
    int getMissingGenotypes();

    /**
     * <code>optional float refAlleleFreq = 9;</code>
     */
    float getRefAlleleFreq();

    /**
     * <code>optional float altAlleleFreq = 10;</code>
     */
    float getAltAlleleFreq();

    /**
     * <code>optional float maf = 11;</code>
     */
    float getMaf();

    /**
     * <code>optional float mgf = 12;</code>
     */
    float getMgf();

    /**
     * <code>optional string mafAllele = 13;</code>
     */
    java.lang.String getMafAllele();
    /**
     * <code>optional string mafAllele = 13;</code>
     */
    com.google.protobuf.ByteString
        getMafAlleleBytes();

    /**
     * <code>optional string mgfGenotype = 14;</code>
     */
    java.lang.String getMgfGenotype();
    /**
     * <code>optional string mgfGenotype = 14;</code>
     */
    com.google.protobuf.ByteString
        getMgfGenotypeBytes();

    /**
     * <code>optional bool passedFilters = 15;</code>
     */
    boolean getPassedFilters();

    /**
     * <code>optional int32 mendelianErrors = 16;</code>
     */
    int getMendelianErrors();

    /**
     * <code>optional float casesPercentDominant = 17;</code>
     */
    float getCasesPercentDominant();

    /**
     * <code>optional float controlsPercentDominant = 18;</code>
     */
    float getControlsPercentDominant();

    /**
     * <code>optional float casesPercentRecessive = 19;</code>
     */
    float getCasesPercentRecessive();

    /**
     * <code>optional float controlsPercentRecessive = 20;</code>
     */
    float getControlsPercentRecessive();

    /**
     * <code>optional float quality = 21;</code>
     */
    float getQuality();

    /**
     * <code>optional int32 numSamples = 22;</code>
     */
    int getNumSamples();

    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
     */
    int getVariantTypeValue();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getVariantType();

    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
     */
    boolean hasHw();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats getHw();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStatsOrBuilder getHwOrBuilder();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantStats}
   */
  public  static final class VariantStats extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VariantStats)
      VariantStatsOrBuilder {
    // Use VariantStats.newBuilder() to construct.
    private VariantStats(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private VariantStats() {
      refAllele_ = "";
      altAllele_ = "";
      refAlleleCount_ = 0;
      altAlleleCount_ = 0;
      missingAlleles_ = 0;
      missingGenotypes_ = 0;
      refAlleleFreq_ = 0F;
      altAlleleFreq_ = 0F;
      maf_ = 0F;
      mgf_ = 0F;
      mafAllele_ = "";
      mgfGenotype_ = "";
      passedFilters_ = false;
      mendelianErrors_ = 0;
      casesPercentDominant_ = 0F;
      controlsPercentDominant_ = 0F;
      casesPercentRecessive_ = 0F;
      controlsPercentRecessive_ = 0F;
      quality_ = 0F;
      numSamples_ = 0;
      variantType_ = 0;
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private VariantStats(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();

              refAllele_ = s;
              break;
            }
            case 18: {
              String s = input.readStringRequireUtf8();

              altAllele_ = s;
              break;
            }
            case 24: {

              refAlleleCount_ = input.readInt32();
              break;
            }
            case 32: {

              altAlleleCount_ = input.readInt32();
              break;
            }
            case 42: {
              if (!((mutable_bitField0_ & 0x00000010) == 0x00000010)) {
                genotypesCount_ = com.google.protobuf.MapField.newMapField(
                    GenotypesCountDefaultEntryHolder.defaultEntry);
                mutable_bitField0_ |= 0x00000010;
              }
              com.google.protobuf.MapEntry<java.lang.String, java.lang.Integer>
              genotypesCount = input.readMessage(
                  GenotypesCountDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
              genotypesCount_.getMutableMap().put(genotypesCount.getKey(), genotypesCount.getValue());
              break;
            }
            case 50: {
              if (!((mutable_bitField0_ & 0x00000020) == 0x00000020)) {
                genotypesFreq_ = com.google.protobuf.MapField.newMapField(
                    GenotypesFreqDefaultEntryHolder.defaultEntry);
                mutable_bitField0_ |= 0x00000020;
              }
              com.google.protobuf.MapEntry<java.lang.String, java.lang.Float>
              genotypesFreq = input.readMessage(
                  GenotypesFreqDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
              genotypesFreq_.getMutableMap().put(genotypesFreq.getKey(), genotypesFreq.getValue());
              break;
            }
            case 56: {

              missingAlleles_ = input.readInt32();
              break;
            }
            case 64: {

              missingGenotypes_ = input.readInt32();
              break;
            }
            case 77: {

              refAlleleFreq_ = input.readFloat();
              break;
            }
            case 85: {

              altAlleleFreq_ = input.readFloat();
              break;
            }
            case 93: {

              maf_ = input.readFloat();
              break;
            }
            case 101: {

              mgf_ = input.readFloat();
              break;
            }
            case 106: {
              String s = input.readStringRequireUtf8();

              mafAllele_ = s;
              break;
            }
            case 114: {
              String s = input.readStringRequireUtf8();

              mgfGenotype_ = s;
              break;
            }
            case 120: {

              passedFilters_ = input.readBool();
              break;
            }
            case 128: {

              mendelianErrors_ = input.readInt32();
              break;
            }
            case 141: {

              casesPercentDominant_ = input.readFloat();
              break;
            }
            case 149: {

              controlsPercentDominant_ = input.readFloat();
              break;
            }
            case 157: {

              casesPercentRecessive_ = input.readFloat();
              break;
            }
            case 165: {

              controlsPercentRecessive_ = input.readFloat();
              break;
            }
            case 173: {

              quality_ = input.readFloat();
              break;
            }
            case 176: {

              numSamples_ = input.readInt32();
              break;
            }
            case 184: {
              int rawValue = input.readEnum();

              variantType_ = rawValue;
              break;
            }
            case 194: {
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder subBuilder = null;
              if (hw_ != null) {
                subBuilder = hw_.toBuilder();
              }
              hw_ = input.readMessage(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.parser(), extensionRegistry);
              if (subBuilder != null) {
                subBuilder.mergeFrom(hw_);
                hw_ = subBuilder.buildPartial();
              }

              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 5:
          return internalGetGenotypesCount();
        case 6:
          return internalGetGenotypesFreq();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.Builder.class);
    }

    private int bitField0_;
    public static final int REFALLELE_FIELD_NUMBER = 1;
    private volatile java.lang.Object refAllele_;
    /**
     * <code>optional string refAllele = 1;</code>
     */
    public java.lang.String getRefAllele() {
      java.lang.Object ref = refAllele_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        refAllele_ = s;
        return s;
      }
    }
    /**
     * <code>optional string refAllele = 1;</code>
     */
    public com.google.protobuf.ByteString
        getRefAlleleBytes() {
      java.lang.Object ref = refAllele_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        refAllele_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int ALTALLELE_FIELD_NUMBER = 2;
    private volatile java.lang.Object altAllele_;
    /**
     * <code>optional string altAllele = 2;</code>
     */
    public java.lang.String getAltAllele() {
      java.lang.Object ref = altAllele_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        altAllele_ = s;
        return s;
      }
    }
    /**
     * <code>optional string altAllele = 2;</code>
     */
    public com.google.protobuf.ByteString
        getAltAlleleBytes() {
      java.lang.Object ref = altAllele_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        altAllele_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int REFALLELECOUNT_FIELD_NUMBER = 3;
    private int refAlleleCount_;
    /**
     * <code>optional int32 refAlleleCount = 3;</code>
     */
    public int getRefAlleleCount() {
      return refAlleleCount_;
    }

    public static final int ALTALLELECOUNT_FIELD_NUMBER = 4;
    private int altAlleleCount_;
    /**
     * <code>optional int32 altAlleleCount = 4;</code>
     */
    public int getAltAlleleCount() {
      return altAlleleCount_;
    }

    public static final int GENOTYPESCOUNT_FIELD_NUMBER = 5;
    private static final class GenotypesCountDefaultEntryHolder {
      static final com.google.protobuf.MapEntry<
          java.lang.String, java.lang.Integer> defaultEntry =
              com.google.protobuf.MapEntry
              .<java.lang.String, java.lang.Integer>newDefaultInstance(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesCountEntry_descriptor, 
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "",
                  com.google.protobuf.WireFormat.FieldType.INT32,
                  0);
    }
    private com.google.protobuf.MapField<
        java.lang.String, java.lang.Integer> genotypesCount_;
    private com.google.protobuf.MapField<java.lang.String, java.lang.Integer>
    internalGetGenotypesCount() {
      if (genotypesCount_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            GenotypesCountDefaultEntryHolder.defaultEntry);
     }
      return genotypesCount_;
    }
    /**
     * <code>map&lt;string, int32&gt; genotypesCount = 5;</code>
     */

    public java.util.Map<java.lang.String, java.lang.Integer> getGenotypesCount() {
      return internalGetGenotypesCount().getMap();
    }

    public static final int GENOTYPESFREQ_FIELD_NUMBER = 6;
    private static final class GenotypesFreqDefaultEntryHolder {
      static final com.google.protobuf.MapEntry<
          java.lang.String, java.lang.Float> defaultEntry =
              com.google.protobuf.MapEntry
              .<java.lang.String, java.lang.Float>newDefaultInstance(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesFreqEntry_descriptor, 
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "",
                  com.google.protobuf.WireFormat.FieldType.FLOAT,
                  0F);
    }
    private com.google.protobuf.MapField<
        java.lang.String, java.lang.Float> genotypesFreq_;
    private com.google.protobuf.MapField<java.lang.String, java.lang.Float>
    internalGetGenotypesFreq() {
      if (genotypesFreq_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            GenotypesFreqDefaultEntryHolder.defaultEntry);
     }
      return genotypesFreq_;
    }
    /**
     * <code>map&lt;string, float&gt; genotypesFreq = 6;</code>
     */

    public java.util.Map<java.lang.String, java.lang.Float> getGenotypesFreq() {
      return internalGetGenotypesFreq().getMap();
    }

    public static final int MISSINGALLELES_FIELD_NUMBER = 7;
    private int missingAlleles_;
    /**
     * <code>optional int32 missingAlleles = 7;</code>
     */
    public int getMissingAlleles() {
      return missingAlleles_;
    }

    public static final int MISSINGGENOTYPES_FIELD_NUMBER = 8;
    private int missingGenotypes_;
    /**
     * <code>optional int32 missingGenotypes = 8;</code>
     */
    public int getMissingGenotypes() {
      return missingGenotypes_;
    }

    public static final int REFALLELEFREQ_FIELD_NUMBER = 9;
    private float refAlleleFreq_;
    /**
     * <code>optional float refAlleleFreq = 9;</code>
     */
    public float getRefAlleleFreq() {
      return refAlleleFreq_;
    }

    public static final int ALTALLELEFREQ_FIELD_NUMBER = 10;
    private float altAlleleFreq_;
    /**
     * <code>optional float altAlleleFreq = 10;</code>
     */
    public float getAltAlleleFreq() {
      return altAlleleFreq_;
    }

    public static final int MAF_FIELD_NUMBER = 11;
    private float maf_;
    /**
     * <code>optional float maf = 11;</code>
     */
    public float getMaf() {
      return maf_;
    }

    public static final int MGF_FIELD_NUMBER = 12;
    private float mgf_;
    /**
     * <code>optional float mgf = 12;</code>
     */
    public float getMgf() {
      return mgf_;
    }

    public static final int MAFALLELE_FIELD_NUMBER = 13;
    private volatile java.lang.Object mafAllele_;
    /**
     * <code>optional string mafAllele = 13;</code>
     */
    public java.lang.String getMafAllele() {
      java.lang.Object ref = mafAllele_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        mafAllele_ = s;
        return s;
      }
    }
    /**
     * <code>optional string mafAllele = 13;</code>
     */
    public com.google.protobuf.ByteString
        getMafAlleleBytes() {
      java.lang.Object ref = mafAllele_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        mafAllele_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int MGFGENOTYPE_FIELD_NUMBER = 14;
    private volatile java.lang.Object mgfGenotype_;
    /**
     * <code>optional string mgfGenotype = 14;</code>
     */
    public java.lang.String getMgfGenotype() {
      java.lang.Object ref = mgfGenotype_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        mgfGenotype_ = s;
        return s;
      }
    }
    /**
     * <code>optional string mgfGenotype = 14;</code>
     */
    public com.google.protobuf.ByteString
        getMgfGenotypeBytes() {
      java.lang.Object ref = mgfGenotype_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        mgfGenotype_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int PASSEDFILTERS_FIELD_NUMBER = 15;
    private boolean passedFilters_;
    /**
     * <code>optional bool passedFilters = 15;</code>
     */
    public boolean getPassedFilters() {
      return passedFilters_;
    }

    public static final int MENDELIANERRORS_FIELD_NUMBER = 16;
    private int mendelianErrors_;
    /**
     * <code>optional int32 mendelianErrors = 16;</code>
     */
    public int getMendelianErrors() {
      return mendelianErrors_;
    }

    public static final int CASESPERCENTDOMINANT_FIELD_NUMBER = 17;
    private float casesPercentDominant_;
    /**
     * <code>optional float casesPercentDominant = 17;</code>
     */
    public float getCasesPercentDominant() {
      return casesPercentDominant_;
    }

    public static final int CONTROLSPERCENTDOMINANT_FIELD_NUMBER = 18;
    private float controlsPercentDominant_;
    /**
     * <code>optional float controlsPercentDominant = 18;</code>
     */
    public float getControlsPercentDominant() {
      return controlsPercentDominant_;
    }

    public static final int CASESPERCENTRECESSIVE_FIELD_NUMBER = 19;
    private float casesPercentRecessive_;
    /**
     * <code>optional float casesPercentRecessive = 19;</code>
     */
    public float getCasesPercentRecessive() {
      return casesPercentRecessive_;
    }

    public static final int CONTROLSPERCENTRECESSIVE_FIELD_NUMBER = 20;
    private float controlsPercentRecessive_;
    /**
     * <code>optional float controlsPercentRecessive = 20;</code>
     */
    public float getControlsPercentRecessive() {
      return controlsPercentRecessive_;
    }

    public static final int QUALITY_FIELD_NUMBER = 21;
    private float quality_;
    /**
     * <code>optional float quality = 21;</code>
     */
    public float getQuality() {
      return quality_;
    }

    public static final int NUMSAMPLES_FIELD_NUMBER = 22;
    private int numSamples_;
    /**
     * <code>optional int32 numSamples = 22;</code>
     */
    public int getNumSamples() {
      return numSamples_;
    }

    public static final int VARIANTTYPE_FIELD_NUMBER = 23;
    private int variantType_;
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
     */
    public int getVariantTypeValue() {
      return variantType_;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getVariantType() {
      org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType result = org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.valueOf(variantType_);
      return result == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.UNRECOGNIZED : result;
    }

    public static final int HW_FIELD_NUMBER = 24;
    private org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats hw_;
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
     */
    public boolean hasHw() {
      return hw_ != null;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats getHw() {
      return hw_ == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.getDefaultInstance() : hw_;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStatsOrBuilder getHwOrBuilder() {
      return getHw();
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!getRefAlleleBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, refAllele_);
      }
      if (!getAltAlleleBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 2, altAllele_);
      }
      if (refAlleleCount_ != 0) {
        output.writeInt32(3, refAlleleCount_);
      }
      if (altAlleleCount_ != 0) {
        output.writeInt32(4, altAlleleCount_);
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.Integer> entry
           : internalGetGenotypesCount().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.Integer>
        genotypesCount = GenotypesCountDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        output.writeMessage(5, genotypesCount);
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.Float> entry
           : internalGetGenotypesFreq().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.Float>
        genotypesFreq = GenotypesFreqDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        output.writeMessage(6, genotypesFreq);
      }
      if (missingAlleles_ != 0) {
        output.writeInt32(7, missingAlleles_);
      }
      if (missingGenotypes_ != 0) {
        output.writeInt32(8, missingGenotypes_);
      }
      if (refAlleleFreq_ != 0F) {
        output.writeFloat(9, refAlleleFreq_);
      }
      if (altAlleleFreq_ != 0F) {
        output.writeFloat(10, altAlleleFreq_);
      }
      if (maf_ != 0F) {
        output.writeFloat(11, maf_);
      }
      if (mgf_ != 0F) {
        output.writeFloat(12, mgf_);
      }
      if (!getMafAlleleBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 13, mafAllele_);
      }
      if (!getMgfGenotypeBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 14, mgfGenotype_);
      }
      if (passedFilters_ != false) {
        output.writeBool(15, passedFilters_);
      }
      if (mendelianErrors_ != 0) {
        output.writeInt32(16, mendelianErrors_);
      }
      if (casesPercentDominant_ != 0F) {
        output.writeFloat(17, casesPercentDominant_);
      }
      if (controlsPercentDominant_ != 0F) {
        output.writeFloat(18, controlsPercentDominant_);
      }
      if (casesPercentRecessive_ != 0F) {
        output.writeFloat(19, casesPercentRecessive_);
      }
      if (controlsPercentRecessive_ != 0F) {
        output.writeFloat(20, controlsPercentRecessive_);
      }
      if (quality_ != 0F) {
        output.writeFloat(21, quality_);
      }
      if (numSamples_ != 0) {
        output.writeInt32(22, numSamples_);
      }
      if (variantType_ != org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.SNP.getNumber()) {
        output.writeEnum(23, variantType_);
      }
      if (hw_ != null) {
        output.writeMessage(24, getHw());
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!getRefAlleleBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(1, refAllele_);
      }
      if (!getAltAlleleBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(2, altAllele_);
      }
      if (refAlleleCount_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(3, refAlleleCount_);
      }
      if (altAlleleCount_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(4, altAlleleCount_);
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.Integer> entry
           : internalGetGenotypesCount().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.Integer>
        genotypesCount = GenotypesCountDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        size += com.google.protobuf.CodedOutputStream
            .computeMessageSize(5, genotypesCount);
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.Float> entry
           : internalGetGenotypesFreq().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.Float>
        genotypesFreq = GenotypesFreqDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        size += com.google.protobuf.CodedOutputStream
            .computeMessageSize(6, genotypesFreq);
      }
      if (missingAlleles_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(7, missingAlleles_);
      }
      if (missingGenotypes_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(8, missingGenotypes_);
      }
      if (refAlleleFreq_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(9, refAlleleFreq_);
      }
      if (altAlleleFreq_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(10, altAlleleFreq_);
      }
      if (maf_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(11, maf_);
      }
      if (mgf_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(12, mgf_);
      }
      if (!getMafAlleleBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(13, mafAllele_);
      }
      if (!getMgfGenotypeBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(14, mgfGenotype_);
      }
      if (passedFilters_ != false) {
        size += com.google.protobuf.CodedOutputStream
          .computeBoolSize(15, passedFilters_);
      }
      if (mendelianErrors_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(16, mendelianErrors_);
      }
      if (casesPercentDominant_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(17, casesPercentDominant_);
      }
      if (controlsPercentDominant_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(18, controlsPercentDominant_);
      }
      if (casesPercentRecessive_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(19, casesPercentRecessive_);
      }
      if (controlsPercentRecessive_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(20, controlsPercentRecessive_);
      }
      if (quality_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(21, quality_);
      }
      if (numSamples_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(22, numSamples_);
      }
      if (variantType_ != org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.SNP.getNumber()) {
        size += com.google.protobuf.CodedOutputStream
          .computeEnumSize(23, variantType_);
      }
      if (hw_ != null) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(24, getHw());
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantStats}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VariantStats)
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStatsOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor;
      }

      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMapField(
          int number) {
        switch (number) {
          case 5:
            return internalGetGenotypesCount();
          case 6:
            return internalGetGenotypesFreq();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMutableMapField(
          int number) {
        switch (number) {
          case 5:
            return internalGetMutableGenotypesCount();
          case 6:
            return internalGetMutableGenotypesFreq();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        refAllele_ = "";

        altAllele_ = "";

        refAlleleCount_ = 0;

        altAlleleCount_ = 0;

        internalGetMutableGenotypesCount().clear();
        internalGetMutableGenotypesFreq().clear();
        missingAlleles_ = 0;

        missingGenotypes_ = 0;

        refAlleleFreq_ = 0F;

        altAlleleFreq_ = 0F;

        maf_ = 0F;

        mgf_ = 0F;

        mafAllele_ = "";

        mgfGenotype_ = "";

        passedFilters_ = false;

        mendelianErrors_ = 0;

        casesPercentDominant_ = 0F;

        controlsPercentDominant_ = 0F;

        casesPercentRecessive_ = 0F;

        controlsPercentRecessive_ = 0F;

        quality_ = 0F;

        numSamples_ = 0;

        variantType_ = 0;

        if (hwBuilder_ == null) {
          hw_ = null;
        } else {
          hw_ = null;
          hwBuilder_ = null;
        }
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats build() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats result = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.refAllele_ = refAllele_;
        result.altAllele_ = altAllele_;
        result.refAlleleCount_ = refAlleleCount_;
        result.altAlleleCount_ = altAlleleCount_;
        result.genotypesCount_ = internalGetGenotypesCount();
        result.genotypesCount_.makeImmutable();
        result.genotypesFreq_ = internalGetGenotypesFreq();
        result.genotypesFreq_.makeImmutable();
        result.missingAlleles_ = missingAlleles_;
        result.missingGenotypes_ = missingGenotypes_;
        result.refAlleleFreq_ = refAlleleFreq_;
        result.altAlleleFreq_ = altAlleleFreq_;
        result.maf_ = maf_;
        result.mgf_ = mgf_;
        result.mafAllele_ = mafAllele_;
        result.mgfGenotype_ = mgfGenotype_;
        result.passedFilters_ = passedFilters_;
        result.mendelianErrors_ = mendelianErrors_;
        result.casesPercentDominant_ = casesPercentDominant_;
        result.controlsPercentDominant_ = controlsPercentDominant_;
        result.casesPercentRecessive_ = casesPercentRecessive_;
        result.controlsPercentRecessive_ = controlsPercentRecessive_;
        result.quality_ = quality_;
        result.numSamples_ = numSamples_;
        result.variantType_ = variantType_;
        if (hwBuilder_ == null) {
          result.hw_ = hw_;
        } else {
          result.hw_ = hwBuilder_.build();
        }
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.getDefaultInstance()) return this;
        if (!other.getRefAllele().isEmpty()) {
          refAllele_ = other.refAllele_;
          onChanged();
        }
        if (!other.getAltAllele().isEmpty()) {
          altAllele_ = other.altAllele_;
          onChanged();
        }
        if (other.getRefAlleleCount() != 0) {
          setRefAlleleCount(other.getRefAlleleCount());
        }
        if (other.getAltAlleleCount() != 0) {
          setAltAlleleCount(other.getAltAlleleCount());
        }
        internalGetMutableGenotypesCount().mergeFrom(
            other.internalGetGenotypesCount());
        internalGetMutableGenotypesFreq().mergeFrom(
            other.internalGetGenotypesFreq());
        if (other.getMissingAlleles() != 0) {
          setMissingAlleles(other.getMissingAlleles());
        }
        if (other.getMissingGenotypes() != 0) {
          setMissingGenotypes(other.getMissingGenotypes());
        }
        if (other.getRefAlleleFreq() != 0F) {
          setRefAlleleFreq(other.getRefAlleleFreq());
        }
        if (other.getAltAlleleFreq() != 0F) {
          setAltAlleleFreq(other.getAltAlleleFreq());
        }
        if (other.getMaf() != 0F) {
          setMaf(other.getMaf());
        }
        if (other.getMgf() != 0F) {
          setMgf(other.getMgf());
        }
        if (!other.getMafAllele().isEmpty()) {
          mafAllele_ = other.mafAllele_;
          onChanged();
        }
        if (!other.getMgfGenotype().isEmpty()) {
          mgfGenotype_ = other.mgfGenotype_;
          onChanged();
        }
        if (other.getPassedFilters() != false) {
          setPassedFilters(other.getPassedFilters());
        }
        if (other.getMendelianErrors() != 0) {
          setMendelianErrors(other.getMendelianErrors());
        }
        if (other.getCasesPercentDominant() != 0F) {
          setCasesPercentDominant(other.getCasesPercentDominant());
        }
        if (other.getControlsPercentDominant() != 0F) {
          setControlsPercentDominant(other.getControlsPercentDominant());
        }
        if (other.getCasesPercentRecessive() != 0F) {
          setCasesPercentRecessive(other.getCasesPercentRecessive());
        }
        if (other.getControlsPercentRecessive() != 0F) {
          setControlsPercentRecessive(other.getControlsPercentRecessive());
        }
        if (other.getQuality() != 0F) {
          setQuality(other.getQuality());
        }
        if (other.getNumSamples() != 0) {
          setNumSamples(other.getNumSamples());
        }
        if (other.variantType_ != 0) {
          setVariantTypeValue(other.getVariantTypeValue());
        }
        if (other.hasHw()) {
          mergeHw(other.getHw());
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private java.lang.Object refAllele_ = "";
      /**
       * <code>optional string refAllele = 1;</code>
       */
      public java.lang.String getRefAllele() {
        java.lang.Object ref = refAllele_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          refAllele_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string refAllele = 1;</code>
       */
      public com.google.protobuf.ByteString
          getRefAlleleBytes() {
        java.lang.Object ref = refAllele_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          refAllele_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string refAllele = 1;</code>
       */
      public Builder setRefAllele(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        refAllele_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string refAllele = 1;</code>
       */
      public Builder clearRefAllele() {
        
        refAllele_ = getDefaultInstance().getRefAllele();
        onChanged();
        return this;
      }
      /**
       * <code>optional string refAllele = 1;</code>
       */
      public Builder setRefAlleleBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        refAllele_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object altAllele_ = "";
      /**
       * <code>optional string altAllele = 2;</code>
       */
      public java.lang.String getAltAllele() {
        java.lang.Object ref = altAllele_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          altAllele_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string altAllele = 2;</code>
       */
      public com.google.protobuf.ByteString
          getAltAlleleBytes() {
        java.lang.Object ref = altAllele_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          altAllele_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string altAllele = 2;</code>
       */
      public Builder setAltAllele(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        altAllele_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string altAllele = 2;</code>
       */
      public Builder clearAltAllele() {
        
        altAllele_ = getDefaultInstance().getAltAllele();
        onChanged();
        return this;
      }
      /**
       * <code>optional string altAllele = 2;</code>
       */
      public Builder setAltAlleleBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        altAllele_ = value;
        onChanged();
        return this;
      }

      private int refAlleleCount_ ;
      /**
       * <code>optional int32 refAlleleCount = 3;</code>
       */
      public int getRefAlleleCount() {
        return refAlleleCount_;
      }
      /**
       * <code>optional int32 refAlleleCount = 3;</code>
       */
      public Builder setRefAlleleCount(int value) {
        
        refAlleleCount_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 refAlleleCount = 3;</code>
       */
      public Builder clearRefAlleleCount() {
        
        refAlleleCount_ = 0;
        onChanged();
        return this;
      }

      private int altAlleleCount_ ;
      /**
       * <code>optional int32 altAlleleCount = 4;</code>
       */
      public int getAltAlleleCount() {
        return altAlleleCount_;
      }
      /**
       * <code>optional int32 altAlleleCount = 4;</code>
       */
      public Builder setAltAlleleCount(int value) {
        
        altAlleleCount_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 altAlleleCount = 4;</code>
       */
      public Builder clearAltAlleleCount() {
        
        altAlleleCount_ = 0;
        onChanged();
        return this;
      }

      private com.google.protobuf.MapField<
          java.lang.String, java.lang.Integer> genotypesCount_;
      private com.google.protobuf.MapField<java.lang.String, java.lang.Integer>
      internalGetGenotypesCount() {
        if (genotypesCount_ == null) {
          return com.google.protobuf.MapField.emptyMapField(
              GenotypesCountDefaultEntryHolder.defaultEntry);
       }
        return genotypesCount_;
      }
      private com.google.protobuf.MapField<java.lang.String, java.lang.Integer>
      internalGetMutableGenotypesCount() {
        onChanged();;
        if (genotypesCount_ == null) {
          genotypesCount_ = com.google.protobuf.MapField.newMapField(
              GenotypesCountDefaultEntryHolder.defaultEntry);
        }
        if (!genotypesCount_.isMutable()) {
          genotypesCount_ = genotypesCount_.copy();
        }
        return genotypesCount_;
      }
      /**
       * <code>map&lt;string, int32&gt; genotypesCount = 5;</code>
       */
      public java.util.Map<java.lang.String, java.lang.Integer> getGenotypesCount() {
        return internalGetGenotypesCount().getMap();
      }
      /**
       * <code>map&lt;string, int32&gt; genotypesCount = 5;</code>
       */
      public java.util.Map<java.lang.String, java.lang.Integer>
      getMutableGenotypesCount() {
        return internalGetMutableGenotypesCount().getMutableMap();
      }
      /**
       * <code>map&lt;string, int32&gt; genotypesCount = 5;</code>
       */
      public Builder putAllGenotypesCount(
          java.util.Map<java.lang.String, java.lang.Integer> values) {
        getMutableGenotypesCount().putAll(values);
        return this;
      }

      private com.google.protobuf.MapField<
          java.lang.String, java.lang.Float> genotypesFreq_;
      private com.google.protobuf.MapField<java.lang.String, java.lang.Float>
      internalGetGenotypesFreq() {
        if (genotypesFreq_ == null) {
          return com.google.protobuf.MapField.emptyMapField(
              GenotypesFreqDefaultEntryHolder.defaultEntry);
       }
        return genotypesFreq_;
      }
      private com.google.protobuf.MapField<java.lang.String, java.lang.Float>
      internalGetMutableGenotypesFreq() {
        onChanged();;
        if (genotypesFreq_ == null) {
          genotypesFreq_ = com.google.protobuf.MapField.newMapField(
              GenotypesFreqDefaultEntryHolder.defaultEntry);
        }
        if (!genotypesFreq_.isMutable()) {
          genotypesFreq_ = genotypesFreq_.copy();
        }
        return genotypesFreq_;
      }
      /**
       * <code>map&lt;string, float&gt; genotypesFreq = 6;</code>
       */
      public java.util.Map<java.lang.String, java.lang.Float> getGenotypesFreq() {
        return internalGetGenotypesFreq().getMap();
      }
      /**
       * <code>map&lt;string, float&gt; genotypesFreq = 6;</code>
       */
      public java.util.Map<java.lang.String, java.lang.Float>
      getMutableGenotypesFreq() {
        return internalGetMutableGenotypesFreq().getMutableMap();
      }
      /**
       * <code>map&lt;string, float&gt; genotypesFreq = 6;</code>
       */
      public Builder putAllGenotypesFreq(
          java.util.Map<java.lang.String, java.lang.Float> values) {
        getMutableGenotypesFreq().putAll(values);
        return this;
      }

      private int missingAlleles_ ;
      /**
       * <code>optional int32 missingAlleles = 7;</code>
       */
      public int getMissingAlleles() {
        return missingAlleles_;
      }
      /**
       * <code>optional int32 missingAlleles = 7;</code>
       */
      public Builder setMissingAlleles(int value) {
        
        missingAlleles_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 missingAlleles = 7;</code>
       */
      public Builder clearMissingAlleles() {
        
        missingAlleles_ = 0;
        onChanged();
        return this;
      }

      private int missingGenotypes_ ;
      /**
       * <code>optional int32 missingGenotypes = 8;</code>
       */
      public int getMissingGenotypes() {
        return missingGenotypes_;
      }
      /**
       * <code>optional int32 missingGenotypes = 8;</code>
       */
      public Builder setMissingGenotypes(int value) {
        
        missingGenotypes_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 missingGenotypes = 8;</code>
       */
      public Builder clearMissingGenotypes() {
        
        missingGenotypes_ = 0;
        onChanged();
        return this;
      }

      private float refAlleleFreq_ ;
      /**
       * <code>optional float refAlleleFreq = 9;</code>
       */
      public float getRefAlleleFreq() {
        return refAlleleFreq_;
      }
      /**
       * <code>optional float refAlleleFreq = 9;</code>
       */
      public Builder setRefAlleleFreq(float value) {
        
        refAlleleFreq_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float refAlleleFreq = 9;</code>
       */
      public Builder clearRefAlleleFreq() {
        
        refAlleleFreq_ = 0F;
        onChanged();
        return this;
      }

      private float altAlleleFreq_ ;
      /**
       * <code>optional float altAlleleFreq = 10;</code>
       */
      public float getAltAlleleFreq() {
        return altAlleleFreq_;
      }
      /**
       * <code>optional float altAlleleFreq = 10;</code>
       */
      public Builder setAltAlleleFreq(float value) {
        
        altAlleleFreq_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float altAlleleFreq = 10;</code>
       */
      public Builder clearAltAlleleFreq() {
        
        altAlleleFreq_ = 0F;
        onChanged();
        return this;
      }

      private float maf_ ;
      /**
       * <code>optional float maf = 11;</code>
       */
      public float getMaf() {
        return maf_;
      }
      /**
       * <code>optional float maf = 11;</code>
       */
      public Builder setMaf(float value) {
        
        maf_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float maf = 11;</code>
       */
      public Builder clearMaf() {
        
        maf_ = 0F;
        onChanged();
        return this;
      }

      private float mgf_ ;
      /**
       * <code>optional float mgf = 12;</code>
       */
      public float getMgf() {
        return mgf_;
      }
      /**
       * <code>optional float mgf = 12;</code>
       */
      public Builder setMgf(float value) {
        
        mgf_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float mgf = 12;</code>
       */
      public Builder clearMgf() {
        
        mgf_ = 0F;
        onChanged();
        return this;
      }

      private java.lang.Object mafAllele_ = "";
      /**
       * <code>optional string mafAllele = 13;</code>
       */
      public java.lang.String getMafAllele() {
        java.lang.Object ref = mafAllele_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          mafAllele_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string mafAllele = 13;</code>
       */
      public com.google.protobuf.ByteString
          getMafAlleleBytes() {
        java.lang.Object ref = mafAllele_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          mafAllele_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string mafAllele = 13;</code>
       */
      public Builder setMafAllele(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        mafAllele_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string mafAllele = 13;</code>
       */
      public Builder clearMafAllele() {
        
        mafAllele_ = getDefaultInstance().getMafAllele();
        onChanged();
        return this;
      }
      /**
       * <code>optional string mafAllele = 13;</code>
       */
      public Builder setMafAlleleBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        mafAllele_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object mgfGenotype_ = "";
      /**
       * <code>optional string mgfGenotype = 14;</code>
       */
      public java.lang.String getMgfGenotype() {
        java.lang.Object ref = mgfGenotype_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          mgfGenotype_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string mgfGenotype = 14;</code>
       */
      public com.google.protobuf.ByteString
          getMgfGenotypeBytes() {
        java.lang.Object ref = mgfGenotype_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          mgfGenotype_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string mgfGenotype = 14;</code>
       */
      public Builder setMgfGenotype(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        mgfGenotype_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string mgfGenotype = 14;</code>
       */
      public Builder clearMgfGenotype() {
        
        mgfGenotype_ = getDefaultInstance().getMgfGenotype();
        onChanged();
        return this;
      }
      /**
       * <code>optional string mgfGenotype = 14;</code>
       */
      public Builder setMgfGenotypeBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        mgfGenotype_ = value;
        onChanged();
        return this;
      }

      private boolean passedFilters_ ;
      /**
       * <code>optional bool passedFilters = 15;</code>
       */
      public boolean getPassedFilters() {
        return passedFilters_;
      }
      /**
       * <code>optional bool passedFilters = 15;</code>
       */
      public Builder setPassedFilters(boolean value) {
        
        passedFilters_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional bool passedFilters = 15;</code>
       */
      public Builder clearPassedFilters() {
        
        passedFilters_ = false;
        onChanged();
        return this;
      }

      private int mendelianErrors_ ;
      /**
       * <code>optional int32 mendelianErrors = 16;</code>
       */
      public int getMendelianErrors() {
        return mendelianErrors_;
      }
      /**
       * <code>optional int32 mendelianErrors = 16;</code>
       */
      public Builder setMendelianErrors(int value) {
        
        mendelianErrors_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 mendelianErrors = 16;</code>
       */
      public Builder clearMendelianErrors() {
        
        mendelianErrors_ = 0;
        onChanged();
        return this;
      }

      private float casesPercentDominant_ ;
      /**
       * <code>optional float casesPercentDominant = 17;</code>
       */
      public float getCasesPercentDominant() {
        return casesPercentDominant_;
      }
      /**
       * <code>optional float casesPercentDominant = 17;</code>
       */
      public Builder setCasesPercentDominant(float value) {
        
        casesPercentDominant_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float casesPercentDominant = 17;</code>
       */
      public Builder clearCasesPercentDominant() {
        
        casesPercentDominant_ = 0F;
        onChanged();
        return this;
      }

      private float controlsPercentDominant_ ;
      /**
       * <code>optional float controlsPercentDominant = 18;</code>
       */
      public float getControlsPercentDominant() {
        return controlsPercentDominant_;
      }
      /**
       * <code>optional float controlsPercentDominant = 18;</code>
       */
      public Builder setControlsPercentDominant(float value) {
        
        controlsPercentDominant_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float controlsPercentDominant = 18;</code>
       */
      public Builder clearControlsPercentDominant() {
        
        controlsPercentDominant_ = 0F;
        onChanged();
        return this;
      }

      private float casesPercentRecessive_ ;
      /**
       * <code>optional float casesPercentRecessive = 19;</code>
       */
      public float getCasesPercentRecessive() {
        return casesPercentRecessive_;
      }
      /**
       * <code>optional float casesPercentRecessive = 19;</code>
       */
      public Builder setCasesPercentRecessive(float value) {
        
        casesPercentRecessive_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float casesPercentRecessive = 19;</code>
       */
      public Builder clearCasesPercentRecessive() {
        
        casesPercentRecessive_ = 0F;
        onChanged();
        return this;
      }

      private float controlsPercentRecessive_ ;
      /**
       * <code>optional float controlsPercentRecessive = 20;</code>
       */
      public float getControlsPercentRecessive() {
        return controlsPercentRecessive_;
      }
      /**
       * <code>optional float controlsPercentRecessive = 20;</code>
       */
      public Builder setControlsPercentRecessive(float value) {
        
        controlsPercentRecessive_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float controlsPercentRecessive = 20;</code>
       */
      public Builder clearControlsPercentRecessive() {
        
        controlsPercentRecessive_ = 0F;
        onChanged();
        return this;
      }

      private float quality_ ;
      /**
       * <code>optional float quality = 21;</code>
       */
      public float getQuality() {
        return quality_;
      }
      /**
       * <code>optional float quality = 21;</code>
       */
      public Builder setQuality(float value) {
        
        quality_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float quality = 21;</code>
       */
      public Builder clearQuality() {
        
        quality_ = 0F;
        onChanged();
        return this;
      }

      private int numSamples_ ;
      /**
       * <code>optional int32 numSamples = 22;</code>
       */
      public int getNumSamples() {
        return numSamples_;
      }
      /**
       * <code>optional int32 numSamples = 22;</code>
       */
      public Builder setNumSamples(int value) {
        
        numSamples_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 numSamples = 22;</code>
       */
      public Builder clearNumSamples() {
        
        numSamples_ = 0;
        onChanged();
        return this;
      }

      private int variantType_ = 0;
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
       */
      public int getVariantTypeValue() {
        return variantType_;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
       */
      public Builder setVariantTypeValue(int value) {
        variantType_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getVariantType() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType result = org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.valueOf(variantType_);
        return result == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.UNRECOGNIZED : result;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
       */
      public Builder setVariantType(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType value) {
        if (value == null) {
          throw new NullPointerException();
        }
        
        variantType_ = value.getNumber();
        onChanged();
        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType variantType = 23;</code>
       */
      public Builder clearVariantType() {
        
        variantType_ = 0;
        onChanged();
        return this;
      }

      private org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats hw_ = null;
      private com.google.protobuf.SingleFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStatsOrBuilder> hwBuilder_;
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public boolean hasHw() {
        return hwBuilder_ != null || hw_ != null;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats getHw() {
        if (hwBuilder_ == null) {
          return hw_ == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.getDefaultInstance() : hw_;
        } else {
          return hwBuilder_.getMessage();
        }
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public Builder setHw(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats value) {
        if (hwBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          hw_ = value;
          onChanged();
        } else {
          hwBuilder_.setMessage(value);
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public Builder setHw(
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder builderForValue) {
        if (hwBuilder_ == null) {
          hw_ = builderForValue.build();
          onChanged();
        } else {
          hwBuilder_.setMessage(builderForValue.build());
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public Builder mergeHw(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats value) {
        if (hwBuilder_ == null) {
          if (hw_ != null) {
            hw_ =
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.newBuilder(hw_).mergeFrom(value).buildPartial();
          } else {
            hw_ = value;
          }
          onChanged();
        } else {
          hwBuilder_.mergeFrom(value);
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public Builder clearHw() {
        if (hwBuilder_ == null) {
          hw_ = null;
          onChanged();
        } else {
          hw_ = null;
          hwBuilder_ = null;
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder getHwBuilder() {
        
        onChanged();
        return getHwFieldBuilder().getBuilder();
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStatsOrBuilder getHwOrBuilder() {
        if (hwBuilder_ != null) {
          return hwBuilder_.getMessageOrBuilder();
        } else {
          return hw_ == null ?
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.getDefaultInstance() : hw_;
        }
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantHardyWeinbergStats hw = 24;</code>
       */
      private com.google.protobuf.SingleFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStatsOrBuilder> 
          getHwFieldBuilder() {
        if (hwBuilder_ == null) {
          hwBuilder_ = new com.google.protobuf.SingleFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStats.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantHardyWeinbergStatsOrBuilder>(
                  getHw(),
                  getParentForChildren(),
                  isClean());
          hw_ = null;
        }
        return hwBuilder_;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VariantStats)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VariantStats)
    private static final org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats();
    }

    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<VariantStats>
        PARSER = new com.google.protobuf.AbstractParser<VariantStats>() {
      public VariantStats parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new VariantStats(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<VariantStats> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<VariantStats> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface FileEntryOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.FileEntry)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional string fileId = 1;</code>
     */
    java.lang.String getFileId();
    /**
     * <code>optional string fileId = 1;</code>
     */
    com.google.protobuf.ByteString
        getFileIdBytes();

    /**
     * <code>optional string call = 2;</code>
     */
    java.lang.String getCall();
    /**
     * <code>optional string call = 2;</code>
     */
    com.google.protobuf.ByteString
        getCallBytes();

    /**
     * <code>map&lt;string, string&gt; attributes = 3;</code>
     */
    java.util.Map<java.lang.String, java.lang.String>
    getAttributes();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.FileEntry}
   */
  public  static final class FileEntry extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.FileEntry)
      FileEntryOrBuilder {
    // Use FileEntry.newBuilder() to construct.
    private FileEntry(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private FileEntry() {
      fileId_ = "";
      call_ = "";
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private FileEntry(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();

              fileId_ = s;
              break;
            }
            case 18: {
              String s = input.readStringRequireUtf8();

              call_ = s;
              break;
            }
            case 26: {
              if (!((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
                attributes_ = com.google.protobuf.MapField.newMapField(
                    AttributesDefaultEntryHolder.defaultEntry);
                mutable_bitField0_ |= 0x00000004;
              }
              com.google.protobuf.MapEntry<java.lang.String, java.lang.String>
              attributes = input.readMessage(
                  AttributesDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
              attributes_.getMutableMap().put(attributes.getKey(), attributes.getValue());
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 3:
          return internalGetAttributes();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder.class);
    }

    private int bitField0_;
    public static final int FILEID_FIELD_NUMBER = 1;
    private volatile java.lang.Object fileId_;
    /**
     * <code>optional string fileId = 1;</code>
     */
    public java.lang.String getFileId() {
      java.lang.Object ref = fileId_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        fileId_ = s;
        return s;
      }
    }
    /**
     * <code>optional string fileId = 1;</code>
     */
    public com.google.protobuf.ByteString
        getFileIdBytes() {
      java.lang.Object ref = fileId_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        fileId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int CALL_FIELD_NUMBER = 2;
    private volatile java.lang.Object call_;
    /**
     * <code>optional string call = 2;</code>
     */
    public java.lang.String getCall() {
      java.lang.Object ref = call_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        call_ = s;
        return s;
      }
    }
    /**
     * <code>optional string call = 2;</code>
     */
    public com.google.protobuf.ByteString
        getCallBytes() {
      java.lang.Object ref = call_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        call_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int ATTRIBUTES_FIELD_NUMBER = 3;
    private static final class AttributesDefaultEntryHolder {
      static final com.google.protobuf.MapEntry<
          java.lang.String, java.lang.String> defaultEntry =
              com.google.protobuf.MapEntry
              .<java.lang.String, java.lang.String>newDefaultInstance(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_AttributesEntry_descriptor, 
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "",
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "");
    }
    private com.google.protobuf.MapField<
        java.lang.String, java.lang.String> attributes_;
    private com.google.protobuf.MapField<java.lang.String, java.lang.String>
    internalGetAttributes() {
      if (attributes_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            AttributesDefaultEntryHolder.defaultEntry);
     }
      return attributes_;
    }
    /**
     * <code>map&lt;string, string&gt; attributes = 3;</code>
     */

    public java.util.Map<java.lang.String, java.lang.String> getAttributes() {
      return internalGetAttributes().getMap();
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!getFileIdBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, fileId_);
      }
      if (!getCallBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 2, call_);
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.String> entry
           : internalGetAttributes().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.String>
        attributes = AttributesDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        output.writeMessage(3, attributes);
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!getFileIdBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(1, fileId_);
      }
      if (!getCallBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(2, call_);
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.String> entry
           : internalGetAttributes().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.String>
        attributes = AttributesDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        size += com.google.protobuf.CodedOutputStream
            .computeMessageSize(3, attributes);
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.FileEntry}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.FileEntry)
        org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_descriptor;
      }

      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMapField(
          int number) {
        switch (number) {
          case 3:
            return internalGetAttributes();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMutableMapField(
          int number) {
        switch (number) {
          case 3:
            return internalGetMutableAttributes();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        fileId_ = "";

        call_ = "";

        internalGetMutableAttributes().clear();
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry build() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry result = new org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.fileId_ = fileId_;
        result.call_ = call_;
        result.attributes_ = internalGetAttributes();
        result.attributes_.makeImmutable();
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.getDefaultInstance()) return this;
        if (!other.getFileId().isEmpty()) {
          fileId_ = other.fileId_;
          onChanged();
        }
        if (!other.getCall().isEmpty()) {
          call_ = other.call_;
          onChanged();
        }
        internalGetMutableAttributes().mergeFrom(
            other.internalGetAttributes());
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private java.lang.Object fileId_ = "";
      /**
       * <code>optional string fileId = 1;</code>
       */
      public java.lang.String getFileId() {
        java.lang.Object ref = fileId_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          fileId_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public com.google.protobuf.ByteString
          getFileIdBytes() {
        java.lang.Object ref = fileId_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          fileId_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public Builder setFileId(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        fileId_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public Builder clearFileId() {
        
        fileId_ = getDefaultInstance().getFileId();
        onChanged();
        return this;
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public Builder setFileIdBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        fileId_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object call_ = "";
      /**
       * <code>optional string call = 2;</code>
       */
      public java.lang.String getCall() {
        java.lang.Object ref = call_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          call_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string call = 2;</code>
       */
      public com.google.protobuf.ByteString
          getCallBytes() {
        java.lang.Object ref = call_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          call_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string call = 2;</code>
       */
      public Builder setCall(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        call_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string call = 2;</code>
       */
      public Builder clearCall() {
        
        call_ = getDefaultInstance().getCall();
        onChanged();
        return this;
      }
      /**
       * <code>optional string call = 2;</code>
       */
      public Builder setCallBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        call_ = value;
        onChanged();
        return this;
      }

      private com.google.protobuf.MapField<
          java.lang.String, java.lang.String> attributes_;
      private com.google.protobuf.MapField<java.lang.String, java.lang.String>
      internalGetAttributes() {
        if (attributes_ == null) {
          return com.google.protobuf.MapField.emptyMapField(
              AttributesDefaultEntryHolder.defaultEntry);
       }
        return attributes_;
      }
      private com.google.protobuf.MapField<java.lang.String, java.lang.String>
      internalGetMutableAttributes() {
        onChanged();;
        if (attributes_ == null) {
          attributes_ = com.google.protobuf.MapField.newMapField(
              AttributesDefaultEntryHolder.defaultEntry);
        }
        if (!attributes_.isMutable()) {
          attributes_ = attributes_.copy();
        }
        return attributes_;
      }
      /**
       * <code>map&lt;string, string&gt; attributes = 3;</code>
       */
      public java.util.Map<java.lang.String, java.lang.String> getAttributes() {
        return internalGetAttributes().getMap();
      }
      /**
       * <code>map&lt;string, string&gt; attributes = 3;</code>
       */
      public java.util.Map<java.lang.String, java.lang.String>
      getMutableAttributes() {
        return internalGetMutableAttributes().getMutableMap();
      }
      /**
       * <code>map&lt;string, string&gt; attributes = 3;</code>
       */
      public Builder putAllAttributes(
          java.util.Map<java.lang.String, java.lang.String> values) {
        getMutableAttributes().putAll(values);
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.FileEntry)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.FileEntry)
    private static final org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry();
    }

    public static org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<FileEntry>
        PARSER = new com.google.protobuf.AbstractParser<FileEntry>() {
      public FileEntry parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new FileEntry(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<FileEntry> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<FileEntry> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface VariantSourceEntryOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional string studyId = 1;</code>
     */
    java.lang.String getStudyId();
    /**
     * <code>optional string studyId = 1;</code>
     */
    com.google.protobuf.ByteString
        getStudyIdBytes();

    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry> 
        getFilesList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry getFiles(int index);
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    int getFilesCount();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder> 
        getFilesOrBuilderList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder getFilesOrBuilder(
        int index);

    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    com.google.protobuf.ProtocolStringList
        getSecondaryAlternatesList();
    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    int getSecondaryAlternatesCount();
    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    java.lang.String getSecondaryAlternates(int index);
    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    com.google.protobuf.ByteString
        getSecondaryAlternatesBytes(int index);

    /**
     * <code>repeated string format = 4;</code>
     */
    com.google.protobuf.ProtocolStringList
        getFormatList();
    /**
     * <code>repeated string format = 4;</code>
     */
    int getFormatCount();
    /**
     * <code>repeated string format = 4;</code>
     */
    java.lang.String getFormat(int index);
    /**
     * <code>repeated string format = 4;</code>
     */
    com.google.protobuf.ByteString
        getFormatBytes(int index);

    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry> 
        getSamplesDataList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry getSamplesData(int index);
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    int getSamplesDataCount();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder> 
        getSamplesDataOrBuilderList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder getSamplesDataOrBuilder(
        int index);

    /**
     * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.VariantStats&gt; stats = 6;</code>
     */
    java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
    getStats();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantSourceEntry}
   */
  public  static final class VariantSourceEntry extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry)
      VariantSourceEntryOrBuilder {
    // Use VariantSourceEntry.newBuilder() to construct.
    private VariantSourceEntry(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private VariantSourceEntry() {
      studyId_ = "";
      files_ = java.util.Collections.emptyList();
      secondaryAlternates_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      format_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      samplesData_ = java.util.Collections.emptyList();
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private VariantSourceEntry(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();

              studyId_ = s;
              break;
            }
            case 18: {
              if (!((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
                files_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry>();
                mutable_bitField0_ |= 0x00000002;
              }
              files_.add(input.readMessage(org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.parser(), extensionRegistry));
              break;
            }
            case 26: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
                secondaryAlternates_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000004;
              }
              secondaryAlternates_.add(s);
              break;
            }
            case 34: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000008) == 0x00000008)) {
                format_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000008;
              }
              format_.add(s);
              break;
            }
            case 42: {
              if (!((mutable_bitField0_ & 0x00000010) == 0x00000010)) {
                samplesData_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry>();
                mutable_bitField0_ |= 0x00000010;
              }
              samplesData_.add(input.readMessage(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.parser(), extensionRegistry));
              break;
            }
            case 50: {
              if (!((mutable_bitField0_ & 0x00000020) == 0x00000020)) {
                stats_ = com.google.protobuf.MapField.newMapField(
                    StatsDefaultEntryHolder.defaultEntry);
                mutable_bitField0_ |= 0x00000020;
              }
              com.google.protobuf.MapEntry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
              stats = input.readMessage(
                  StatsDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
              stats_.getMutableMap().put(stats.getKey(), stats.getValue());
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
          files_ = java.util.Collections.unmodifiableList(files_);
        }
        if (((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
          secondaryAlternates_ = secondaryAlternates_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00000008) == 0x00000008)) {
          format_ = format_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00000010) == 0x00000010)) {
          samplesData_ = java.util.Collections.unmodifiableList(samplesData_);
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 6:
          return internalGetStats();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder.class);
    }

    public interface SamplesDataInfoEntryOrBuilder extends
        // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry)
        com.google.protobuf.MessageOrBuilder {

      /**
       * <code>repeated string info = 1;</code>
       */
      com.google.protobuf.ProtocolStringList
          getInfoList();
      /**
       * <code>repeated string info = 1;</code>
       */
      int getInfoCount();
      /**
       * <code>repeated string info = 1;</code>
       */
      java.lang.String getInfo(int index);
      /**
       * <code>repeated string info = 1;</code>
       */
      com.google.protobuf.ByteString
          getInfoBytes(int index);
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry}
     */
    public  static final class SamplesDataInfoEntry extends
        com.google.protobuf.GeneratedMessage implements
        // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry)
        SamplesDataInfoEntryOrBuilder {
      // Use SamplesDataInfoEntry.newBuilder() to construct.
      private SamplesDataInfoEntry(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
        super(builder);
      }
      private SamplesDataInfoEntry() {
        info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      }

      @java.lang.Override
      public final com.google.protobuf.UnknownFieldSet
      getUnknownFields() {
        return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
      }
      private SamplesDataInfoEntry(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
        this();
        int mutable_bitField0_ = 0;
        try {
          boolean done = false;
          while (!done) {
            int tag = input.readTag();
            switch (tag) {
              case 0:
                done = true;
                break;
              default: {
                if (!input.skipField(tag)) {
                  done = true;
                }
                break;
              }
              case 10: {
                String s = input.readStringRequireUtf8();
                if (!((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
                  info_ = new com.google.protobuf.LazyStringArrayList();
                  mutable_bitField0_ |= 0x00000001;
                }
                info_.add(s);
                break;
              }
            }
          }
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          throw new RuntimeException(e.setUnfinishedMessage(this));
        } catch (java.io.IOException e) {
          throw new RuntimeException(
              new com.google.protobuf.InvalidProtocolBufferException(
                  e.getMessage()).setUnfinishedMessage(this));
        } finally {
          if (((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
            info_ = info_.getUnmodifiableView();
          }
          makeExtensionsImmutable();
        }
      }
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder.class);
      }

      public static final int INFO_FIELD_NUMBER = 1;
      private com.google.protobuf.LazyStringList info_;
      /**
       * <code>repeated string info = 1;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getInfoList() {
        return info_;
      }
      /**
       * <code>repeated string info = 1;</code>
       */
      public int getInfoCount() {
        return info_.size();
      }
      /**
       * <code>repeated string info = 1;</code>
       */
      public java.lang.String getInfo(int index) {
        return info_.get(index);
      }
      /**
       * <code>repeated string info = 1;</code>
       */
      public com.google.protobuf.ByteString
          getInfoBytes(int index) {
        return info_.getByteString(index);
      }

      private byte memoizedIsInitialized = -1;
      public final boolean isInitialized() {
        byte isInitialized = memoizedIsInitialized;
        if (isInitialized == 1) return true;
        if (isInitialized == 0) return false;

        memoizedIsInitialized = 1;
        return true;
      }

      public void writeTo(com.google.protobuf.CodedOutputStream output)
                          throws java.io.IOException {
        for (int i = 0; i < info_.size(); i++) {
          com.google.protobuf.GeneratedMessage.writeString(output, 1, info_.getRaw(i));
        }
      }

      public int getSerializedSize() {
        int size = memoizedSize;
        if (size != -1) return size;

        size = 0;
        {
          int dataSize = 0;
          for (int i = 0; i < info_.size(); i++) {
            dataSize += computeStringSizeNoTag(info_.getRaw(i));
          }
          size += dataSize;
          size += 1 * getInfoList().size();
        }
        memoizedSize = size;
        return size;
      }

      private static final long serialVersionUID = 0L;
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(
          com.google.protobuf.ByteString data)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(
          com.google.protobuf.ByteString data,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(byte[] data)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(
          byte[] data,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(java.io.InputStream input)
          throws java.io.IOException {
        return PARSER.parseFrom(input);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(
          java.io.InputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        return PARSER.parseFrom(input, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseDelimitedFrom(java.io.InputStream input)
          throws java.io.IOException {
        return PARSER.parseDelimitedFrom(input);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseDelimitedFrom(
          java.io.InputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        return PARSER.parseDelimitedFrom(input, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(
          com.google.protobuf.CodedInputStream input)
          throws java.io.IOException {
        return PARSER.parseFrom(input);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parseFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        return PARSER.parseFrom(input, extensionRegistry);
      }

      public Builder newBuilderForType() { return newBuilder(); }
      public static Builder newBuilder() {
        return DEFAULT_INSTANCE.toBuilder();
      }
      public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry prototype) {
        return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
      }
      public Builder toBuilder() {
        return this == DEFAULT_INSTANCE
            ? new Builder() : new Builder().mergeFrom(this);
      }

      @java.lang.Override
      protected Builder newBuilderForType(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        Builder builder = new Builder(parent);
        return builder;
      }
      /**
       * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry}
       */
      public static final class Builder extends
          com.google.protobuf.GeneratedMessage.Builder<Builder> implements
          // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry)
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder {
        public static final com.google.protobuf.Descriptors.Descriptor
            getDescriptor() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_descriptor;
        }

        protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
            internalGetFieldAccessorTable() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_fieldAccessorTable
              .ensureFieldAccessorsInitialized(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder.class);
        }

        // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.newBuilder()
        private Builder() {
          maybeForceBuilderInitialization();
        }

        private Builder(
            com.google.protobuf.GeneratedMessage.BuilderParent parent) {
          super(parent);
          maybeForceBuilderInitialization();
        }
        private void maybeForceBuilderInitialization() {
          if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
          }
        }
        public Builder clear() {
          super.clear();
          info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
          bitField0_ = (bitField0_ & ~0x00000001);
          return this;
        }

        public com.google.protobuf.Descriptors.Descriptor
            getDescriptorForType() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_descriptor;
        }

        public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry getDefaultInstanceForType() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.getDefaultInstance();
        }

        public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry build() {
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry result = buildPartial();
          if (!result.isInitialized()) {
            throw newUninitializedMessageException(result);
          }
          return result;
        }

        public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry buildPartial() {
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry result = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry(this);
          int from_bitField0_ = bitField0_;
          if (((bitField0_ & 0x00000001) == 0x00000001)) {
            info_ = info_.getUnmodifiableView();
            bitField0_ = (bitField0_ & ~0x00000001);
          }
          result.info_ = info_;
          onBuilt();
          return result;
        }

        public Builder mergeFrom(com.google.protobuf.Message other) {
          if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry) {
            return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry)other);
          } else {
            super.mergeFrom(other);
            return this;
          }
        }

        public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry other) {
          if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.getDefaultInstance()) return this;
          if (!other.info_.isEmpty()) {
            if (info_.isEmpty()) {
              info_ = other.info_;
              bitField0_ = (bitField0_ & ~0x00000001);
            } else {
              ensureInfoIsMutable();
              info_.addAll(other.info_);
            }
            onChanged();
          }
          onChanged();
          return this;
        }

        public final boolean isInitialized() {
          return true;
        }

        public Builder mergeFrom(
            com.google.protobuf.CodedInputStream input,
            com.google.protobuf.ExtensionRegistryLite extensionRegistry)
            throws java.io.IOException {
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry parsedMessage = null;
          try {
            parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
          } catch (com.google.protobuf.InvalidProtocolBufferException e) {
            parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry) e.getUnfinishedMessage();
            throw e;
          } finally {
            if (parsedMessage != null) {
              mergeFrom(parsedMessage);
            }
          }
          return this;
        }
        private int bitField0_;

        private com.google.protobuf.LazyStringList info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        private void ensureInfoIsMutable() {
          if (!((bitField0_ & 0x00000001) == 0x00000001)) {
            info_ = new com.google.protobuf.LazyStringArrayList(info_);
            bitField0_ |= 0x00000001;
           }
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public com.google.protobuf.ProtocolStringList
            getInfoList() {
          return info_.getUnmodifiableView();
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public int getInfoCount() {
          return info_.size();
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public java.lang.String getInfo(int index) {
          return info_.get(index);
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public com.google.protobuf.ByteString
            getInfoBytes(int index) {
          return info_.getByteString(index);
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder setInfo(
            int index, java.lang.String value) {
          if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoIsMutable();
          info_.set(index, value);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder addInfo(
            java.lang.String value) {
          if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoIsMutable();
          info_.add(value);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder addAllInfo(
            java.lang.Iterable<java.lang.String> values) {
          ensureInfoIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, info_);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder clearInfo() {
          info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
          bitField0_ = (bitField0_ & ~0x00000001);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder addInfoBytes(
            com.google.protobuf.ByteString value) {
          if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
          ensureInfoIsMutable();
          info_.add(value);
          onChanged();
          return this;
        }
        public final Builder setUnknownFields(
            final com.google.protobuf.UnknownFieldSet unknownFields) {
          return this;
        }

        public final Builder mergeUnknownFields(
            final com.google.protobuf.UnknownFieldSet unknownFields) {
          return this;
        }


        // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry)
      }

      // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry)
      private static final org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry DEFAULT_INSTANCE;
      static {
        DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry();
      }

      public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry getDefaultInstance() {
        return DEFAULT_INSTANCE;
      }

      private static final com.google.protobuf.Parser<SamplesDataInfoEntry>
          PARSER = new com.google.protobuf.AbstractParser<SamplesDataInfoEntry>() {
        public SamplesDataInfoEntry parsePartialFrom(
            com.google.protobuf.CodedInputStream input,
            com.google.protobuf.ExtensionRegistryLite extensionRegistry)
            throws com.google.protobuf.InvalidProtocolBufferException {
          try {
            return new SamplesDataInfoEntry(input, extensionRegistry);
          } catch (RuntimeException e) {
            if (e.getCause() instanceof
                com.google.protobuf.InvalidProtocolBufferException) {
              throw (com.google.protobuf.InvalidProtocolBufferException)
                  e.getCause();
            }
            throw e;
          }
        }
      };

      public static com.google.protobuf.Parser<SamplesDataInfoEntry> parser() {
        return PARSER;
      }

      @java.lang.Override
      public com.google.protobuf.Parser<SamplesDataInfoEntry> getParserForType() {
        return PARSER;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry getDefaultInstanceForType() {
        return DEFAULT_INSTANCE;
      }

    }

    private int bitField0_;
    public static final int STUDYID_FIELD_NUMBER = 1;
    private volatile java.lang.Object studyId_;
    /**
     * <code>optional string studyId = 1;</code>
     */
    public java.lang.String getStudyId() {
      java.lang.Object ref = studyId_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        studyId_ = s;
        return s;
      }
    }
    /**
     * <code>optional string studyId = 1;</code>
     */
    public com.google.protobuf.ByteString
        getStudyIdBytes() {
      java.lang.Object ref = studyId_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        studyId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int FILES_FIELD_NUMBER = 2;
    private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry> files_;
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry> getFilesList() {
      return files_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder> 
        getFilesOrBuilderList() {
      return files_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    public int getFilesCount() {
      return files_.size();
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry getFiles(int index) {
      return files_.get(index);
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder getFilesOrBuilder(
        int index) {
      return files_.get(index);
    }

    public static final int SECONDARYALTERNATES_FIELD_NUMBER = 3;
    private com.google.protobuf.LazyStringList secondaryAlternates_;
    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getSecondaryAlternatesList() {
      return secondaryAlternates_;
    }
    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    public int getSecondaryAlternatesCount() {
      return secondaryAlternates_.size();
    }
    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    public java.lang.String getSecondaryAlternates(int index) {
      return secondaryAlternates_.get(index);
    }
    /**
     * <code>repeated string secondaryAlternates = 3;</code>
     */
    public com.google.protobuf.ByteString
        getSecondaryAlternatesBytes(int index) {
      return secondaryAlternates_.getByteString(index);
    }

    public static final int FORMAT_FIELD_NUMBER = 4;
    private com.google.protobuf.LazyStringList format_;
    /**
     * <code>repeated string format = 4;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getFormatList() {
      return format_;
    }
    /**
     * <code>repeated string format = 4;</code>
     */
    public int getFormatCount() {
      return format_.size();
    }
    /**
     * <code>repeated string format = 4;</code>
     */
    public java.lang.String getFormat(int index) {
      return format_.get(index);
    }
    /**
     * <code>repeated string format = 4;</code>
     */
    public com.google.protobuf.ByteString
        getFormatBytes(int index) {
      return format_.getByteString(index);
    }

    public static final int SAMPLESDATA_FIELD_NUMBER = 5;
    private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry> samplesData_;
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry> getSamplesDataList() {
      return samplesData_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder> 
        getSamplesDataOrBuilderList() {
      return samplesData_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    public int getSamplesDataCount() {
      return samplesData_.size();
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry getSamplesData(int index) {
      return samplesData_.get(index);
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder getSamplesDataOrBuilder(
        int index) {
      return samplesData_.get(index);
    }

    public static final int STATS_FIELD_NUMBER = 6;
    private static final class StatsDefaultEntryHolder {
      static final com.google.protobuf.MapEntry<
          java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> defaultEntry =
              com.google.protobuf.MapEntry
              .<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>newDefaultInstance(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_StatsEntry_descriptor, 
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "",
                  com.google.protobuf.WireFormat.FieldType.MESSAGE,
                  org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats.getDefaultInstance());
    }
    private com.google.protobuf.MapField<
        java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> stats_;
    private com.google.protobuf.MapField<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
    internalGetStats() {
      if (stats_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            StatsDefaultEntryHolder.defaultEntry);
     }
      return stats_;
    }
    /**
     * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.VariantStats&gt; stats = 6;</code>
     */

    public java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> getStats() {
      return internalGetStats().getMap();
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!getStudyIdBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, studyId_);
      }
      for (int i = 0; i < files_.size(); i++) {
        output.writeMessage(2, files_.get(i));
      }
      for (int i = 0; i < secondaryAlternates_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 3, secondaryAlternates_.getRaw(i));
      }
      for (int i = 0; i < format_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 4, format_.getRaw(i));
      }
      for (int i = 0; i < samplesData_.size(); i++) {
        output.writeMessage(5, samplesData_.get(i));
      }
      for (java.util.Map.Entry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> entry
           : internalGetStats().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
        stats = StatsDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        output.writeMessage(6, stats);
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!getStudyIdBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(1, studyId_);
      }
      for (int i = 0; i < files_.size(); i++) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(2, files_.get(i));
      }
      {
        int dataSize = 0;
        for (int i = 0; i < secondaryAlternates_.size(); i++) {
          dataSize += computeStringSizeNoTag(secondaryAlternates_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getSecondaryAlternatesList().size();
      }
      {
        int dataSize = 0;
        for (int i = 0; i < format_.size(); i++) {
          dataSize += computeStringSizeNoTag(format_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getFormatList().size();
      }
      for (int i = 0; i < samplesData_.size(); i++) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(5, samplesData_.get(i));
      }
      for (java.util.Map.Entry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> entry
           : internalGetStats().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
        stats = StatsDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        size += com.google.protobuf.CodedOutputStream
            .computeMessageSize(6, stats);
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantSourceEntry}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry)
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor;
      }

      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMapField(
          int number) {
        switch (number) {
          case 6:
            return internalGetStats();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMutableMapField(
          int number) {
        switch (number) {
          case 6:
            return internalGetMutableStats();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
          getFilesFieldBuilder();
          getSamplesDataFieldBuilder();
        }
      }
      public Builder clear() {
        super.clear();
        studyId_ = "";

        if (filesBuilder_ == null) {
          files_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000002);
        } else {
          filesBuilder_.clear();
        }
        secondaryAlternates_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000004);
        format_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000008);
        if (samplesDataBuilder_ == null) {
          samplesData_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000010);
        } else {
          samplesDataBuilder_.clear();
        }
        internalGetMutableStats().clear();
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry build() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry result = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.studyId_ = studyId_;
        if (filesBuilder_ == null) {
          if (((bitField0_ & 0x00000002) == 0x00000002)) {
            files_ = java.util.Collections.unmodifiableList(files_);
            bitField0_ = (bitField0_ & ~0x00000002);
          }
          result.files_ = files_;
        } else {
          result.files_ = filesBuilder_.build();
        }
        if (((bitField0_ & 0x00000004) == 0x00000004)) {
          secondaryAlternates_ = secondaryAlternates_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000004);
        }
        result.secondaryAlternates_ = secondaryAlternates_;
        if (((bitField0_ & 0x00000008) == 0x00000008)) {
          format_ = format_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000008);
        }
        result.format_ = format_;
        if (samplesDataBuilder_ == null) {
          if (((bitField0_ & 0x00000010) == 0x00000010)) {
            samplesData_ = java.util.Collections.unmodifiableList(samplesData_);
            bitField0_ = (bitField0_ & ~0x00000010);
          }
          result.samplesData_ = samplesData_;
        } else {
          result.samplesData_ = samplesDataBuilder_.build();
        }
        result.stats_ = internalGetStats();
        result.stats_.makeImmutable();
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.getDefaultInstance()) return this;
        if (!other.getStudyId().isEmpty()) {
          studyId_ = other.studyId_;
          onChanged();
        }
        if (filesBuilder_ == null) {
          if (!other.files_.isEmpty()) {
            if (files_.isEmpty()) {
              files_ = other.files_;
              bitField0_ = (bitField0_ & ~0x00000002);
            } else {
              ensureFilesIsMutable();
              files_.addAll(other.files_);
            }
            onChanged();
          }
        } else {
          if (!other.files_.isEmpty()) {
            if (filesBuilder_.isEmpty()) {
              filesBuilder_.dispose();
              filesBuilder_ = null;
              files_ = other.files_;
              bitField0_ = (bitField0_ & ~0x00000002);
              filesBuilder_ = 
                com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders ?
                   getFilesFieldBuilder() : null;
            } else {
              filesBuilder_.addAllMessages(other.files_);
            }
          }
        }
        if (!other.secondaryAlternates_.isEmpty()) {
          if (secondaryAlternates_.isEmpty()) {
            secondaryAlternates_ = other.secondaryAlternates_;
            bitField0_ = (bitField0_ & ~0x00000004);
          } else {
            ensureSecondaryAlternatesIsMutable();
            secondaryAlternates_.addAll(other.secondaryAlternates_);
          }
          onChanged();
        }
        if (!other.format_.isEmpty()) {
          if (format_.isEmpty()) {
            format_ = other.format_;
            bitField0_ = (bitField0_ & ~0x00000008);
          } else {
            ensureFormatIsMutable();
            format_.addAll(other.format_);
          }
          onChanged();
        }
        if (samplesDataBuilder_ == null) {
          if (!other.samplesData_.isEmpty()) {
            if (samplesData_.isEmpty()) {
              samplesData_ = other.samplesData_;
              bitField0_ = (bitField0_ & ~0x00000010);
            } else {
              ensureSamplesDataIsMutable();
              samplesData_.addAll(other.samplesData_);
            }
            onChanged();
          }
        } else {
          if (!other.samplesData_.isEmpty()) {
            if (samplesDataBuilder_.isEmpty()) {
              samplesDataBuilder_.dispose();
              samplesDataBuilder_ = null;
              samplesData_ = other.samplesData_;
              bitField0_ = (bitField0_ & ~0x00000010);
              samplesDataBuilder_ = 
                com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders ?
                   getSamplesDataFieldBuilder() : null;
            } else {
              samplesDataBuilder_.addAllMessages(other.samplesData_);
            }
          }
        }
        internalGetMutableStats().mergeFrom(
            other.internalGetStats());
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private java.lang.Object studyId_ = "";
      /**
       * <code>optional string studyId = 1;</code>
       */
      public java.lang.String getStudyId() {
        java.lang.Object ref = studyId_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          studyId_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string studyId = 1;</code>
       */
      public com.google.protobuf.ByteString
          getStudyIdBytes() {
        java.lang.Object ref = studyId_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          studyId_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string studyId = 1;</code>
       */
      public Builder setStudyId(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        studyId_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string studyId = 1;</code>
       */
      public Builder clearStudyId() {
        
        studyId_ = getDefaultInstance().getStudyId();
        onChanged();
        return this;
      }
      /**
       * <code>optional string studyId = 1;</code>
       */
      public Builder setStudyIdBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        studyId_ = value;
        onChanged();
        return this;
      }

      private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry> files_ =
        java.util.Collections.emptyList();
      private void ensureFilesIsMutable() {
        if (!((bitField0_ & 0x00000002) == 0x00000002)) {
          files_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry>(files_);
          bitField0_ |= 0x00000002;
         }
      }

      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder> filesBuilder_;

      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry> getFilesList() {
        if (filesBuilder_ == null) {
          return java.util.Collections.unmodifiableList(files_);
        } else {
          return filesBuilder_.getMessageList();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public int getFilesCount() {
        if (filesBuilder_ == null) {
          return files_.size();
        } else {
          return filesBuilder_.getCount();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry getFiles(int index) {
        if (filesBuilder_ == null) {
          return files_.get(index);
        } else {
          return filesBuilder_.getMessage(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder setFiles(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry value) {
        if (filesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureFilesIsMutable();
          files_.set(index, value);
          onChanged();
        } else {
          filesBuilder_.setMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder setFiles(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder builderForValue) {
        if (filesBuilder_ == null) {
          ensureFilesIsMutable();
          files_.set(index, builderForValue.build());
          onChanged();
        } else {
          filesBuilder_.setMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder addFiles(org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry value) {
        if (filesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureFilesIsMutable();
          files_.add(value);
          onChanged();
        } else {
          filesBuilder_.addMessage(value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder addFiles(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry value) {
        if (filesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureFilesIsMutable();
          files_.add(index, value);
          onChanged();
        } else {
          filesBuilder_.addMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder addFiles(
          org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder builderForValue) {
        if (filesBuilder_ == null) {
          ensureFilesIsMutable();
          files_.add(builderForValue.build());
          onChanged();
        } else {
          filesBuilder_.addMessage(builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder addFiles(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder builderForValue) {
        if (filesBuilder_ == null) {
          ensureFilesIsMutable();
          files_.add(index, builderForValue.build());
          onChanged();
        } else {
          filesBuilder_.addMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder addAllFiles(
          java.lang.Iterable<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry> values) {
        if (filesBuilder_ == null) {
          ensureFilesIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, files_);
          onChanged();
        } else {
          filesBuilder_.addAllMessages(values);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder clearFiles() {
        if (filesBuilder_ == null) {
          files_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000002);
          onChanged();
        } else {
          filesBuilder_.clear();
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public Builder removeFiles(int index) {
        if (filesBuilder_ == null) {
          ensureFilesIsMutable();
          files_.remove(index);
          onChanged();
        } else {
          filesBuilder_.remove(index);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder getFilesBuilder(
          int index) {
        return getFilesFieldBuilder().getBuilder(index);
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder getFilesOrBuilder(
          int index) {
        if (filesBuilder_ == null) {
          return files_.get(index);  } else {
          return filesBuilder_.getMessageOrBuilder(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder> 
           getFilesOrBuilderList() {
        if (filesBuilder_ != null) {
          return filesBuilder_.getMessageOrBuilderList();
        } else {
          return java.util.Collections.unmodifiableList(files_);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder addFilesBuilder() {
        return getFilesFieldBuilder().addBuilder(
            org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder addFilesBuilder(
          int index) {
        return getFilesFieldBuilder().addBuilder(
            index, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.FileEntry files = 2;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder> 
           getFilesBuilderList() {
        return getFilesFieldBuilder().getBuilderList();
      }
      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder> 
          getFilesFieldBuilder() {
        if (filesBuilder_ == null) {
          filesBuilder_ = new com.google.protobuf.RepeatedFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.FileEntryOrBuilder>(
                  files_,
                  ((bitField0_ & 0x00000002) == 0x00000002),
                  getParentForChildren(),
                  isClean());
          files_ = null;
        }
        return filesBuilder_;
      }

      private com.google.protobuf.LazyStringList secondaryAlternates_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureSecondaryAlternatesIsMutable() {
        if (!((bitField0_ & 0x00000004) == 0x00000004)) {
          secondaryAlternates_ = new com.google.protobuf.LazyStringArrayList(secondaryAlternates_);
          bitField0_ |= 0x00000004;
         }
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getSecondaryAlternatesList() {
        return secondaryAlternates_.getUnmodifiableView();
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public int getSecondaryAlternatesCount() {
        return secondaryAlternates_.size();
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public java.lang.String getSecondaryAlternates(int index) {
        return secondaryAlternates_.get(index);
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public com.google.protobuf.ByteString
          getSecondaryAlternatesBytes(int index) {
        return secondaryAlternates_.getByteString(index);
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public Builder setSecondaryAlternates(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureSecondaryAlternatesIsMutable();
        secondaryAlternates_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public Builder addSecondaryAlternates(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureSecondaryAlternatesIsMutable();
        secondaryAlternates_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public Builder addAllSecondaryAlternates(
          java.lang.Iterable<java.lang.String> values) {
        ensureSecondaryAlternatesIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, secondaryAlternates_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public Builder clearSecondaryAlternates() {
        secondaryAlternates_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000004);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string secondaryAlternates = 3;</code>
       */
      public Builder addSecondaryAlternatesBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureSecondaryAlternatesIsMutable();
        secondaryAlternates_.add(value);
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList format_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureFormatIsMutable() {
        if (!((bitField0_ & 0x00000008) == 0x00000008)) {
          format_ = new com.google.protobuf.LazyStringArrayList(format_);
          bitField0_ |= 0x00000008;
         }
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getFormatList() {
        return format_.getUnmodifiableView();
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public int getFormatCount() {
        return format_.size();
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public java.lang.String getFormat(int index) {
        return format_.get(index);
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public com.google.protobuf.ByteString
          getFormatBytes(int index) {
        return format_.getByteString(index);
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public Builder setFormat(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureFormatIsMutable();
        format_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public Builder addFormat(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureFormatIsMutable();
        format_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public Builder addAllFormat(
          java.lang.Iterable<java.lang.String> values) {
        ensureFormatIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, format_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public Builder clearFormat() {
        format_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000008);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string format = 4;</code>
       */
      public Builder addFormatBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureFormatIsMutable();
        format_.add(value);
        onChanged();
        return this;
      }

      private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry> samplesData_ =
        java.util.Collections.emptyList();
      private void ensureSamplesDataIsMutable() {
        if (!((bitField0_ & 0x00000010) == 0x00000010)) {
          samplesData_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry>(samplesData_);
          bitField0_ |= 0x00000010;
         }
      }

      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder> samplesDataBuilder_;

      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry> getSamplesDataList() {
        if (samplesDataBuilder_ == null) {
          return java.util.Collections.unmodifiableList(samplesData_);
        } else {
          return samplesDataBuilder_.getMessageList();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public int getSamplesDataCount() {
        if (samplesDataBuilder_ == null) {
          return samplesData_.size();
        } else {
          return samplesDataBuilder_.getCount();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry getSamplesData(int index) {
        if (samplesDataBuilder_ == null) {
          return samplesData_.get(index);
        } else {
          return samplesDataBuilder_.getMessage(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder setSamplesData(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry value) {
        if (samplesDataBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSamplesDataIsMutable();
          samplesData_.set(index, value);
          onChanged();
        } else {
          samplesDataBuilder_.setMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder setSamplesData(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder builderForValue) {
        if (samplesDataBuilder_ == null) {
          ensureSamplesDataIsMutable();
          samplesData_.set(index, builderForValue.build());
          onChanged();
        } else {
          samplesDataBuilder_.setMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder addSamplesData(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry value) {
        if (samplesDataBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSamplesDataIsMutable();
          samplesData_.add(value);
          onChanged();
        } else {
          samplesDataBuilder_.addMessage(value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder addSamplesData(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry value) {
        if (samplesDataBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSamplesDataIsMutable();
          samplesData_.add(index, value);
          onChanged();
        } else {
          samplesDataBuilder_.addMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder addSamplesData(
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder builderForValue) {
        if (samplesDataBuilder_ == null) {
          ensureSamplesDataIsMutable();
          samplesData_.add(builderForValue.build());
          onChanged();
        } else {
          samplesDataBuilder_.addMessage(builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder addSamplesData(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder builderForValue) {
        if (samplesDataBuilder_ == null) {
          ensureSamplesDataIsMutable();
          samplesData_.add(index, builderForValue.build());
          onChanged();
        } else {
          samplesDataBuilder_.addMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder addAllSamplesData(
          java.lang.Iterable<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry> values) {
        if (samplesDataBuilder_ == null) {
          ensureSamplesDataIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, samplesData_);
          onChanged();
        } else {
          samplesDataBuilder_.addAllMessages(values);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder clearSamplesData() {
        if (samplesDataBuilder_ == null) {
          samplesData_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000010);
          onChanged();
        } else {
          samplesDataBuilder_.clear();
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public Builder removeSamplesData(int index) {
        if (samplesDataBuilder_ == null) {
          ensureSamplesDataIsMutable();
          samplesData_.remove(index);
          onChanged();
        } else {
          samplesDataBuilder_.remove(index);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder getSamplesDataBuilder(
          int index) {
        return getSamplesDataFieldBuilder().getBuilder(index);
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder getSamplesDataOrBuilder(
          int index) {
        if (samplesDataBuilder_ == null) {
          return samplesData_.get(index);  } else {
          return samplesDataBuilder_.getMessageOrBuilder(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder> 
           getSamplesDataOrBuilderList() {
        if (samplesDataBuilder_ != null) {
          return samplesDataBuilder_.getMessageOrBuilderList();
        } else {
          return java.util.Collections.unmodifiableList(samplesData_);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder addSamplesDataBuilder() {
        return getSamplesDataFieldBuilder().addBuilder(
            org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder addSamplesDataBuilder(
          int index) {
        return getSamplesDataFieldBuilder().addBuilder(
            index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry.SamplesDataInfoEntry samplesData = 5;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder> 
           getSamplesDataBuilderList() {
        return getSamplesDataFieldBuilder().getBuilderList();
      }
      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder> 
          getSamplesDataFieldBuilder() {
        if (samplesDataBuilder_ == null) {
          samplesDataBuilder_ = new com.google.protobuf.RepeatedFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.SamplesDataInfoEntryOrBuilder>(
                  samplesData_,
                  ((bitField0_ & 0x00000010) == 0x00000010),
                  getParentForChildren(),
                  isClean());
          samplesData_ = null;
        }
        return samplesDataBuilder_;
      }

      private com.google.protobuf.MapField<
          java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> stats_;
      private com.google.protobuf.MapField<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
      internalGetStats() {
        if (stats_ == null) {
          return com.google.protobuf.MapField.emptyMapField(
              StatsDefaultEntryHolder.defaultEntry);
       }
        return stats_;
      }
      private com.google.protobuf.MapField<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
      internalGetMutableStats() {
        onChanged();;
        if (stats_ == null) {
          stats_ = com.google.protobuf.MapField.newMapField(
              StatsDefaultEntryHolder.defaultEntry);
        }
        if (!stats_.isMutable()) {
          stats_ = stats_.copy();
        }
        return stats_;
      }
      /**
       * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.VariantStats&gt; stats = 6;</code>
       */
      public java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> getStats() {
        return internalGetStats().getMap();
      }
      /**
       * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.VariantStats&gt; stats = 6;</code>
       */
      public java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats>
      getMutableStats() {
        return internalGetMutableStats().getMutableMap();
      }
      /**
       * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.VariantStats&gt; stats = 6;</code>
       */
      public Builder putAllStats(
          java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantStats> values) {
        getMutableStats().putAll(values);
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VariantSourceEntry)
    private static final org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry();
    }

    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<VariantSourceEntry>
        PARSER = new com.google.protobuf.AbstractParser<VariantSourceEntry>() {
      public VariantSourceEntry parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new VariantSourceEntry(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<VariantSourceEntry> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<VariantSourceEntry> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface GenotypeOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.Genotype)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional string reference = 1;</code>
     */
    java.lang.String getReference();
    /**
     * <code>optional string reference = 1;</code>
     */
    com.google.protobuf.ByteString
        getReferenceBytes();

    /**
     * <code>optional string alternate = 2;</code>
     */
    java.lang.String getAlternate();
    /**
     * <code>optional string alternate = 2;</code>
     */
    com.google.protobuf.ByteString
        getAlternateBytes();

    /**
     * <code>repeated int32 allelesIdx = 3;</code>
     */
    java.util.List<java.lang.Integer> getAllelesIdxList();
    /**
     * <code>repeated int32 allelesIdx = 3;</code>
     */
    int getAllelesIdxCount();
    /**
     * <code>repeated int32 allelesIdx = 3;</code>
     */
    int getAllelesIdx(int index);

    /**
     * <code>optional bool phased = 4;</code>
     */
    boolean getPhased();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Genotype}
   */
  public  static final class Genotype extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.Genotype)
      GenotypeOrBuilder {
    // Use Genotype.newBuilder() to construct.
    private Genotype(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private Genotype() {
      reference_ = "";
      alternate_ = "";
      allelesIdx_ = java.util.Collections.emptyList();
      phased_ = false;
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private Genotype(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();

              reference_ = s;
              break;
            }
            case 18: {
              String s = input.readStringRequireUtf8();

              alternate_ = s;
              break;
            }
            case 24: {
              if (!((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
                allelesIdx_ = new java.util.ArrayList<java.lang.Integer>();
                mutable_bitField0_ |= 0x00000004;
              }
              allelesIdx_.add(input.readInt32());
              break;
            }
            case 26: {
              int length = input.readRawVarint32();
              int limit = input.pushLimit(length);
              if (!((mutable_bitField0_ & 0x00000004) == 0x00000004) && input.getBytesUntilLimit() > 0) {
                allelesIdx_ = new java.util.ArrayList<java.lang.Integer>();
                mutable_bitField0_ |= 0x00000004;
              }
              while (input.getBytesUntilLimit() > 0) {
                allelesIdx_.add(input.readInt32());
              }
              input.popLimit(limit);
              break;
            }
            case 32: {

              phased_ = input.readBool();
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
          allelesIdx_ = java.util.Collections.unmodifiableList(allelesIdx_);
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype.class, org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype.Builder.class);
    }

    private int bitField0_;
    public static final int REFERENCE_FIELD_NUMBER = 1;
    private volatile java.lang.Object reference_;
    /**
     * <code>optional string reference = 1;</code>
     */
    public java.lang.String getReference() {
      java.lang.Object ref = reference_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        reference_ = s;
        return s;
      }
    }
    /**
     * <code>optional string reference = 1;</code>
     */
    public com.google.protobuf.ByteString
        getReferenceBytes() {
      java.lang.Object ref = reference_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        reference_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int ALTERNATE_FIELD_NUMBER = 2;
    private volatile java.lang.Object alternate_;
    /**
     * <code>optional string alternate = 2;</code>
     */
    public java.lang.String getAlternate() {
      java.lang.Object ref = alternate_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        alternate_ = s;
        return s;
      }
    }
    /**
     * <code>optional string alternate = 2;</code>
     */
    public com.google.protobuf.ByteString
        getAlternateBytes() {
      java.lang.Object ref = alternate_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        alternate_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int ALLELESIDX_FIELD_NUMBER = 3;
    private java.util.List<java.lang.Integer> allelesIdx_;
    /**
     * <code>repeated int32 allelesIdx = 3;</code>
     */
    public java.util.List<java.lang.Integer>
        getAllelesIdxList() {
      return allelesIdx_;
    }
    /**
     * <code>repeated int32 allelesIdx = 3;</code>
     */
    public int getAllelesIdxCount() {
      return allelesIdx_.size();
    }
    /**
     * <code>repeated int32 allelesIdx = 3;</code>
     */
    public int getAllelesIdx(int index) {
      return allelesIdx_.get(index);
    }
    private int allelesIdxMemoizedSerializedSize = -1;

    public static final int PHASED_FIELD_NUMBER = 4;
    private boolean phased_;
    /**
     * <code>optional bool phased = 4;</code>
     */
    public boolean getPhased() {
      return phased_;
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      getSerializedSize();
      if (!getReferenceBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, reference_);
      }
      if (!getAlternateBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 2, alternate_);
      }
      if (getAllelesIdxList().size() > 0) {
        output.writeRawVarint32(26);
        output.writeRawVarint32(allelesIdxMemoizedSerializedSize);
      }
      for (int i = 0; i < allelesIdx_.size(); i++) {
        output.writeInt32NoTag(allelesIdx_.get(i));
      }
      if (phased_ != false) {
        output.writeBool(4, phased_);
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!getReferenceBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(1, reference_);
      }
      if (!getAlternateBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(2, alternate_);
      }
      {
        int dataSize = 0;
        for (int i = 0; i < allelesIdx_.size(); i++) {
          dataSize += com.google.protobuf.CodedOutputStream
            .computeInt32SizeNoTag(allelesIdx_.get(i));
        }
        size += dataSize;
        if (!getAllelesIdxList().isEmpty()) {
          size += 1;
          size += com.google.protobuf.CodedOutputStream
              .computeInt32SizeNoTag(dataSize);
        }
        allelesIdxMemoizedSerializedSize = dataSize;
      }
      if (phased_ != false) {
        size += com.google.protobuf.CodedOutputStream
          .computeBoolSize(4, phased_);
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Genotype}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.Genotype)
        org.opencb.biodata.models.variant.protobuf.VariantProto.GenotypeOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype.class, org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        reference_ = "";

        alternate_ = "";

        allelesIdx_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000004);
        phased_ = false;

        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype build() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype result = new org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.reference_ = reference_;
        result.alternate_ = alternate_;
        if (((bitField0_ & 0x00000004) == 0x00000004)) {
          allelesIdx_ = java.util.Collections.unmodifiableList(allelesIdx_);
          bitField0_ = (bitField0_ & ~0x00000004);
        }
        result.allelesIdx_ = allelesIdx_;
        result.phased_ = phased_;
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype.getDefaultInstance()) return this;
        if (!other.getReference().isEmpty()) {
          reference_ = other.reference_;
          onChanged();
        }
        if (!other.getAlternate().isEmpty()) {
          alternate_ = other.alternate_;
          onChanged();
        }
        if (!other.allelesIdx_.isEmpty()) {
          if (allelesIdx_.isEmpty()) {
            allelesIdx_ = other.allelesIdx_;
            bitField0_ = (bitField0_ & ~0x00000004);
          } else {
            ensureAllelesIdxIsMutable();
            allelesIdx_.addAll(other.allelesIdx_);
          }
          onChanged();
        }
        if (other.getPhased() != false) {
          setPhased(other.getPhased());
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private java.lang.Object reference_ = "";
      /**
       * <code>optional string reference = 1;</code>
       */
      public java.lang.String getReference() {
        java.lang.Object ref = reference_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          reference_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string reference = 1;</code>
       */
      public com.google.protobuf.ByteString
          getReferenceBytes() {
        java.lang.Object ref = reference_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          reference_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string reference = 1;</code>
       */
      public Builder setReference(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        reference_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string reference = 1;</code>
       */
      public Builder clearReference() {
        
        reference_ = getDefaultInstance().getReference();
        onChanged();
        return this;
      }
      /**
       * <code>optional string reference = 1;</code>
       */
      public Builder setReferenceBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        reference_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object alternate_ = "";
      /**
       * <code>optional string alternate = 2;</code>
       */
      public java.lang.String getAlternate() {
        java.lang.Object ref = alternate_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          alternate_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string alternate = 2;</code>
       */
      public com.google.protobuf.ByteString
          getAlternateBytes() {
        java.lang.Object ref = alternate_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          alternate_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string alternate = 2;</code>
       */
      public Builder setAlternate(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        alternate_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string alternate = 2;</code>
       */
      public Builder clearAlternate() {
        
        alternate_ = getDefaultInstance().getAlternate();
        onChanged();
        return this;
      }
      /**
       * <code>optional string alternate = 2;</code>
       */
      public Builder setAlternateBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        alternate_ = value;
        onChanged();
        return this;
      }

      private java.util.List<java.lang.Integer> allelesIdx_ = java.util.Collections.emptyList();
      private void ensureAllelesIdxIsMutable() {
        if (!((bitField0_ & 0x00000004) == 0x00000004)) {
          allelesIdx_ = new java.util.ArrayList<java.lang.Integer>(allelesIdx_);
          bitField0_ |= 0x00000004;
         }
      }
      /**
       * <code>repeated int32 allelesIdx = 3;</code>
       */
      public java.util.List<java.lang.Integer>
          getAllelesIdxList() {
        return java.util.Collections.unmodifiableList(allelesIdx_);
      }
      /**
       * <code>repeated int32 allelesIdx = 3;</code>
       */
      public int getAllelesIdxCount() {
        return allelesIdx_.size();
      }
      /**
       * <code>repeated int32 allelesIdx = 3;</code>
       */
      public int getAllelesIdx(int index) {
        return allelesIdx_.get(index);
      }
      /**
       * <code>repeated int32 allelesIdx = 3;</code>
       */
      public Builder setAllelesIdx(
          int index, int value) {
        ensureAllelesIdxIsMutable();
        allelesIdx_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated int32 allelesIdx = 3;</code>
       */
      public Builder addAllelesIdx(int value) {
        ensureAllelesIdxIsMutable();
        allelesIdx_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated int32 allelesIdx = 3;</code>
       */
      public Builder addAllAllelesIdx(
          java.lang.Iterable<? extends java.lang.Integer> values) {
        ensureAllelesIdxIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, allelesIdx_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated int32 allelesIdx = 3;</code>
       */
      public Builder clearAllelesIdx() {
        allelesIdx_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000004);
        onChanged();
        return this;
      }

      private boolean phased_ ;
      /**
       * <code>optional bool phased = 4;</code>
       */
      public boolean getPhased() {
        return phased_;
      }
      /**
       * <code>optional bool phased = 4;</code>
       */
      public Builder setPhased(boolean value) {
        
        phased_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional bool phased = 4;</code>
       */
      public Builder clearPhased() {
        
        phased_ = false;
        onChanged();
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.Genotype)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.Genotype)
    private static final org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype();
    }

    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<Genotype>
        PARSER = new com.google.protobuf.AbstractParser<Genotype>() {
      public Genotype parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new Genotype(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<Genotype> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<Genotype> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VariantProto.Genotype getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface VariantOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.Variant)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional string chromosome = 1;</code>
     *
     * <pre>
     **
     * Chromosome where the genomic variation occurred.
     * </pre>
     */
    java.lang.String getChromosome();
    /**
     * <code>optional string chromosome = 1;</code>
     *
     * <pre>
     **
     * Chromosome where the genomic variation occurred.
     * </pre>
     */
    com.google.protobuf.ByteString
        getChromosomeBytes();

    /**
     * <code>optional int32 start = 2;</code>
     *
     * <pre>
     **
     * Normalized position where the genomic variation starts.
     * &lt;ul&gt;
     * &lt;li&gt;SNVs have the same start and end position&lt;/li&gt;
     * &lt;li&gt;Insertions start in the last present position: if the first nucleotide
     * is inserted in position 6, the start is position 5&lt;/li&gt;
     * &lt;li&gt;Deletions start in the first previously present position: if the first
     * deleted nucleotide is in position 6, the start is position 6&lt;/li&gt;
     * &lt;/ul&gt;
     * </pre>
     */
    int getStart();

    /**
     * <code>optional int32 end = 3;</code>
     *
     * <pre>
     **
     * Normalized position where the genomic variation ends.
     * &lt;ul&gt;
     * &lt;li&gt;SNVs have the same start and end positions&lt;/li&gt;
     * &lt;li&gt;Insertions end in the first present position: if the last nucleotide
     * is inserted in position 9, the end is position 10&lt;/li&gt;
     * &lt;li&gt;Deletions ends in the last previously present position: if the last
     * deleted nucleotide is in position 9, the end is position 9&lt;/li&gt;
     * &lt;/ul&gt;
     * </pre>
     */
    int getEnd();

    /**
     * <code>optional string reference = 4;</code>
     *
     * <pre>
     **
     * Reference allele.
     * </pre>
     */
    java.lang.String getReference();
    /**
     * <code>optional string reference = 4;</code>
     *
     * <pre>
     **
     * Reference allele.
     * </pre>
     */
    com.google.protobuf.ByteString
        getReferenceBytes();

    /**
     * <code>optional string alternate = 5;</code>
     *
     * <pre>
     **
     * Alternate allele.
     * </pre>
     */
    java.lang.String getAlternate();
    /**
     * <code>optional string alternate = 5;</code>
     *
     * <pre>
     **
     * Alternate allele.
     * </pre>
     */
    com.google.protobuf.ByteString
        getAlternateBytes();

    /**
     * <code>optional string strand = 6;</code>
     *
     * <pre>
     **
     * Reference strand for this variant
     * </pre>
     */
    java.lang.String getStrand();
    /**
     * <code>optional string strand = 6;</code>
     *
     * <pre>
     **
     * Reference strand for this variant
     * </pre>
     */
    com.google.protobuf.ByteString
        getStrandBytes();

    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    com.google.protobuf.ProtocolStringList
        getIdsList();
    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    int getIdsCount();
    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    java.lang.String getIds(int index);
    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    com.google.protobuf.ByteString
        getIdsBytes(int index);

    /**
     * <code>optional int32 length = 8;</code>
     *
     * <pre>
     **
     * Length of the genomic variation, which depends on the variation type.
     * &lt;ul&gt;
     * &lt;li&gt;SNVs have a length of 1 nucleotide&lt;/li&gt;
     * &lt;li&gt;Indels have the length of the largest allele&lt;/li&gt;
     * &lt;/ul&gt;
     * </pre>
     */
    int getLength();

    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
     *
     * <pre>
     **
     * Type of variation: single nucleotide, indel or structural variation.
     * </pre>
     */
    int getTypeValue();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
     *
     * <pre>
     **
     * Type of variation: single nucleotide, indel or structural variation.
     * </pre>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getType();

    /**
     * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry&gt; hgvss = 10;</code>
     */
    java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
    getHgvss();

    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry> 
        getStudiesList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry getStudies(int index);
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    int getStudiesCount();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder> 
        getStudiesOrBuilderList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder getStudiesOrBuilder(
        int index);

    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
     *
     * <pre>
     **
     * Annotations of the genomic variation.
     * </pre>
     */
    boolean hasAnnotation();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
     *
     * <pre>
     **
     * Annotations of the genomic variation.
     * </pre>
     */
    org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation getAnnotation();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
     *
     * <pre>
     **
     * Annotations of the genomic variation.
     * </pre>
     */
    org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotationOrBuilder getAnnotationOrBuilder();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Variant}
   */
  public  static final class Variant extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.Variant)
      VariantOrBuilder {
    // Use Variant.newBuilder() to construct.
    private Variant(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private Variant() {
      chromosome_ = "";
      start_ = 0;
      end_ = 0;
      reference_ = "";
      alternate_ = "";
      strand_ = "";
      ids_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      length_ = 0;
      type_ = 0;
      studies_ = java.util.Collections.emptyList();
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private Variant(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();

              chromosome_ = s;
              break;
            }
            case 16: {

              start_ = input.readInt32();
              break;
            }
            case 24: {

              end_ = input.readInt32();
              break;
            }
            case 34: {
              String s = input.readStringRequireUtf8();

              reference_ = s;
              break;
            }
            case 42: {
              String s = input.readStringRequireUtf8();

              alternate_ = s;
              break;
            }
            case 50: {
              String s = input.readStringRequireUtf8();

              strand_ = s;
              break;
            }
            case 58: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000040) == 0x00000040)) {
                ids_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000040;
              }
              ids_.add(s);
              break;
            }
            case 64: {

              length_ = input.readInt32();
              break;
            }
            case 72: {
              int rawValue = input.readEnum();

              type_ = rawValue;
              break;
            }
            case 82: {
              if (!((mutable_bitField0_ & 0x00000200) == 0x00000200)) {
                hgvss_ = com.google.protobuf.MapField.newMapField(
                    HgvssDefaultEntryHolder.defaultEntry);
                mutable_bitField0_ |= 0x00000200;
              }
              com.google.protobuf.MapEntry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
              hgvss = input.readMessage(
                  HgvssDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
              hgvss_.getMutableMap().put(hgvss.getKey(), hgvss.getValue());
              break;
            }
            case 90: {
              if (!((mutable_bitField0_ & 0x00000400) == 0x00000400)) {
                studies_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry>();
                mutable_bitField0_ |= 0x00000400;
              }
              studies_.add(input.readMessage(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.parser(), extensionRegistry));
              break;
            }
            case 98: {
              org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.Builder subBuilder = null;
              if (annotation_ != null) {
                subBuilder = annotation_.toBuilder();
              }
              annotation_ = input.readMessage(org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.parser(), extensionRegistry);
              if (subBuilder != null) {
                subBuilder.mergeFrom(annotation_);
                annotation_ = subBuilder.buildPartial();
              }

              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000040) == 0x00000040)) {
          ids_ = ids_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00000400) == 0x00000400)) {
          studies_ = java.util.Collections.unmodifiableList(studies_);
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 10:
          return internalGetHgvss();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.class, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.Builder.class);
    }

    public interface HgvsEntryOrBuilder extends
        // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry)
        com.google.protobuf.MessageOrBuilder {

      /**
       * <code>repeated string info = 1;</code>
       */
      com.google.protobuf.ProtocolStringList
          getInfoList();
      /**
       * <code>repeated string info = 1;</code>
       */
      int getInfoCount();
      /**
       * <code>repeated string info = 1;</code>
       */
      java.lang.String getInfo(int index);
      /**
       * <code>repeated string info = 1;</code>
       */
      com.google.protobuf.ByteString
          getInfoBytes(int index);
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry}
     *
     * <pre>
     **
     * Unique identifier following the HGVS nomenclature.
     * </pre>
     */
    public  static final class HgvsEntry extends
        com.google.protobuf.GeneratedMessage implements
        // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry)
        HgvsEntryOrBuilder {
      // Use HgvsEntry.newBuilder() to construct.
      private HgvsEntry(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
        super(builder);
      }
      private HgvsEntry() {
        info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      }

      @java.lang.Override
      public final com.google.protobuf.UnknownFieldSet
      getUnknownFields() {
        return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
      }
      private HgvsEntry(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
        this();
        int mutable_bitField0_ = 0;
        try {
          boolean done = false;
          while (!done) {
            int tag = input.readTag();
            switch (tag) {
              case 0:
                done = true;
                break;
              default: {
                if (!input.skipField(tag)) {
                  done = true;
                }
                break;
              }
              case 10: {
                String s = input.readStringRequireUtf8();
                if (!((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
                  info_ = new com.google.protobuf.LazyStringArrayList();
                  mutable_bitField0_ |= 0x00000001;
                }
                info_.add(s);
                break;
              }
            }
          }
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          throw new RuntimeException(e.setUnfinishedMessage(this));
        } catch (java.io.IOException e) {
          throw new RuntimeException(
              new com.google.protobuf.InvalidProtocolBufferException(
                  e.getMessage()).setUnfinishedMessage(this));
        } finally {
          if (((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
            info_ = info_.getUnmodifiableView();
          }
          makeExtensionsImmutable();
        }
      }
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.Builder.class);
      }

      public static final int INFO_FIELD_NUMBER = 1;
      private com.google.protobuf.LazyStringList info_;
      /**
       * <code>repeated string info = 1;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getInfoList() {
        return info_;
      }
      /**
       * <code>repeated string info = 1;</code>
       */
      public int getInfoCount() {
        return info_.size();
      }
      /**
       * <code>repeated string info = 1;</code>
       */
      public java.lang.String getInfo(int index) {
        return info_.get(index);
      }
      /**
       * <code>repeated string info = 1;</code>
       */
      public com.google.protobuf.ByteString
          getInfoBytes(int index) {
        return info_.getByteString(index);
      }

      private byte memoizedIsInitialized = -1;
      public final boolean isInitialized() {
        byte isInitialized = memoizedIsInitialized;
        if (isInitialized == 1) return true;
        if (isInitialized == 0) return false;

        memoizedIsInitialized = 1;
        return true;
      }

      public void writeTo(com.google.protobuf.CodedOutputStream output)
                          throws java.io.IOException {
        for (int i = 0; i < info_.size(); i++) {
          com.google.protobuf.GeneratedMessage.writeString(output, 1, info_.getRaw(i));
        }
      }

      public int getSerializedSize() {
        int size = memoizedSize;
        if (size != -1) return size;

        size = 0;
        {
          int dataSize = 0;
          for (int i = 0; i < info_.size(); i++) {
            dataSize += computeStringSizeNoTag(info_.getRaw(i));
          }
          size += dataSize;
          size += 1 * getInfoList().size();
        }
        memoizedSize = size;
        return size;
      }

      private static final long serialVersionUID = 0L;
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(
          com.google.protobuf.ByteString data)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(
          com.google.protobuf.ByteString data,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(byte[] data)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(
          byte[] data,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        return PARSER.parseFrom(data, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(java.io.InputStream input)
          throws java.io.IOException {
        return PARSER.parseFrom(input);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(
          java.io.InputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        return PARSER.parseFrom(input, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseDelimitedFrom(java.io.InputStream input)
          throws java.io.IOException {
        return PARSER.parseDelimitedFrom(input);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseDelimitedFrom(
          java.io.InputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        return PARSER.parseDelimitedFrom(input, extensionRegistry);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(
          com.google.protobuf.CodedInputStream input)
          throws java.io.IOException {
        return PARSER.parseFrom(input);
      }
      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parseFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        return PARSER.parseFrom(input, extensionRegistry);
      }

      public Builder newBuilderForType() { return newBuilder(); }
      public static Builder newBuilder() {
        return DEFAULT_INSTANCE.toBuilder();
      }
      public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry prototype) {
        return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
      }
      public Builder toBuilder() {
        return this == DEFAULT_INSTANCE
            ? new Builder() : new Builder().mergeFrom(this);
      }

      @java.lang.Override
      protected Builder newBuilderForType(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        Builder builder = new Builder(parent);
        return builder;
      }
      /**
       * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry}
       *
       * <pre>
       **
       * Unique identifier following the HGVS nomenclature.
       * </pre>
       */
      public static final class Builder extends
          com.google.protobuf.GeneratedMessage.Builder<Builder> implements
          // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry)
          org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntryOrBuilder {
        public static final com.google.protobuf.Descriptors.Descriptor
            getDescriptor() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_descriptor;
        }

        protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
            internalGetFieldAccessorTable() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_fieldAccessorTable
              .ensureFieldAccessorsInitialized(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.class, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.Builder.class);
        }

        // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.newBuilder()
        private Builder() {
          maybeForceBuilderInitialization();
        }

        private Builder(
            com.google.protobuf.GeneratedMessage.BuilderParent parent) {
          super(parent);
          maybeForceBuilderInitialization();
        }
        private void maybeForceBuilderInitialization() {
          if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
          }
        }
        public Builder clear() {
          super.clear();
          info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
          bitField0_ = (bitField0_ & ~0x00000001);
          return this;
        }

        public com.google.protobuf.Descriptors.Descriptor
            getDescriptorForType() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_descriptor;
        }

        public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry getDefaultInstanceForType() {
          return org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.getDefaultInstance();
        }

        public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry build() {
          org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry result = buildPartial();
          if (!result.isInitialized()) {
            throw newUninitializedMessageException(result);
          }
          return result;
        }

        public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry buildPartial() {
          org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry result = new org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry(this);
          int from_bitField0_ = bitField0_;
          if (((bitField0_ & 0x00000001) == 0x00000001)) {
            info_ = info_.getUnmodifiableView();
            bitField0_ = (bitField0_ & ~0x00000001);
          }
          result.info_ = info_;
          onBuilt();
          return result;
        }

        public Builder mergeFrom(com.google.protobuf.Message other) {
          if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry) {
            return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry)other);
          } else {
            super.mergeFrom(other);
            return this;
          }
        }

        public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry other) {
          if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.getDefaultInstance()) return this;
          if (!other.info_.isEmpty()) {
            if (info_.isEmpty()) {
              info_ = other.info_;
              bitField0_ = (bitField0_ & ~0x00000001);
            } else {
              ensureInfoIsMutable();
              info_.addAll(other.info_);
            }
            onChanged();
          }
          onChanged();
          return this;
        }

        public final boolean isInitialized() {
          return true;
        }

        public Builder mergeFrom(
            com.google.protobuf.CodedInputStream input,
            com.google.protobuf.ExtensionRegistryLite extensionRegistry)
            throws java.io.IOException {
          org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry parsedMessage = null;
          try {
            parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
          } catch (com.google.protobuf.InvalidProtocolBufferException e) {
            parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry) e.getUnfinishedMessage();
            throw e;
          } finally {
            if (parsedMessage != null) {
              mergeFrom(parsedMessage);
            }
          }
          return this;
        }
        private int bitField0_;

        private com.google.protobuf.LazyStringList info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        private void ensureInfoIsMutable() {
          if (!((bitField0_ & 0x00000001) == 0x00000001)) {
            info_ = new com.google.protobuf.LazyStringArrayList(info_);
            bitField0_ |= 0x00000001;
           }
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public com.google.protobuf.ProtocolStringList
            getInfoList() {
          return info_.getUnmodifiableView();
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public int getInfoCount() {
          return info_.size();
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public java.lang.String getInfo(int index) {
          return info_.get(index);
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public com.google.protobuf.ByteString
            getInfoBytes(int index) {
          return info_.getByteString(index);
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder setInfo(
            int index, java.lang.String value) {
          if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoIsMutable();
          info_.set(index, value);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder addInfo(
            java.lang.String value) {
          if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoIsMutable();
          info_.add(value);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder addAllInfo(
            java.lang.Iterable<java.lang.String> values) {
          ensureInfoIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, info_);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder clearInfo() {
          info_ = com.google.protobuf.LazyStringArrayList.EMPTY;
          bitField0_ = (bitField0_ & ~0x00000001);
          onChanged();
          return this;
        }
        /**
         * <code>repeated string info = 1;</code>
         */
        public Builder addInfoBytes(
            com.google.protobuf.ByteString value) {
          if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
          ensureInfoIsMutable();
          info_.add(value);
          onChanged();
          return this;
        }
        public final Builder setUnknownFields(
            final com.google.protobuf.UnknownFieldSet unknownFields) {
          return this;
        }

        public final Builder mergeUnknownFields(
            final com.google.protobuf.UnknownFieldSet unknownFields) {
          return this;
        }


        // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry)
      }

      // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry)
      private static final org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry DEFAULT_INSTANCE;
      static {
        DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry();
      }

      public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry getDefaultInstance() {
        return DEFAULT_INSTANCE;
      }

      private static final com.google.protobuf.Parser<HgvsEntry>
          PARSER = new com.google.protobuf.AbstractParser<HgvsEntry>() {
        public HgvsEntry parsePartialFrom(
            com.google.protobuf.CodedInputStream input,
            com.google.protobuf.ExtensionRegistryLite extensionRegistry)
            throws com.google.protobuf.InvalidProtocolBufferException {
          try {
            return new HgvsEntry(input, extensionRegistry);
          } catch (RuntimeException e) {
            if (e.getCause() instanceof
                com.google.protobuf.InvalidProtocolBufferException) {
              throw (com.google.protobuf.InvalidProtocolBufferException)
                  e.getCause();
            }
            throw e;
          }
        }
      };

      public static com.google.protobuf.Parser<HgvsEntry> parser() {
        return PARSER;
      }

      @java.lang.Override
      public com.google.protobuf.Parser<HgvsEntry> getParserForType() {
        return PARSER;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry getDefaultInstanceForType() {
        return DEFAULT_INSTANCE;
      }

    }

    private int bitField0_;
    public static final int CHROMOSOME_FIELD_NUMBER = 1;
    private volatile java.lang.Object chromosome_;
    /**
     * <code>optional string chromosome = 1;</code>
     *
     * <pre>
     **
     * Chromosome where the genomic variation occurred.
     * </pre>
     */
    public java.lang.String getChromosome() {
      java.lang.Object ref = chromosome_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        chromosome_ = s;
        return s;
      }
    }
    /**
     * <code>optional string chromosome = 1;</code>
     *
     * <pre>
     **
     * Chromosome where the genomic variation occurred.
     * </pre>
     */
    public com.google.protobuf.ByteString
        getChromosomeBytes() {
      java.lang.Object ref = chromosome_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        chromosome_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int START_FIELD_NUMBER = 2;
    private int start_;
    /**
     * <code>optional int32 start = 2;</code>
     *
     * <pre>
     **
     * Normalized position where the genomic variation starts.
     * &lt;ul&gt;
     * &lt;li&gt;SNVs have the same start and end position&lt;/li&gt;
     * &lt;li&gt;Insertions start in the last present position: if the first nucleotide
     * is inserted in position 6, the start is position 5&lt;/li&gt;
     * &lt;li&gt;Deletions start in the first previously present position: if the first
     * deleted nucleotide is in position 6, the start is position 6&lt;/li&gt;
     * &lt;/ul&gt;
     * </pre>
     */
    public int getStart() {
      return start_;
    }

    public static final int END_FIELD_NUMBER = 3;
    private int end_;
    /**
     * <code>optional int32 end = 3;</code>
     *
     * <pre>
     **
     * Normalized position where the genomic variation ends.
     * &lt;ul&gt;
     * &lt;li&gt;SNVs have the same start and end positions&lt;/li&gt;
     * &lt;li&gt;Insertions end in the first present position: if the last nucleotide
     * is inserted in position 9, the end is position 10&lt;/li&gt;
     * &lt;li&gt;Deletions ends in the last previously present position: if the last
     * deleted nucleotide is in position 9, the end is position 9&lt;/li&gt;
     * &lt;/ul&gt;
     * </pre>
     */
    public int getEnd() {
      return end_;
    }

    public static final int REFERENCE_FIELD_NUMBER = 4;
    private volatile java.lang.Object reference_;
    /**
     * <code>optional string reference = 4;</code>
     *
     * <pre>
     **
     * Reference allele.
     * </pre>
     */
    public java.lang.String getReference() {
      java.lang.Object ref = reference_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        reference_ = s;
        return s;
      }
    }
    /**
     * <code>optional string reference = 4;</code>
     *
     * <pre>
     **
     * Reference allele.
     * </pre>
     */
    public com.google.protobuf.ByteString
        getReferenceBytes() {
      java.lang.Object ref = reference_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        reference_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int ALTERNATE_FIELD_NUMBER = 5;
    private volatile java.lang.Object alternate_;
    /**
     * <code>optional string alternate = 5;</code>
     *
     * <pre>
     **
     * Alternate allele.
     * </pre>
     */
    public java.lang.String getAlternate() {
      java.lang.Object ref = alternate_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        alternate_ = s;
        return s;
      }
    }
    /**
     * <code>optional string alternate = 5;</code>
     *
     * <pre>
     **
     * Alternate allele.
     * </pre>
     */
    public com.google.protobuf.ByteString
        getAlternateBytes() {
      java.lang.Object ref = alternate_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        alternate_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int STRAND_FIELD_NUMBER = 6;
    private volatile java.lang.Object strand_;
    /**
     * <code>optional string strand = 6;</code>
     *
     * <pre>
     **
     * Reference strand for this variant
     * </pre>
     */
    public java.lang.String getStrand() {
      java.lang.Object ref = strand_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        strand_ = s;
        return s;
      }
    }
    /**
     * <code>optional string strand = 6;</code>
     *
     * <pre>
     **
     * Reference strand for this variant
     * </pre>
     */
    public com.google.protobuf.ByteString
        getStrandBytes() {
      java.lang.Object ref = strand_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        strand_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int IDS_FIELD_NUMBER = 7;
    private com.google.protobuf.LazyStringList ids_;
    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    public com.google.protobuf.ProtocolStringList
        getIdsList() {
      return ids_;
    }
    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    public int getIdsCount() {
      return ids_.size();
    }
    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    public java.lang.String getIds(int index) {
      return ids_.get(index);
    }
    /**
     * <code>repeated string ids = 7;</code>
     *
     * <pre>
     **
     * Set of identifiers used for this genomic variation.
     * </pre>
     */
    public com.google.protobuf.ByteString
        getIdsBytes(int index) {
      return ids_.getByteString(index);
    }

    public static final int LENGTH_FIELD_NUMBER = 8;
    private int length_;
    /**
     * <code>optional int32 length = 8;</code>
     *
     * <pre>
     **
     * Length of the genomic variation, which depends on the variation type.
     * &lt;ul&gt;
     * &lt;li&gt;SNVs have a length of 1 nucleotide&lt;/li&gt;
     * &lt;li&gt;Indels have the length of the largest allele&lt;/li&gt;
     * &lt;/ul&gt;
     * </pre>
     */
    public int getLength() {
      return length_;
    }

    public static final int TYPE_FIELD_NUMBER = 9;
    private int type_;
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
     *
     * <pre>
     **
     * Type of variation: single nucleotide, indel or structural variation.
     * </pre>
     */
    public int getTypeValue() {
      return type_;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
     *
     * <pre>
     **
     * Type of variation: single nucleotide, indel or structural variation.
     * </pre>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getType() {
      org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType result = org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.valueOf(type_);
      return result == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.UNRECOGNIZED : result;
    }

    public static final int HGVSS_FIELD_NUMBER = 10;
    private static final class HgvssDefaultEntryHolder {
      static final com.google.protobuf.MapEntry<
          java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> defaultEntry =
              com.google.protobuf.MapEntry
              .<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>newDefaultInstance(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvssEntry_descriptor, 
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "",
                  com.google.protobuf.WireFormat.FieldType.MESSAGE,
                  org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry.getDefaultInstance());
    }
    private com.google.protobuf.MapField<
        java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> hgvss_;
    private com.google.protobuf.MapField<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
    internalGetHgvss() {
      if (hgvss_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            HgvssDefaultEntryHolder.defaultEntry);
     }
      return hgvss_;
    }
    /**
     * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry&gt; hgvss = 10;</code>
     */

    public java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> getHgvss() {
      return internalGetHgvss().getMap();
    }

    public static final int STUDIES_FIELD_NUMBER = 11;
    private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry> studies_;
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry> getStudiesList() {
      return studies_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder> 
        getStudiesOrBuilderList() {
      return studies_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    public int getStudiesCount() {
      return studies_.size();
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry getStudies(int index) {
      return studies_.get(index);
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
     *
     * <pre>
     **
     * Information specific to each study the variant was read from, such as
     * samples or statistics.
     * </pre>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder getStudiesOrBuilder(
        int index) {
      return studies_.get(index);
    }

    public static final int ANNOTATION_FIELD_NUMBER = 12;
    private org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation annotation_;
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
     *
     * <pre>
     **
     * Annotations of the genomic variation.
     * </pre>
     */
    public boolean hasAnnotation() {
      return annotation_ != null;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
     *
     * <pre>
     **
     * Annotations of the genomic variation.
     * </pre>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation getAnnotation() {
      return annotation_ == null ? org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.getDefaultInstance() : annotation_;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
     *
     * <pre>
     **
     * Annotations of the genomic variation.
     * </pre>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotationOrBuilder getAnnotationOrBuilder() {
      return getAnnotation();
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!getChromosomeBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, chromosome_);
      }
      if (start_ != 0) {
        output.writeInt32(2, start_);
      }
      if (end_ != 0) {
        output.writeInt32(3, end_);
      }
      if (!getReferenceBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 4, reference_);
      }
      if (!getAlternateBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 5, alternate_);
      }
      if (!getStrandBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 6, strand_);
      }
      for (int i = 0; i < ids_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 7, ids_.getRaw(i));
      }
      if (length_ != 0) {
        output.writeInt32(8, length_);
      }
      if (type_ != org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.SNP.getNumber()) {
        output.writeEnum(9, type_);
      }
      for (java.util.Map.Entry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> entry
           : internalGetHgvss().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
        hgvss = HgvssDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        output.writeMessage(10, hgvss);
      }
      for (int i = 0; i < studies_.size(); i++) {
        output.writeMessage(11, studies_.get(i));
      }
      if (annotation_ != null) {
        output.writeMessage(12, getAnnotation());
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!getChromosomeBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(1, chromosome_);
      }
      if (start_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(2, start_);
      }
      if (end_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(3, end_);
      }
      if (!getReferenceBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(4, reference_);
      }
      if (!getAlternateBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(5, alternate_);
      }
      if (!getStrandBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(6, strand_);
      }
      {
        int dataSize = 0;
        for (int i = 0; i < ids_.size(); i++) {
          dataSize += computeStringSizeNoTag(ids_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getIdsList().size();
      }
      if (length_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(8, length_);
      }
      if (type_ != org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.SNP.getNumber()) {
        size += com.google.protobuf.CodedOutputStream
          .computeEnumSize(9, type_);
      }
      for (java.util.Map.Entry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> entry
           : internalGetHgvss().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
        hgvss = HgvssDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        size += com.google.protobuf.CodedOutputStream
            .computeMessageSize(10, hgvss);
      }
      for (int i = 0; i < studies_.size(); i++) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(11, studies_.get(i));
      }
      if (annotation_ != null) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(12, getAnnotation());
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.Variant prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Variant}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.Variant)
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor;
      }

      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMapField(
          int number) {
        switch (number) {
          case 10:
            return internalGetHgvss();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMutableMapField(
          int number) {
        switch (number) {
          case 10:
            return internalGetMutableHgvss();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.class, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
          getStudiesFieldBuilder();
        }
      }
      public Builder clear() {
        super.clear();
        chromosome_ = "";

        start_ = 0;

        end_ = 0;

        reference_ = "";

        alternate_ = "";

        strand_ = "";

        ids_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000040);
        length_ = 0;

        type_ = 0;

        internalGetMutableHgvss().clear();
        if (studiesBuilder_ == null) {
          studies_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000400);
        } else {
          studiesBuilder_.clear();
        }
        if (annotationBuilder_ == null) {
          annotation_ = null;
        } else {
          annotation_ = null;
          annotationBuilder_ = null;
        }
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant build() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.Variant result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.Variant result = new org.opencb.biodata.models.variant.protobuf.VariantProto.Variant(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.chromosome_ = chromosome_;
        result.start_ = start_;
        result.end_ = end_;
        result.reference_ = reference_;
        result.alternate_ = alternate_;
        result.strand_ = strand_;
        if (((bitField0_ & 0x00000040) == 0x00000040)) {
          ids_ = ids_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000040);
        }
        result.ids_ = ids_;
        result.length_ = length_;
        result.type_ = type_;
        result.hgvss_ = internalGetHgvss();
        result.hgvss_.makeImmutable();
        if (studiesBuilder_ == null) {
          if (((bitField0_ & 0x00000400) == 0x00000400)) {
            studies_ = java.util.Collections.unmodifiableList(studies_);
            bitField0_ = (bitField0_ & ~0x00000400);
          }
          result.studies_ = studies_;
        } else {
          result.studies_ = studiesBuilder_.build();
        }
        if (annotationBuilder_ == null) {
          result.annotation_ = annotation_;
        } else {
          result.annotation_ = annotationBuilder_.build();
        }
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.Variant) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.Variant)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.Variant other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.getDefaultInstance()) return this;
        if (!other.getChromosome().isEmpty()) {
          chromosome_ = other.chromosome_;
          onChanged();
        }
        if (other.getStart() != 0) {
          setStart(other.getStart());
        }
        if (other.getEnd() != 0) {
          setEnd(other.getEnd());
        }
        if (!other.getReference().isEmpty()) {
          reference_ = other.reference_;
          onChanged();
        }
        if (!other.getAlternate().isEmpty()) {
          alternate_ = other.alternate_;
          onChanged();
        }
        if (!other.getStrand().isEmpty()) {
          strand_ = other.strand_;
          onChanged();
        }
        if (!other.ids_.isEmpty()) {
          if (ids_.isEmpty()) {
            ids_ = other.ids_;
            bitField0_ = (bitField0_ & ~0x00000040);
          } else {
            ensureIdsIsMutable();
            ids_.addAll(other.ids_);
          }
          onChanged();
        }
        if (other.getLength() != 0) {
          setLength(other.getLength());
        }
        if (other.type_ != 0) {
          setTypeValue(other.getTypeValue());
        }
        internalGetMutableHgvss().mergeFrom(
            other.internalGetHgvss());
        if (studiesBuilder_ == null) {
          if (!other.studies_.isEmpty()) {
            if (studies_.isEmpty()) {
              studies_ = other.studies_;
              bitField0_ = (bitField0_ & ~0x00000400);
            } else {
              ensureStudiesIsMutable();
              studies_.addAll(other.studies_);
            }
            onChanged();
          }
        } else {
          if (!other.studies_.isEmpty()) {
            if (studiesBuilder_.isEmpty()) {
              studiesBuilder_.dispose();
              studiesBuilder_ = null;
              studies_ = other.studies_;
              bitField0_ = (bitField0_ & ~0x00000400);
              studiesBuilder_ = 
                com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders ?
                   getStudiesFieldBuilder() : null;
            } else {
              studiesBuilder_.addAllMessages(other.studies_);
            }
          }
        }
        if (other.hasAnnotation()) {
          mergeAnnotation(other.getAnnotation());
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VariantProto.Variant parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.Variant) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private java.lang.Object chromosome_ = "";
      /**
       * <code>optional string chromosome = 1;</code>
       *
       * <pre>
       **
       * Chromosome where the genomic variation occurred.
       * </pre>
       */
      public java.lang.String getChromosome() {
        java.lang.Object ref = chromosome_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          chromosome_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string chromosome = 1;</code>
       *
       * <pre>
       **
       * Chromosome where the genomic variation occurred.
       * </pre>
       */
      public com.google.protobuf.ByteString
          getChromosomeBytes() {
        java.lang.Object ref = chromosome_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          chromosome_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string chromosome = 1;</code>
       *
       * <pre>
       **
       * Chromosome where the genomic variation occurred.
       * </pre>
       */
      public Builder setChromosome(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        chromosome_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string chromosome = 1;</code>
       *
       * <pre>
       **
       * Chromosome where the genomic variation occurred.
       * </pre>
       */
      public Builder clearChromosome() {
        
        chromosome_ = getDefaultInstance().getChromosome();
        onChanged();
        return this;
      }
      /**
       * <code>optional string chromosome = 1;</code>
       *
       * <pre>
       **
       * Chromosome where the genomic variation occurred.
       * </pre>
       */
      public Builder setChromosomeBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        chromosome_ = value;
        onChanged();
        return this;
      }

      private int start_ ;
      /**
       * <code>optional int32 start = 2;</code>
       *
       * <pre>
       **
       * Normalized position where the genomic variation starts.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have the same start and end position&lt;/li&gt;
       * &lt;li&gt;Insertions start in the last present position: if the first nucleotide
       * is inserted in position 6, the start is position 5&lt;/li&gt;
       * &lt;li&gt;Deletions start in the first previously present position: if the first
       * deleted nucleotide is in position 6, the start is position 6&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public int getStart() {
        return start_;
      }
      /**
       * <code>optional int32 start = 2;</code>
       *
       * <pre>
       **
       * Normalized position where the genomic variation starts.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have the same start and end position&lt;/li&gt;
       * &lt;li&gt;Insertions start in the last present position: if the first nucleotide
       * is inserted in position 6, the start is position 5&lt;/li&gt;
       * &lt;li&gt;Deletions start in the first previously present position: if the first
       * deleted nucleotide is in position 6, the start is position 6&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public Builder setStart(int value) {
        
        start_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 start = 2;</code>
       *
       * <pre>
       **
       * Normalized position where the genomic variation starts.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have the same start and end position&lt;/li&gt;
       * &lt;li&gt;Insertions start in the last present position: if the first nucleotide
       * is inserted in position 6, the start is position 5&lt;/li&gt;
       * &lt;li&gt;Deletions start in the first previously present position: if the first
       * deleted nucleotide is in position 6, the start is position 6&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public Builder clearStart() {
        
        start_ = 0;
        onChanged();
        return this;
      }

      private int end_ ;
      /**
       * <code>optional int32 end = 3;</code>
       *
       * <pre>
       **
       * Normalized position where the genomic variation ends.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have the same start and end positions&lt;/li&gt;
       * &lt;li&gt;Insertions end in the first present position: if the last nucleotide
       * is inserted in position 9, the end is position 10&lt;/li&gt;
       * &lt;li&gt;Deletions ends in the last previously present position: if the last
       * deleted nucleotide is in position 9, the end is position 9&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public int getEnd() {
        return end_;
      }
      /**
       * <code>optional int32 end = 3;</code>
       *
       * <pre>
       **
       * Normalized position where the genomic variation ends.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have the same start and end positions&lt;/li&gt;
       * &lt;li&gt;Insertions end in the first present position: if the last nucleotide
       * is inserted in position 9, the end is position 10&lt;/li&gt;
       * &lt;li&gt;Deletions ends in the last previously present position: if the last
       * deleted nucleotide is in position 9, the end is position 9&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public Builder setEnd(int value) {
        
        end_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 end = 3;</code>
       *
       * <pre>
       **
       * Normalized position where the genomic variation ends.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have the same start and end positions&lt;/li&gt;
       * &lt;li&gt;Insertions end in the first present position: if the last nucleotide
       * is inserted in position 9, the end is position 10&lt;/li&gt;
       * &lt;li&gt;Deletions ends in the last previously present position: if the last
       * deleted nucleotide is in position 9, the end is position 9&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public Builder clearEnd() {
        
        end_ = 0;
        onChanged();
        return this;
      }

      private java.lang.Object reference_ = "";
      /**
       * <code>optional string reference = 4;</code>
       *
       * <pre>
       **
       * Reference allele.
       * </pre>
       */
      public java.lang.String getReference() {
        java.lang.Object ref = reference_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          reference_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string reference = 4;</code>
       *
       * <pre>
       **
       * Reference allele.
       * </pre>
       */
      public com.google.protobuf.ByteString
          getReferenceBytes() {
        java.lang.Object ref = reference_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          reference_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string reference = 4;</code>
       *
       * <pre>
       **
       * Reference allele.
       * </pre>
       */
      public Builder setReference(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        reference_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string reference = 4;</code>
       *
       * <pre>
       **
       * Reference allele.
       * </pre>
       */
      public Builder clearReference() {
        
        reference_ = getDefaultInstance().getReference();
        onChanged();
        return this;
      }
      /**
       * <code>optional string reference = 4;</code>
       *
       * <pre>
       **
       * Reference allele.
       * </pre>
       */
      public Builder setReferenceBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        reference_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object alternate_ = "";
      /**
       * <code>optional string alternate = 5;</code>
       *
       * <pre>
       **
       * Alternate allele.
       * </pre>
       */
      public java.lang.String getAlternate() {
        java.lang.Object ref = alternate_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          alternate_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string alternate = 5;</code>
       *
       * <pre>
       **
       * Alternate allele.
       * </pre>
       */
      public com.google.protobuf.ByteString
          getAlternateBytes() {
        java.lang.Object ref = alternate_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          alternate_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string alternate = 5;</code>
       *
       * <pre>
       **
       * Alternate allele.
       * </pre>
       */
      public Builder setAlternate(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        alternate_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string alternate = 5;</code>
       *
       * <pre>
       **
       * Alternate allele.
       * </pre>
       */
      public Builder clearAlternate() {
        
        alternate_ = getDefaultInstance().getAlternate();
        onChanged();
        return this;
      }
      /**
       * <code>optional string alternate = 5;</code>
       *
       * <pre>
       **
       * Alternate allele.
       * </pre>
       */
      public Builder setAlternateBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        alternate_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object strand_ = "";
      /**
       * <code>optional string strand = 6;</code>
       *
       * <pre>
       **
       * Reference strand for this variant
       * </pre>
       */
      public java.lang.String getStrand() {
        java.lang.Object ref = strand_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          strand_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string strand = 6;</code>
       *
       * <pre>
       **
       * Reference strand for this variant
       * </pre>
       */
      public com.google.protobuf.ByteString
          getStrandBytes() {
        java.lang.Object ref = strand_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          strand_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string strand = 6;</code>
       *
       * <pre>
       **
       * Reference strand for this variant
       * </pre>
       */
      public Builder setStrand(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        strand_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string strand = 6;</code>
       *
       * <pre>
       **
       * Reference strand for this variant
       * </pre>
       */
      public Builder clearStrand() {
        
        strand_ = getDefaultInstance().getStrand();
        onChanged();
        return this;
      }
      /**
       * <code>optional string strand = 6;</code>
       *
       * <pre>
       **
       * Reference strand for this variant
       * </pre>
       */
      public Builder setStrandBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        strand_ = value;
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList ids_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureIdsIsMutable() {
        if (!((bitField0_ & 0x00000040) == 0x00000040)) {
          ids_ = new com.google.protobuf.LazyStringArrayList(ids_);
          bitField0_ |= 0x00000040;
         }
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public com.google.protobuf.ProtocolStringList
          getIdsList() {
        return ids_.getUnmodifiableView();
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public int getIdsCount() {
        return ids_.size();
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public java.lang.String getIds(int index) {
        return ids_.get(index);
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public com.google.protobuf.ByteString
          getIdsBytes(int index) {
        return ids_.getByteString(index);
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public Builder setIds(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureIdsIsMutable();
        ids_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public Builder addIds(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureIdsIsMutable();
        ids_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public Builder addAllIds(
          java.lang.Iterable<java.lang.String> values) {
        ensureIdsIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, ids_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public Builder clearIds() {
        ids_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000040);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string ids = 7;</code>
       *
       * <pre>
       **
       * Set of identifiers used for this genomic variation.
       * </pre>
       */
      public Builder addIdsBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureIdsIsMutable();
        ids_.add(value);
        onChanged();
        return this;
      }

      private int length_ ;
      /**
       * <code>optional int32 length = 8;</code>
       *
       * <pre>
       **
       * Length of the genomic variation, which depends on the variation type.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have a length of 1 nucleotide&lt;/li&gt;
       * &lt;li&gt;Indels have the length of the largest allele&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public int getLength() {
        return length_;
      }
      /**
       * <code>optional int32 length = 8;</code>
       *
       * <pre>
       **
       * Length of the genomic variation, which depends on the variation type.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have a length of 1 nucleotide&lt;/li&gt;
       * &lt;li&gt;Indels have the length of the largest allele&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public Builder setLength(int value) {
        
        length_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 length = 8;</code>
       *
       * <pre>
       **
       * Length of the genomic variation, which depends on the variation type.
       * &lt;ul&gt;
       * &lt;li&gt;SNVs have a length of 1 nucleotide&lt;/li&gt;
       * &lt;li&gt;Indels have the length of the largest allele&lt;/li&gt;
       * &lt;/ul&gt;
       * </pre>
       */
      public Builder clearLength() {
        
        length_ = 0;
        onChanged();
        return this;
      }

      private int type_ = 0;
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
       *
       * <pre>
       **
       * Type of variation: single nucleotide, indel or structural variation.
       * </pre>
       */
      public int getTypeValue() {
        return type_;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
       *
       * <pre>
       **
       * Type of variation: single nucleotide, indel or structural variation.
       * </pre>
       */
      public Builder setTypeValue(int value) {
        type_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
       *
       * <pre>
       **
       * Type of variation: single nucleotide, indel or structural variation.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getType() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType result = org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.valueOf(type_);
        return result == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.UNRECOGNIZED : result;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
       *
       * <pre>
       **
       * Type of variation: single nucleotide, indel or structural variation.
       * </pre>
       */
      public Builder setType(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType value) {
        if (value == null) {
          throw new NullPointerException();
        }
        
        type_ = value.getNumber();
        onChanged();
        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 9;</code>
       *
       * <pre>
       **
       * Type of variation: single nucleotide, indel or structural variation.
       * </pre>
       */
      public Builder clearType() {
        
        type_ = 0;
        onChanged();
        return this;
      }

      private com.google.protobuf.MapField<
          java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> hgvss_;
      private com.google.protobuf.MapField<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
      internalGetHgvss() {
        if (hgvss_ == null) {
          return com.google.protobuf.MapField.emptyMapField(
              HgvssDefaultEntryHolder.defaultEntry);
       }
        return hgvss_;
      }
      private com.google.protobuf.MapField<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
      internalGetMutableHgvss() {
        onChanged();;
        if (hgvss_ == null) {
          hgvss_ = com.google.protobuf.MapField.newMapField(
              HgvssDefaultEntryHolder.defaultEntry);
        }
        if (!hgvss_.isMutable()) {
          hgvss_ = hgvss_.copy();
        }
        return hgvss_;
      }
      /**
       * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry&gt; hgvss = 10;</code>
       */
      public java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> getHgvss() {
        return internalGetHgvss().getMap();
      }
      /**
       * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry&gt; hgvss = 10;</code>
       */
      public java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry>
      getMutableHgvss() {
        return internalGetMutableHgvss().getMutableMap();
      }
      /**
       * <code>map&lt;string, .org.opencb.biodata.models.variant.protobuf.Variant.HgvsEntry&gt; hgvss = 10;</code>
       */
      public Builder putAllHgvss(
          java.util.Map<java.lang.String, org.opencb.biodata.models.variant.protobuf.VariantProto.Variant.HgvsEntry> values) {
        getMutableHgvss().putAll(values);
        return this;
      }

      private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry> studies_ =
        java.util.Collections.emptyList();
      private void ensureStudiesIsMutable() {
        if (!((bitField0_ & 0x00000400) == 0x00000400)) {
          studies_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry>(studies_);
          bitField0_ |= 0x00000400;
         }
      }

      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder> studiesBuilder_;

      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry> getStudiesList() {
        if (studiesBuilder_ == null) {
          return java.util.Collections.unmodifiableList(studies_);
        } else {
          return studiesBuilder_.getMessageList();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public int getStudiesCount() {
        if (studiesBuilder_ == null) {
          return studies_.size();
        } else {
          return studiesBuilder_.getCount();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry getStudies(int index) {
        if (studiesBuilder_ == null) {
          return studies_.get(index);
        } else {
          return studiesBuilder_.getMessage(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder setStudies(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry value) {
        if (studiesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureStudiesIsMutable();
          studies_.set(index, value);
          onChanged();
        } else {
          studiesBuilder_.setMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder setStudies(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder builderForValue) {
        if (studiesBuilder_ == null) {
          ensureStudiesIsMutable();
          studies_.set(index, builderForValue.build());
          onChanged();
        } else {
          studiesBuilder_.setMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder addStudies(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry value) {
        if (studiesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureStudiesIsMutable();
          studies_.add(value);
          onChanged();
        } else {
          studiesBuilder_.addMessage(value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder addStudies(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry value) {
        if (studiesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureStudiesIsMutable();
          studies_.add(index, value);
          onChanged();
        } else {
          studiesBuilder_.addMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder addStudies(
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder builderForValue) {
        if (studiesBuilder_ == null) {
          ensureStudiesIsMutable();
          studies_.add(builderForValue.build());
          onChanged();
        } else {
          studiesBuilder_.addMessage(builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder addStudies(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder builderForValue) {
        if (studiesBuilder_ == null) {
          ensureStudiesIsMutable();
          studies_.add(index, builderForValue.build());
          onChanged();
        } else {
          studiesBuilder_.addMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder addAllStudies(
          java.lang.Iterable<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry> values) {
        if (studiesBuilder_ == null) {
          ensureStudiesIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, studies_);
          onChanged();
        } else {
          studiesBuilder_.addAllMessages(values);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder clearStudies() {
        if (studiesBuilder_ == null) {
          studies_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000400);
          onChanged();
        } else {
          studiesBuilder_.clear();
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public Builder removeStudies(int index) {
        if (studiesBuilder_ == null) {
          ensureStudiesIsMutable();
          studies_.remove(index);
          onChanged();
        } else {
          studiesBuilder_.remove(index);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder getStudiesBuilder(
          int index) {
        return getStudiesFieldBuilder().getBuilder(index);
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder getStudiesOrBuilder(
          int index) {
        if (studiesBuilder_ == null) {
          return studies_.get(index);  } else {
          return studiesBuilder_.getMessageOrBuilder(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder> 
           getStudiesOrBuilderList() {
        if (studiesBuilder_ != null) {
          return studiesBuilder_.getMessageOrBuilderList();
        } else {
          return java.util.Collections.unmodifiableList(studies_);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder addStudiesBuilder() {
        return getStudiesFieldBuilder().addBuilder(
            org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder addStudiesBuilder(
          int index) {
        return getStudiesFieldBuilder().addBuilder(
            index, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VariantSourceEntry studies = 11;</code>
       *
       * <pre>
       **
       * Information specific to each study the variant was read from, such as
       * samples or statistics.
       * </pre>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder> 
           getStudiesBuilderList() {
        return getStudiesFieldBuilder().getBuilderList();
      }
      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder> 
          getStudiesFieldBuilder() {
        if (studiesBuilder_ == null) {
          studiesBuilder_ = new com.google.protobuf.RepeatedFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntry.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantSourceEntryOrBuilder>(
                  studies_,
                  ((bitField0_ & 0x00000400) == 0x00000400),
                  getParentForChildren(),
                  isClean());
          studies_ = null;
        }
        return studiesBuilder_;
      }

      private org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation annotation_ = null;
      private com.google.protobuf.SingleFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation, org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.Builder, org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotationOrBuilder> annotationBuilder_;
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public boolean hasAnnotation() {
        return annotationBuilder_ != null || annotation_ != null;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation getAnnotation() {
        if (annotationBuilder_ == null) {
          return annotation_ == null ? org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.getDefaultInstance() : annotation_;
        } else {
          return annotationBuilder_.getMessage();
        }
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public Builder setAnnotation(org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation value) {
        if (annotationBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          annotation_ = value;
          onChanged();
        } else {
          annotationBuilder_.setMessage(value);
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public Builder setAnnotation(
          org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.Builder builderForValue) {
        if (annotationBuilder_ == null) {
          annotation_ = builderForValue.build();
          onChanged();
        } else {
          annotationBuilder_.setMessage(builderForValue.build());
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public Builder mergeAnnotation(org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation value) {
        if (annotationBuilder_ == null) {
          if (annotation_ != null) {
            annotation_ =
              org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.newBuilder(annotation_).mergeFrom(value).buildPartial();
          } else {
            annotation_ = value;
          }
          onChanged();
        } else {
          annotationBuilder_.mergeFrom(value);
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public Builder clearAnnotation() {
        if (annotationBuilder_ == null) {
          annotation_ = null;
          onChanged();
        } else {
          annotation_ = null;
          annotationBuilder_ = null;
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.Builder getAnnotationBuilder() {
        
        onChanged();
        return getAnnotationFieldBuilder().getBuilder();
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotationOrBuilder getAnnotationOrBuilder() {
        if (annotationBuilder_ != null) {
          return annotationBuilder_.getMessageOrBuilder();
        } else {
          return annotation_ == null ?
              org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.getDefaultInstance() : annotation_;
        }
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantAnnotation annotation = 12;</code>
       *
       * <pre>
       **
       * Annotations of the genomic variation.
       * </pre>
       */
      private com.google.protobuf.SingleFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation, org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.Builder, org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotationOrBuilder> 
          getAnnotationFieldBuilder() {
        if (annotationBuilder_ == null) {
          annotationBuilder_ = new com.google.protobuf.SingleFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation, org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotation.Builder, org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.VariantAnnotationOrBuilder>(
                  getAnnotation(),
                  getParentForChildren(),
                  isClean());
          annotation_ = null;
        }
        return annotationBuilder_;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.Variant)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.Variant)
    private static final org.opencb.biodata.models.variant.protobuf.VariantProto.Variant DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.Variant();
    }

    public static org.opencb.biodata.models.variant.protobuf.VariantProto.Variant getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<Variant>
        PARSER = new com.google.protobuf.AbstractParser<Variant>() {
      public Variant parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new Variant(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<Variant> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<Variant> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VariantProto.Variant getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface VariantFileMetadataOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VariantFileMetadata)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional string fileId = 1;</code>
     */
    java.lang.String getFileId();
    /**
     * <code>optional string fileId = 1;</code>
     */
    com.google.protobuf.ByteString
        getFileIdBytes();

    /**
     * <code>optional string studyId = 2;</code>
     */
    java.lang.String getStudyId();
    /**
     * <code>optional string studyId = 2;</code>
     */
    com.google.protobuf.ByteString
        getStudyIdBytes();

    /**
     * <code>optional string fileName = 3;</code>
     */
    java.lang.String getFileName();
    /**
     * <code>optional string fileName = 3;</code>
     */
    com.google.protobuf.ByteString
        getFileNameBytes();

    /**
     * <code>optional string studyName = 4;</code>
     */
    java.lang.String getStudyName();
    /**
     * <code>optional string studyName = 4;</code>
     */
    com.google.protobuf.ByteString
        getStudyNameBytes();

    /**
     * <code>repeated string samples = 5;</code>
     */
    com.google.protobuf.ProtocolStringList
        getSamplesList();
    /**
     * <code>repeated string samples = 5;</code>
     */
    int getSamplesCount();
    /**
     * <code>repeated string samples = 5;</code>
     */
    java.lang.String getSamples(int index);
    /**
     * <code>repeated string samples = 5;</code>
     */
    com.google.protobuf.ByteString
        getSamplesBytes(int index);

    /**
     * <code>map&lt;string, string&gt; metadata = 6;</code>
     */
    java.util.Map<java.lang.String, java.lang.String>
    getMetadata();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantFileMetadata}
   */
  public  static final class VariantFileMetadata extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VariantFileMetadata)
      VariantFileMetadataOrBuilder {
    // Use VariantFileMetadata.newBuilder() to construct.
    private VariantFileMetadata(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private VariantFileMetadata() {
      fileId_ = "";
      studyId_ = "";
      fileName_ = "";
      studyName_ = "";
      samples_ = com.google.protobuf.LazyStringArrayList.EMPTY;
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private VariantFileMetadata(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();

              fileId_ = s;
              break;
            }
            case 18: {
              String s = input.readStringRequireUtf8();

              studyId_ = s;
              break;
            }
            case 26: {
              String s = input.readStringRequireUtf8();

              fileName_ = s;
              break;
            }
            case 34: {
              String s = input.readStringRequireUtf8();

              studyName_ = s;
              break;
            }
            case 42: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000010) == 0x00000010)) {
                samples_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000010;
              }
              samples_.add(s);
              break;
            }
            case 50: {
              if (!((mutable_bitField0_ & 0x00000020) == 0x00000020)) {
                metadata_ = com.google.protobuf.MapField.newMapField(
                    MetadataDefaultEntryHolder.defaultEntry);
                mutable_bitField0_ |= 0x00000020;
              }
              com.google.protobuf.MapEntry<java.lang.String, java.lang.String>
              metadata = input.readMessage(
                  MetadataDefaultEntryHolder.defaultEntry.getParserForType(), extensionRegistry);
              metadata_.getMutableMap().put(metadata.getKey(), metadata.getValue());
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000010) == 0x00000010)) {
          samples_ = samples_.getUnmodifiableView();
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_descriptor;
    }

    @SuppressWarnings({"rawtypes"})
    protected com.google.protobuf.MapField internalGetMapField(
        int number) {
      switch (number) {
        case 6:
          return internalGetMetadata();
        default:
          throw new RuntimeException(
              "Invalid map field number: " + number);
      }
    }
    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata.Builder.class);
    }

    private int bitField0_;
    public static final int FILEID_FIELD_NUMBER = 1;
    private volatile java.lang.Object fileId_;
    /**
     * <code>optional string fileId = 1;</code>
     */
    public java.lang.String getFileId() {
      java.lang.Object ref = fileId_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        fileId_ = s;
        return s;
      }
    }
    /**
     * <code>optional string fileId = 1;</code>
     */
    public com.google.protobuf.ByteString
        getFileIdBytes() {
      java.lang.Object ref = fileId_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        fileId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int STUDYID_FIELD_NUMBER = 2;
    private volatile java.lang.Object studyId_;
    /**
     * <code>optional string studyId = 2;</code>
     */
    public java.lang.String getStudyId() {
      java.lang.Object ref = studyId_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        studyId_ = s;
        return s;
      }
    }
    /**
     * <code>optional string studyId = 2;</code>
     */
    public com.google.protobuf.ByteString
        getStudyIdBytes() {
      java.lang.Object ref = studyId_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        studyId_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int FILENAME_FIELD_NUMBER = 3;
    private volatile java.lang.Object fileName_;
    /**
     * <code>optional string fileName = 3;</code>
     */
    public java.lang.String getFileName() {
      java.lang.Object ref = fileName_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        fileName_ = s;
        return s;
      }
    }
    /**
     * <code>optional string fileName = 3;</code>
     */
    public com.google.protobuf.ByteString
        getFileNameBytes() {
      java.lang.Object ref = fileName_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        fileName_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int STUDYNAME_FIELD_NUMBER = 4;
    private volatile java.lang.Object studyName_;
    /**
     * <code>optional string studyName = 4;</code>
     */
    public java.lang.String getStudyName() {
      java.lang.Object ref = studyName_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        studyName_ = s;
        return s;
      }
    }
    /**
     * <code>optional string studyName = 4;</code>
     */
    public com.google.protobuf.ByteString
        getStudyNameBytes() {
      java.lang.Object ref = studyName_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        studyName_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int SAMPLES_FIELD_NUMBER = 5;
    private com.google.protobuf.LazyStringList samples_;
    /**
     * <code>repeated string samples = 5;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getSamplesList() {
      return samples_;
    }
    /**
     * <code>repeated string samples = 5;</code>
     */
    public int getSamplesCount() {
      return samples_.size();
    }
    /**
     * <code>repeated string samples = 5;</code>
     */
    public java.lang.String getSamples(int index) {
      return samples_.get(index);
    }
    /**
     * <code>repeated string samples = 5;</code>
     */
    public com.google.protobuf.ByteString
        getSamplesBytes(int index) {
      return samples_.getByteString(index);
    }

    public static final int METADATA_FIELD_NUMBER = 6;
    private static final class MetadataDefaultEntryHolder {
      static final com.google.protobuf.MapEntry<
          java.lang.String, java.lang.String> defaultEntry =
              com.google.protobuf.MapEntry
              .<java.lang.String, java.lang.String>newDefaultInstance(
                  org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_MetadataEntry_descriptor, 
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "",
                  com.google.protobuf.WireFormat.FieldType.STRING,
                  "");
    }
    private com.google.protobuf.MapField<
        java.lang.String, java.lang.String> metadata_;
    private com.google.protobuf.MapField<java.lang.String, java.lang.String>
    internalGetMetadata() {
      if (metadata_ == null) {
        return com.google.protobuf.MapField.emptyMapField(
            MetadataDefaultEntryHolder.defaultEntry);
     }
      return metadata_;
    }
    /**
     * <code>map&lt;string, string&gt; metadata = 6;</code>
     */

    public java.util.Map<java.lang.String, java.lang.String> getMetadata() {
      return internalGetMetadata().getMap();
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!getFileIdBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, fileId_);
      }
      if (!getStudyIdBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 2, studyId_);
      }
      if (!getFileNameBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 3, fileName_);
      }
      if (!getStudyNameBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 4, studyName_);
      }
      for (int i = 0; i < samples_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 5, samples_.getRaw(i));
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.String> entry
           : internalGetMetadata().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.String>
        metadata = MetadataDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        output.writeMessage(6, metadata);
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!getFileIdBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(1, fileId_);
      }
      if (!getStudyIdBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(2, studyId_);
      }
      if (!getFileNameBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(3, fileName_);
      }
      if (!getStudyNameBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(4, studyName_);
      }
      {
        int dataSize = 0;
        for (int i = 0; i < samples_.size(); i++) {
          dataSize += computeStringSizeNoTag(samples_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getSamplesList().size();
      }
      for (java.util.Map.Entry<java.lang.String, java.lang.String> entry
           : internalGetMetadata().getMap().entrySet()) {
        com.google.protobuf.MapEntry<java.lang.String, java.lang.String>
        metadata = MetadataDefaultEntryHolder.defaultEntry.newBuilderForType()
            .setKey(entry.getKey())
            .setValue(entry.getValue())
            .build();
        size += com.google.protobuf.CodedOutputStream
            .computeMessageSize(6, metadata);
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VariantFileMetadata}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VariantFileMetadata)
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadataOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_descriptor;
      }

      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMapField(
          int number) {
        switch (number) {
          case 6:
            return internalGetMetadata();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      @SuppressWarnings({"rawtypes"})
      protected com.google.protobuf.MapField internalGetMutableMapField(
          int number) {
        switch (number) {
          case 6:
            return internalGetMutableMetadata();
          default:
            throw new RuntimeException(
                "Invalid map field number: " + number);
        }
      }
      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata.class, org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        fileId_ = "";

        studyId_ = "";

        fileName_ = "";

        studyName_ = "";

        samples_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000010);
        internalGetMutableMetadata().clear();
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata build() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata result = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.fileId_ = fileId_;
        result.studyId_ = studyId_;
        result.fileName_ = fileName_;
        result.studyName_ = studyName_;
        if (((bitField0_ & 0x00000010) == 0x00000010)) {
          samples_ = samples_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000010);
        }
        result.samples_ = samples_;
        result.metadata_ = internalGetMetadata();
        result.metadata_.makeImmutable();
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata.getDefaultInstance()) return this;
        if (!other.getFileId().isEmpty()) {
          fileId_ = other.fileId_;
          onChanged();
        }
        if (!other.getStudyId().isEmpty()) {
          studyId_ = other.studyId_;
          onChanged();
        }
        if (!other.getFileName().isEmpty()) {
          fileName_ = other.fileName_;
          onChanged();
        }
        if (!other.getStudyName().isEmpty()) {
          studyName_ = other.studyName_;
          onChanged();
        }
        if (!other.samples_.isEmpty()) {
          if (samples_.isEmpty()) {
            samples_ = other.samples_;
            bitField0_ = (bitField0_ & ~0x00000010);
          } else {
            ensureSamplesIsMutable();
            samples_.addAll(other.samples_);
          }
          onChanged();
        }
        internalGetMutableMetadata().mergeFrom(
            other.internalGetMetadata());
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private java.lang.Object fileId_ = "";
      /**
       * <code>optional string fileId = 1;</code>
       */
      public java.lang.String getFileId() {
        java.lang.Object ref = fileId_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          fileId_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public com.google.protobuf.ByteString
          getFileIdBytes() {
        java.lang.Object ref = fileId_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          fileId_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public Builder setFileId(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        fileId_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public Builder clearFileId() {
        
        fileId_ = getDefaultInstance().getFileId();
        onChanged();
        return this;
      }
      /**
       * <code>optional string fileId = 1;</code>
       */
      public Builder setFileIdBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        fileId_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object studyId_ = "";
      /**
       * <code>optional string studyId = 2;</code>
       */
      public java.lang.String getStudyId() {
        java.lang.Object ref = studyId_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          studyId_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string studyId = 2;</code>
       */
      public com.google.protobuf.ByteString
          getStudyIdBytes() {
        java.lang.Object ref = studyId_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          studyId_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string studyId = 2;</code>
       */
      public Builder setStudyId(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        studyId_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string studyId = 2;</code>
       */
      public Builder clearStudyId() {
        
        studyId_ = getDefaultInstance().getStudyId();
        onChanged();
        return this;
      }
      /**
       * <code>optional string studyId = 2;</code>
       */
      public Builder setStudyIdBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        studyId_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object fileName_ = "";
      /**
       * <code>optional string fileName = 3;</code>
       */
      public java.lang.String getFileName() {
        java.lang.Object ref = fileName_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          fileName_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string fileName = 3;</code>
       */
      public com.google.protobuf.ByteString
          getFileNameBytes() {
        java.lang.Object ref = fileName_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          fileName_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string fileName = 3;</code>
       */
      public Builder setFileName(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        fileName_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string fileName = 3;</code>
       */
      public Builder clearFileName() {
        
        fileName_ = getDefaultInstance().getFileName();
        onChanged();
        return this;
      }
      /**
       * <code>optional string fileName = 3;</code>
       */
      public Builder setFileNameBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        fileName_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object studyName_ = "";
      /**
       * <code>optional string studyName = 4;</code>
       */
      public java.lang.String getStudyName() {
        java.lang.Object ref = studyName_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          studyName_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string studyName = 4;</code>
       */
      public com.google.protobuf.ByteString
          getStudyNameBytes() {
        java.lang.Object ref = studyName_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          studyName_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string studyName = 4;</code>
       */
      public Builder setStudyName(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        studyName_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string studyName = 4;</code>
       */
      public Builder clearStudyName() {
        
        studyName_ = getDefaultInstance().getStudyName();
        onChanged();
        return this;
      }
      /**
       * <code>optional string studyName = 4;</code>
       */
      public Builder setStudyNameBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        studyName_ = value;
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList samples_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureSamplesIsMutable() {
        if (!((bitField0_ & 0x00000010) == 0x00000010)) {
          samples_ = new com.google.protobuf.LazyStringArrayList(samples_);
          bitField0_ |= 0x00000010;
         }
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getSamplesList() {
        return samples_.getUnmodifiableView();
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public int getSamplesCount() {
        return samples_.size();
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public java.lang.String getSamples(int index) {
        return samples_.get(index);
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public com.google.protobuf.ByteString
          getSamplesBytes(int index) {
        return samples_.getByteString(index);
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public Builder setSamples(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureSamplesIsMutable();
        samples_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public Builder addSamples(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureSamplesIsMutable();
        samples_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public Builder addAllSamples(
          java.lang.Iterable<java.lang.String> values) {
        ensureSamplesIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, samples_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public Builder clearSamples() {
        samples_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000010);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string samples = 5;</code>
       */
      public Builder addSamplesBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureSamplesIsMutable();
        samples_.add(value);
        onChanged();
        return this;
      }

      private com.google.protobuf.MapField<
          java.lang.String, java.lang.String> metadata_;
      private com.google.protobuf.MapField<java.lang.String, java.lang.String>
      internalGetMetadata() {
        if (metadata_ == null) {
          return com.google.protobuf.MapField.emptyMapField(
              MetadataDefaultEntryHolder.defaultEntry);
       }
        return metadata_;
      }
      private com.google.protobuf.MapField<java.lang.String, java.lang.String>
      internalGetMutableMetadata() {
        onChanged();;
        if (metadata_ == null) {
          metadata_ = com.google.protobuf.MapField.newMapField(
              MetadataDefaultEntryHolder.defaultEntry);
        }
        if (!metadata_.isMutable()) {
          metadata_ = metadata_.copy();
        }
        return metadata_;
      }
      /**
       * <code>map&lt;string, string&gt; metadata = 6;</code>
       */
      public java.util.Map<java.lang.String, java.lang.String> getMetadata() {
        return internalGetMetadata().getMap();
      }
      /**
       * <code>map&lt;string, string&gt; metadata = 6;</code>
       */
      public java.util.Map<java.lang.String, java.lang.String>
      getMutableMetadata() {
        return internalGetMutableMetadata().getMutableMap();
      }
      /**
       * <code>map&lt;string, string&gt; metadata = 6;</code>
       */
      public Builder putAllMetadata(
          java.util.Map<java.lang.String, java.lang.String> values) {
        getMutableMetadata().putAll(values);
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VariantFileMetadata)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VariantFileMetadata)
    private static final org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata();
    }

    public static org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<VariantFileMetadata>
        PARSER = new com.google.protobuf.AbstractParser<VariantFileMetadata>() {
      public VariantFileMetadata parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new VariantFileMetadata(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<VariantFileMetadata> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<VariantFileMetadata> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantFileMetadata getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesCountEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesCountEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesFreqEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesFreqEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_AttributesEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_AttributesEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_StatsEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_StatsEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_Variant_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvssEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvssEntry_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_MetadataEntry_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_MetadataEntry_fieldAccessorTable;

  public static com.google.protobuf.Descriptors.FileDescriptor
      getDescriptor() {
    return descriptor;
  }
  private static com.google.protobuf.Descriptors.FileDescriptor
      descriptor;
  static {
    java.lang.String[] descriptorData = {
      "\n\rvariant.proto\022*org.opencb.biodata.mode" +
      "ls.variant.protobuf\032\027variantAnnotation.p" +
      "roto\"\300\001\n\031VariantHardyWeinbergStats\022\014\n\004ch" +
      "i2\030\001 \001(\002\022\016\n\006pValue\030\002 \001(\002\022\t\n\001n\030\003 \001(\005\022\017\n\007n" +
      "_AA_11\030\004 \001(\005\022\017\n\007n_Aa_10\030\005 \001(\005\022\017\n\007n_aa_00" +
      "\030\006 \001(\005\022\017\n\007e_AA_11\030\007 \001(\002\022\017\n\007e_Aa_10\030\010 \001(\002" +
      "\022\017\n\007e_aa_00\030\t \001(\002\022\t\n\001p\030\n \001(\002\022\t\n\001q\030\013 \001(\002\"" +
      "\263\007\n\014VariantStats\022\021\n\trefAllele\030\001 \001(\t\022\021\n\ta" +
      "ltAllele\030\002 \001(\t\022\026\n\016refAlleleCount\030\003 \001(\005\022\026" +
      "\n\016altAlleleCount\030\004 \001(\005\022d\n\016genotypesCount",
      "\030\005 \003(\0132L.org.opencb.biodata.models.varia" +
      "nt.protobuf.VariantStats.GenotypesCountE" +
      "ntry\022b\n\rgenotypesFreq\030\006 \003(\0132K.org.opencb" +
      ".biodata.models.variant.protobuf.Variant" +
      "Stats.GenotypesFreqEntry\022\026\n\016missingAllel" +
      "es\030\007 \001(\005\022\030\n\020missingGenotypes\030\010 \001(\005\022\025\n\rre" +
      "fAlleleFreq\030\t \001(\002\022\025\n\raltAlleleFreq\030\n \001(\002" +
      "\022\013\n\003maf\030\013 \001(\002\022\013\n\003mgf\030\014 \001(\002\022\021\n\tmafAllele\030" +
      "\r \001(\t\022\023\n\013mgfGenotype\030\016 \001(\t\022\025\n\rpassedFilt" +
      "ers\030\017 \001(\010\022\027\n\017mendelianErrors\030\020 \001(\005\022\034\n\024ca",
      "sesPercentDominant\030\021 \001(\002\022\037\n\027controlsPerc" +
      "entDominant\030\022 \001(\002\022\035\n\025casesPercentRecessi" +
      "ve\030\023 \001(\002\022 \n\030controlsPercentRecessive\030\024 \001" +
      "(\002\022\017\n\007quality\030\025 \001(\002\022\022\n\nnumSamples\030\026 \001(\005\022" +
      "L\n\013variantType\030\027 \001(\01627.org.opencb.biodat" +
      "a.models.variant.protobuf.VariantType\022Q\n" +
      "\002hw\030\030 \001(\0132E.org.opencb.biodata.models.va" +
      "riant.protobuf.VariantHardyWeinbergStats" +
      "\0325\n\023GenotypesCountEntry\022\013\n\003key\030\001 \001(\t\022\r\n\005" +
      "value\030\002 \001(\005:\0028\001\0324\n\022GenotypesFreqEntry\022\013\n",
      "\003key\030\001 \001(\t\022\r\n\005value\030\002 \001(\002:\0028\001\"\267\001\n\tFileEn" +
      "try\022\016\n\006fileId\030\001 \001(\t\022\014\n\004call\030\002 \001(\t\022Y\n\natt" +
      "ributes\030\003 \003(\0132E.org.opencb.biodata.model" +
      "s.variant.protobuf.FileEntry.AttributesE" +
      "ntry\0321\n\017AttributesEntry\022\013\n\003key\030\001 \001(\t\022\r\n\005" +
      "value\030\002 \001(\t:\0028\001\"\352\003\n\022VariantSourceEntry\022\017" +
      "\n\007studyId\030\001 \001(\t\022D\n\005files\030\002 \003(\01325.org.ope" +
      "ncb.biodata.models.variant.protobuf.File" +
      "Entry\022\033\n\023secondaryAlternates\030\003 \003(\t\022\016\n\006fo" +
      "rmat\030\004 \003(\t\022h\n\013samplesData\030\005 \003(\0132S.org.op",
      "encb.biodata.models.variant.protobuf.Var" +
      "iantSourceEntry.SamplesDataInfoEntry\022X\n\005" +
      "stats\030\006 \003(\0132I.org.opencb.biodata.models." +
      "variant.protobuf.VariantSourceEntry.Stat" +
      "sEntry\032$\n\024SamplesDataInfoEntry\022\014\n\004info\030\001" +
      " \003(\t\032f\n\nStatsEntry\022\013\n\003key\030\001 \001(\t\022G\n\005value" +
      "\030\002 \001(\01328.org.opencb.biodata.models.varia" +
      "nt.protobuf.VariantStats:\0028\001\"T\n\010Genotype" +
      "\022\021\n\treference\030\001 \001(\t\022\021\n\talternate\030\002 \001(\t\022\022" +
      "\n\nallelesIdx\030\003 \003(\005\022\016\n\006phased\030\004 \001(\010\"\316\004\n\007V",
      "ariant\022\022\n\nchromosome\030\001 \001(\t\022\r\n\005start\030\002 \001(" +
      "\005\022\013\n\003end\030\003 \001(\005\022\021\n\treference\030\004 \001(\t\022\021\n\talt" +
      "ernate\030\005 \001(\t\022\016\n\006strand\030\006 \001(\t\022\013\n\003ids\030\007 \003(" +
      "\t\022\016\n\006length\030\010 \001(\005\022E\n\004type\030\t \001(\01627.org.op" +
      "encb.biodata.models.variant.protobuf.Var" +
      "iantType\022M\n\005hgvss\030\n \003(\0132>.org.opencb.bio" +
      "data.models.variant.protobuf.Variant.Hgv" +
      "ssEntry\022O\n\007studies\030\013 \003(\0132>.org.opencb.bi" +
      "odata.models.variant.protobuf.VariantSou" +
      "rceEntry\022Q\n\nannotation\030\014 \001(\0132=.org.openc",
      "b.biodata.models.variant.protobuf.Varian" +
      "tAnnotation\032\031\n\tHgvsEntry\022\014\n\004info\030\001 \003(\t\032k" +
      "\n\nHgvssEntry\022\013\n\003key\030\001 \001(\t\022L\n\005value\030\002 \001(\013" +
      "2=.org.opencb.biodata.models.variant.pro" +
      "tobuf.Variant.HgvsEntry:\0028\001\"\376\001\n\023VariantF" +
      "ileMetadata\022\016\n\006fileId\030\001 \001(\t\022\017\n\007studyId\030\002" +
      " \001(\t\022\020\n\010fileName\030\003 \001(\t\022\021\n\tstudyName\030\004 \001(" +
      "\t\022\017\n\007samples\030\005 \003(\t\022_\n\010metadata\030\006 \003(\0132M.o" +
      "rg.opencb.biodata.models.variant.protobu" +
      "f.VariantFileMetadata.MetadataEntry\032/\n\rM",
      "etadataEntry\022\013\n\003key\030\001 \001(\t\022\r\n\005value\030\002 \001(\t" +
      ":\0028\001*\267\001\n\013VariantType\022\007\n\003SNP\020\000\022\007\n\003SNV\020\001\022\007" +
      "\n\003MNP\020\002\022\007\n\003MNV\020\003\022\t\n\005INDEL\020\004\022\006\n\002SV\020\005\022\007\n\003C" +
      "NV\020\006\022\020\n\014NO_VARIATION\020\007\022\014\n\010SYMBOLIC\020\010\022\t\n\005" +
      "MIXED\020\t\022\r\n\tINSERTION\020\n\022\014\n\010DELETION\020\013\022\021\n\r" +
      "TRANSLOCATION\020\014\022\r\n\tINVERSION\020\r*X\n\013Allele" +
      "sCode\022\016\n\nALLELES_OK\020\000\022\023\n\017ALLELES_MISSING" +
      "\020\001\022\027\n\023MULTIPLE_ALTERNATES\020\002\022\013\n\007HAPLOID\020\003" +
      "B\016B\014VariantProtob\006proto3"
    };
    com.google.protobuf.Descriptors.FileDescriptor.InternalDescriptorAssigner assigner =
        new com.google.protobuf.Descriptors.FileDescriptor.    InternalDescriptorAssigner() {
          public com.google.protobuf.ExtensionRegistry assignDescriptors(
              com.google.protobuf.Descriptors.FileDescriptor root) {
            descriptor = root;
            return null;
          }
        };
    com.google.protobuf.Descriptors.FileDescriptor
      .internalBuildGeneratedFileFrom(descriptorData,
        new com.google.protobuf.Descriptors.FileDescriptor[] {
          org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.getDescriptor(),
        }, assigner);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_descriptor =
      getDescriptor().getMessageTypes().get(0);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantHardyWeinbergStats_descriptor,
        new java.lang.String[] { "Chi2", "PValue", "N", "NAA11", "NAa10", "NAa00", "EAA11", "EAa10", "EAa00", "P", "Q", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor =
      getDescriptor().getMessageTypes().get(1);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor,
        new java.lang.String[] { "RefAllele", "AltAllele", "RefAlleleCount", "AltAlleleCount", "GenotypesCount", "GenotypesFreq", "MissingAlleles", "MissingGenotypes", "RefAlleleFreq", "AltAlleleFreq", "Maf", "Mgf", "MafAllele", "MgfGenotype", "PassedFilters", "MendelianErrors", "CasesPercentDominant", "ControlsPercentDominant", "CasesPercentRecessive", "ControlsPercentRecessive", "Quality", "NumSamples", "VariantType", "Hw", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesCountEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor.getNestedTypes().get(0);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesCountEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesCountEntry_descriptor,
        new java.lang.String[] { "Key", "Value", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesFreqEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_descriptor.getNestedTypes().get(1);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesFreqEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantStats_GenotypesFreqEntry_descriptor,
        new java.lang.String[] { "Key", "Value", });
    internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_descriptor =
      getDescriptor().getMessageTypes().get(2);
    internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_descriptor,
        new java.lang.String[] { "FileId", "Call", "Attributes", });
    internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_AttributesEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_descriptor.getNestedTypes().get(0);
    internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_AttributesEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_FileEntry_AttributesEntry_descriptor,
        new java.lang.String[] { "Key", "Value", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor =
      getDescriptor().getMessageTypes().get(3);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor,
        new java.lang.String[] { "StudyId", "Files", "SecondaryAlternates", "Format", "SamplesData", "Stats", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor.getNestedTypes().get(0);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_SamplesDataInfoEntry_descriptor,
        new java.lang.String[] { "Info", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_StatsEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_descriptor.getNestedTypes().get(1);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_StatsEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantSourceEntry_StatsEntry_descriptor,
        new java.lang.String[] { "Key", "Value", });
    internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_descriptor =
      getDescriptor().getMessageTypes().get(4);
    internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_Genotype_descriptor,
        new java.lang.String[] { "Reference", "Alternate", "AllelesIdx", "Phased", });
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor =
      getDescriptor().getMessageTypes().get(5);
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor,
        new java.lang.String[] { "Chromosome", "Start", "End", "Reference", "Alternate", "Strand", "Ids", "Length", "Type", "Hgvss", "Studies", "Annotation", });
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor.getNestedTypes().get(0);
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvsEntry_descriptor,
        new java.lang.String[] { "Info", });
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvssEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_Variant_descriptor.getNestedTypes().get(1);
    internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvssEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_Variant_HgvssEntry_descriptor,
        new java.lang.String[] { "Key", "Value", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_descriptor =
      getDescriptor().getMessageTypes().get(6);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_descriptor,
        new java.lang.String[] { "FileId", "StudyId", "FileName", "StudyName", "Samples", "Metadata", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_MetadataEntry_descriptor =
      internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_descriptor.getNestedTypes().get(0);
    internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_MetadataEntry_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VariantFileMetadata_MetadataEntry_descriptor,
        new java.lang.String[] { "Key", "Value", });
    org.opencb.biodata.models.variant.protobuf.VariantAnnotationProto.getDescriptor();
  }

  // @@protoc_insertion_point(outer_class_scope)
}
