/**
 * Autogenerated by Avro
 * 
 * DO NOT EDIT DIRECTLY
 */
package org.opencb.biodata.external.gel_participant_6_2_0;

import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;

@SuppressWarnings("all")
@org.apache.avro.specific.AvroGenerated
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
public class CipapiFacade extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  public static final org.apache.avro.Schema SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\",\"name\":\"CipapiFacade\",\"namespace\":\"org.dummy.avro\",\"fields\":[{\"name\":\"files\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DummyFile\",\"fields\":[{\"name\":\"url\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"file_name\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"file_type\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}},{\"name\":\"number_of_samples\",\"type\":\"int\"},{\"name\":\"program\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"family_id\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"labkey_links\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},{\"name\":\"created_at\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"last_modified\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"proband\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"status\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DummyStatus\",\"fields\":[{\"name\":\"status\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"created_at\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"user\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}}},{\"name\":\"case_priority\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"assembly\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"case_id\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"interpretation_request_id\",\"type\":\"int\"},{\"name\":\"cohort_id\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"clinical_report\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DummyClinicalReport\",\"fields\":[{\"name\":\"valid\",\"type\":\"boolean\"},{\"name\":\"exit_questionnaire\",\"type\":{\"type\":\"record\",\"name\":\"DummyExit\",\"fields\":[{\"name\":\"created_at\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"exit_questionnaire_data\",\"type\":{\"type\":\"record\",\"name\":\"RareDiseaseExitQuestionnaire\",\"namespace\":\"org.gel.models.report.avro\",\"doc\":\"The rare disease program exit questionnaire\",\"fields\":[{\"name\":\"eventDate\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The date when the questionnaire was submitted\"},{\"name\":\"reporter\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The person that submitted the questionnaire\"},{\"name\":\"familyLevelQuestions\",\"type\":{\"type\":\"record\",\"name\":\"FamilyLevelQuestions\",\"doc\":\"The family level questions\",\"fields\":[{\"name\":\"caseSolvedFamily\",\"type\":{\"type\":\"enum\",\"name\":\"CaseSolvedFamily\",\"symbols\":[\"yes\",\"no\",\"partially\",\"unknown\"]},\"doc\":\"Have the results reported here explained the genetic basis of the familyâ€™s presenting phenotype(s)?\"},{\"name\":\"segregationQuestion\",\"type\":{\"type\":\"enum\",\"name\":\"SegregationQuestion\",\"symbols\":[\"yes\",\"no\"]},\"doc\":\"Have you done any segregation testing in non-participating family members?\"},{\"name\":\"additionalComments\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Comments regarding report\"}]},\"doc\":\"The set of questions at family level\"},{\"name\":\"variantGroupLevelQuestions\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"VariantGroupLevelQuestions\",\"doc\":\"The variant group level questions\",\"fields\":[{\"name\":\"variantGroup\",\"type\":\"int\",\"doc\":\"This value groups variants that together could explain the phenotype according to the mode of inheritance used.\\n        (e.g.: compound heterozygous). All the variants in the same report sharing the same value will be considered in\\n        the same group (i.e.: reported together). This value is an integer unique in the whole report.\\n        These values are only relevant within the same report.\"},{\"name\":\"variantLevelQuestions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"VariantLevelQuestions\",\"doc\":\"The variant level questions\",\"fields\":[{\"name\":\"variantCoordinates\",\"type\":{\"type\":\"record\",\"name\":\"VariantCoordinates\",\"doc\":\"The variant coordinates representing uniquely a small variant.\\n    No multi-allelic variant supported, alternate only represents one alternate allele.\",\"fields\":[{\"name\":\"chromosome\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Chromosome without \\\"chr\\\" prefix (e.g. X rather than chrX)\"},{\"name\":\"position\",\"type\":\"int\",\"doc\":\"Genomic position\"},{\"name\":\"reference\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The reference bases.\"},{\"name\":\"alternate\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The alternate bases\"},{\"name\":\"assembly\",\"type\":{\"type\":\"enum\",\"name\":\"Assembly\",\"doc\":\"The reference genome assembly\",\"symbols\":[\"GRCh38\",\"GRCh37\"]},\"doc\":\"The assembly to which this variant corresponds\"}]},\"doc\":\"Variant coordinates\"},{\"name\":\"confirmationDecision\",\"type\":{\"type\":\"enum\",\"name\":\"ConfirmationDecision\",\"symbols\":[\"yes\",\"no\",\"na\"]},\"doc\":\"Did you carry out technical confirmation of this variant via an alternative test?\"},{\"name\":\"confirmationOutcome\",\"type\":{\"type\":\"enum\",\"name\":\"ConfirmationOutcome\",\"symbols\":[\"yes\",\"no\",\"na\"]},\"doc\":\"Did the test confirm that the variant is present?\"},{\"name\":\"reportingQuestion\",\"type\":{\"type\":\"enum\",\"name\":\"ReportingQuestion\",\"symbols\":[\"yes\",\"no\",\"na\"]},\"doc\":\"Did you include the variant in your report to the clinician?\"},{\"name\":\"acmgClassification\",\"type\":{\"type\":\"enum\",\"name\":\"ACMGClassification\",\"symbols\":[\"pathogenic_variant\",\"likely_pathogenic_variant\",\"variant_of_unknown_clinical_significance\",\"likely_benign_variant\",\"benign_variant\",\"not_assessed\",\"na\"]},\"doc\":\"What ACMG pathogenicity score (1-5) did you assign to this variant?\"},{\"name\":\"publications\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Please provide PMIDs for papers which you have used to inform your assessment for this variant, separated by a `;` for multiple papers\"}]}}],\"doc\":\"Variant level questions for each of the variants in the group\"},{\"name\":\"shortTandemRepeatLevelQuestions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ShortTandemRepeatLevelQuestions\",\"doc\":\"The variant level questions\",\"fields\":[{\"name\":\"coordinates\",\"type\":{\"type\":\"record\",\"name\":\"Coordinates\",\"fields\":[{\"name\":\"assembly\",\"type\":\"Assembly\",\"doc\":\"The assembly to which this variant corresponds\"},{\"name\":\"chromosome\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Chromosome without \\\"chr\\\" prefix (e.g. X rather than chrX)\"},{\"name\":\"start\",\"type\":\"int\",\"doc\":\"Start genomic position for variant (1-based)\"},{\"name\":\"end\",\"type\":\"int\",\"doc\":\"End genomic position for variant\"},{\"name\":\"ciStart\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"ConfidenceInterval\",\"fields\":[{\"name\":\"left\",\"type\":\"int\"},{\"name\":\"right\",\"type\":\"int\"}]}]},{\"name\":\"ciEnd\",\"type\":[\"null\",\"ConfidenceInterval\"]}]},\"doc\":\"Variant coordinates\"},{\"name\":\"confirmationDecision\",\"type\":\"ConfirmationDecision\",\"doc\":\"Did you carry out technical confirmation of this variant via an alternative test?\"},{\"name\":\"confirmationOutcome\",\"type\":\"ConfirmationOutcome\",\"doc\":\"Did the test confirm that the variant is present?\"},{\"name\":\"reportingQuestion\",\"type\":\"ReportingQuestion\",\"doc\":\"Did you include the variant in your report to the clinician?\"},{\"name\":\"acmgClassification\",\"type\":\"ACMGClassification\",\"doc\":\"What ACMG pathogenicity score (1-5) did you assign to this variant?\"},{\"name\":\"publications\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Please provide PMIDs for papers which you have used to inform your assessment for this variant, separated by a `;` for multiple papers\"}]}}],\"doc\":\"STR level questions for each of the variants in the group\"},{\"name\":\"structuralVariantLevelQuestions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"StructuralVariantLevelQuestions\",\"doc\":\"Structural variant level questions\",\"fields\":[{\"name\":\"variantType\",\"type\":{\"type\":\"enum\",\"name\":\"StructuralVariantType\",\"symbols\":[\"ins\",\"dup\",\"inv\",\"amplification\",\"deletion\",\"dup_tandem\",\"del_me\",\"ins_me\"]},\"doc\":\"Structural variant type\"},{\"name\":\"coordinates\",\"type\":\"Coordinates\",\"doc\":\"Variant coordinates\"},{\"name\":\"confirmationDecision\",\"type\":\"ConfirmationDecision\",\"doc\":\"Did you carry out technical confirmation of this variant via an alternative test?\"},{\"nam","e\":\"confirmationOutcome\",\"type\":\"ConfirmationOutcome\",\"doc\":\"Did the test confirm that the variant is present?\"},{\"name\":\"reportingQuestion\",\"type\":\"ReportingQuestion\",\"doc\":\"Did you include the variant in your report to the clinician?\"},{\"name\":\"acmgClassification\",\"type\":\"ACMGClassification\",\"doc\":\"What ACMG pathogenicity score (1-5) did you assign to this variant?\"},{\"name\":\"publications\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Please provide PMIDs for papers which you have used to inform your assessment for this variant, separated by a `;` for multiple papers\"}]}}],\"doc\":\"Structural level questions for each of the variants in the group\"},{\"name\":\"actionability\",\"type\":{\"type\":\"enum\",\"name\":\"Actionability\",\"symbols\":[\"yes\",\"no\",\"not_yet\",\"na\"]},\"doc\":\"Is evidence for this variant/variant pair sufficient to use it for clinical purposes such as prenatal diagnosis or predictive testing?\"},{\"name\":\"clinicalUtility\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"enum\",\"name\":\"ClinicalUtility\",\"symbols\":[\"none\",\"change_in_medication\",\"surgical_option\",\"additional_surveillance_for_proband_or_relatives\",\"clinical_trial_eligibility\",\"informs_reproductive_choice\",\"unknown\",\"other\"]}},\"doc\":\"Has the clinical team identified any changes to clinical care which could potentially arise as a result of this variant/variant pair?\"},{\"name\":\"phenotypesSolved\",\"type\":{\"type\":\"enum\",\"name\":\"PhenotypesSolved\",\"symbols\":[\"yes\",\"no\",\"partially\",\"unknown\"]},\"doc\":\"Did you report the variant(s) as being partially or completely causative of the family's presenting phenotype(s)?\"},{\"name\":\"phenotypesExplained\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"If you indicated that the variant(s) only partially explained the familyâ€™s presenting phenotypes, please indicate which HPO terms you are confident that they DO explain\"}]}},\"doc\":\"The list of variant group level variants (ungrouped variants are to be set in single variant group)\"}]}}]}}]}}},{\"name\":\"interpretation_request_data\",\"type\":{\"type\":\"record\",\"name\":\"DummyInterpretationRequest\",\"fields\":[{\"name\":\"json_request\",\"type\":{\"type\":\"record\",\"name\":\"InterpretationRequestRD\",\"namespace\":\"org.gel.models.report.avro\",\"doc\":\"This record represents basic information for this report\",\"fields\":[{\"name\":\"versionControl\",\"type\":{\"type\":\"record\",\"name\":\"ReportVersionControl\",\"fields\":[{\"name\":\"gitVersionControl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"This is the version for the entire set of data models as referred to the Git release tag\",\"default\":\"6.1.1\"}]},\"doc\":\"Model version number\"},{\"name\":\"interpretationRequestId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Identifier for this interpretation request\"},{\"name\":\"interpretationRequestVersion\",\"type\":\"int\",\"doc\":\"Version for this interpretation request\"},{\"name\":\"internalStudyId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Internal study identifier\"},{\"name\":\"familyInternalId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Family internal identifier\"},{\"name\":\"genomeAssembly\",\"type\":\"Assembly\",\"doc\":\"This is the version of the assembly used to align the reads\"},{\"name\":\"workspace\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},\"doc\":\"The genome shall be assigned to the workspaces(projects or domains with a predefined set of users) to control user access\"},{\"name\":\"bams\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"File\",\"doc\":\"This defines a file\\n    This record is uniquely defined by the sample identfier and an URI\\n    Currently sample identifier can be a single string or a list of strings if multiple samples are associated with the same file\\n    *\",\"fields\":[{\"name\":\"sampleId\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Unique identifier(s) of the sample. For example in a multisample vcf this would have an array of all the sample identifiers\"},{\"name\":\"uriFile\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"URI path of the file\"},{\"name\":\"fileType\",\"type\":{\"type\":\"enum\",\"name\":\"FileType\",\"symbols\":[\"BAM\",\"gVCF\",\"VCF_small\",\"VCF_somatic_small\",\"VCF_CNV\",\"VCF_somatic_CNV\",\"VCF_SV\",\"VCF_somatic_SV\",\"VCF_SV_CNV\",\"SVG\",\"ANN\",\"BigWig\",\"MD5Sum\",\"ROH\",\"OTHER\",\"PARTITION\",\"VARIANT_FREQUENCIES\",\"COVERAGE\"]},\"doc\":\"The type of the file\"},{\"name\":\"md5Sum\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The MD5 checksum\"}]}}],\"doc\":\"BAMs Files\"},{\"name\":\"vcfs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"File\"}],\"doc\":\"VCFs Files where SVs and CNVs are represented\"},{\"name\":\"bigWigs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"File\"}],\"doc\":\"BigWig Files\"},{\"name\":\"pedigreeDiagram\",\"type\":[\"null\",\"File\"],\"doc\":\"Pedigree Diagram Files as an SGV\"},{\"name\":\"annotationFile\",\"type\":[\"null\",\"File\"],\"doc\":\"Variant Annotation File\"},{\"name\":\"otherFiles\",\"type\":[\"null\",{\"type\":\"map\",\"values\":\"File\",\"avro.java.string\":\"String\"}],\"doc\":\"Other files that may be vendor specific\\n        map of key: type of file, value: record of type File\"},{\"name\":\"pedigree\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Pedigree\",\"namespace\":\"org.gel.models.participant.avro\",\"doc\":\"This is the concept of a family with associated phenotypes as present in the record RDParticipant\",\"fields\":[{\"name\":\"versionControl\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VersionControl\",\"fields\":[{\"name\":\"GitVersionControl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"This is the version for the entire set of data models as referred to the Git release tag\",\"default\":\"1.3.0\"}]}],\"doc\":\"Model version number\"},{\"name\":\"LDPCode\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"LDP Code (Local Delivery Partner)\"},{\"name\":\"familyId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Family identifier which internally translate to a sample set\"},{\"name\":\"members\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"PedigreeMember\",\"doc\":\"This defines a RD Participant (demographics and pedigree information)\",\"fields\":[{\"name\":\"pedigreeId\",\"type\":[\"null\",\"int\"],\"doc\":\"Numbering used to refer to each member of the pedigree\"},{\"name\":\"isProband\",\"type\":[\"null\",\"boolean\"],\"doc\":\"If this field is true, the member should be considered the proband of this family\"},{\"name\":\"participantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"participantId.  This is the human readable ID in GMS\"},{\"name\":\"participantQCState\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"ParticipantQCState\",\"doc\":\"QCState Status\",\"symbols\":[\"noState\",\"passedMedicalReviewReadyForInterpretation\",\"passedMedicalReviewNotReadyForInterpretation\",\"queryToGel\",\"queryToGMC\",\"failed\"]}],\"doc\":\"participantQCState\"},{\"name\":\"gelSuperFamilyId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"superFamily id, this id is built as a concatenation of all\\n        families id in this superfamily i.e, fam10024_fam100457.\\n        NOTE: Retired for GMS\"},{\"name\":\"sex\",\"type\":{\"type\":\"enum\",\"name\":\"Sex\",\"doc\":\"Sex\",\"symbols\":[\"MALE\",\"FEMALE\",\"UNKNOWN\"]},\"doc\":\"Sex of the Participant\"},{\"name\":\"personKaryotypicSex\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"PersonKaryotipicSex\",\"doc\":\"Karyotipic Sex\",\"symbols\":[\"UNKNOWN\",\"XX\",\"XY\",\"XO\",\"XXY\",\"XXX\",\"XXYY\",\"XXXY\",\"XXXX\",\"XYY\",\"OTHER\"]}],\"doc\":\"Karyotypic sex of the participant as previously established or by looking at the GEL genome\"},{\"name\":\"yearOfBirth\",\"type\":[\"null\",\"int\"],\"doc\":\"Year of Birth\"},{\"name\":\"fatherId\",\"type\":[\"null\",\"int\"],\"doc\":\"refers to the pedigreeId of the father\\n        Id of the parent, if unknown then no parent is referenced. Parents may need to be entered even if no data is known\\n        about them in order to unambiguously reconstruct the pedigree.\"},{\"name\":\"motherId\",\"type\":[\"null\",\"int\"],\"doc\":\"refers to the pedigreeId of the mother\\n        Id of the parent, if unknown then no parent is referenced. Parents may need to be entered even if no data is known\\n        about them in order to unambiguously reconstruct the pedigree.\"},{\"name\":\"superFatherId\",\"type\":[\"null\",\"int\"],\"doc","\":\"this id is built using the original familyId and the original pedigreeId of the father\\n        NOTE: retired for GMS\"},{\"name\":\"superMotherId\",\"type\":[\"null\",\"int\"],\"doc\":\"this id is built using the original familyId and the original pedigreeId of the mother\\n        NOTE: Retired for GMS\"},{\"name\":\"twinGroup\",\"type\":[\"null\",\"int\"],\"doc\":\"Each twin group is numbered, i.e. all members of a group of multiparous births receive the same number\"},{\"name\":\"monozygotic\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"TernaryOption\",\"doc\":\"This defines a yes/no/unknown case\",\"symbols\":[\"yes\",\"no\",\"unknown\"]}],\"doc\":\"A property of the twinning group but should be entered for all members\"},{\"name\":\"adoptedStatus\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"AdoptedStatus\",\"doc\":\"adoptedin means adopted into the family\\n    adoptedout means child belonged to the family and was adopted out\",\"symbols\":[\"notadopted\",\"adoptedin\",\"adoptedout\"]}],\"doc\":\"Adopted Status\"},{\"name\":\"lifeStatus\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"LifeStatus\",\"doc\":\"Life Status\",\"symbols\":[\"ALIVE\",\"ABORTED\",\"DECEASED\",\"UNBORN\",\"STILLBORN\",\"MISCARRIAGE\"]}],\"doc\":\"Life Status\"},{\"name\":\"consanguineousParents\",\"type\":[\"null\",\"TernaryOption\"],\"doc\":\"The parents of this participant has a consanguineous relationship\"},{\"name\":\"affectionStatus\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"AffectionStatus\",\"doc\":\"Affection Status\",\"symbols\":[\"UNAFFECTED\",\"AFFECTED\",\"UNCERTAIN\"]}],\"doc\":\"Affection Status\"},{\"name\":\"disorderList\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Disorder\",\"doc\":\"This is quite GEL specific. This is the way is stored in ModelCatalogue and PanelApp.\\n    Currently all specific disease titles are assigned to a disease subgroup so really only specificDisease needs to be\\n    completed but we add the others for generality\",\"fields\":[{\"name\":\"diseaseGroup\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"This is Level2 Title for this disorder\"},{\"name\":\"diseaseSubGroup\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"This is Level3 Title for this disorder\"},{\"name\":\"specificDisease\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"This is Level4 Title for this disorder.\\n        In GMS, this is the clinicalIndicationFullName/clinicalIndicationCode.\"},{\"name\":\"ageOfOnset\",\"type\":[\"null\",\"float\"],\"doc\":\"Age of onset in years\"}]}}],\"doc\":\"Clinical Data (disorders). If the family member is unaffected as per affectionStatus then this list is empty.\"},{\"name\":\"hpoTermList\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"HpoTerm\",\"doc\":\"This defines an HPO term and its modifiers (possibly multiple)\\n    If HPO term presence is unknown we don't have a entry on the list\",\"fields\":[{\"name\":\"term\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Identifier of the HPO term\"},{\"name\":\"termPresence\",\"type\":[\"null\",\"TernaryOption\"],\"doc\":\"This is whether the term is present in the participant (default is unknown) yes=term is present in participant,\\n        no=term is not present\"},{\"name\":\"hpoBuildNumber\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"hpoBuildNumber\"},{\"name\":\"modifiers\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"HpoTermModifiers\",\"doc\":\"HPO Modifiers\\n    For GMS, hpoModifierCode and hpoModifierVersion will be used\",\"fields\":[{\"name\":\"laterality\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Laterality\",\"symbols\":[\"RIGHT\",\"UNILATERAL\",\"BILATERAL\",\"LEFT\"]}]},{\"name\":\"progression\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Progression\",\"symbols\":[\"PROGRESSIVE\",\"NONPROGRESSIVE\"]}]},{\"name\":\"severity\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Severity\",\"symbols\":[\"BORDERLINE\",\"MILD\",\"MODERATE\",\"SEVERE\",\"PROFOUND\"]}]},{\"name\":\"spatialPattern\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"SpatialPattern\",\"symbols\":[\"DISTAL\",\"GENERALIZED\",\"LOCALIZED\",\"PROXIMAL\"]}]},{\"name\":\"hpoModifierCode\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"hpoModifierVersion\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}],\"doc\":\"Modifier associated with the HPO term\"},{\"name\":\"ageOfOnset\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"AgeOfOnset\",\"symbols\":[\"EMBRYONAL_ONSET\",\"FETAL_ONSET\",\"NEONATAL_ONSET\",\"INFANTILE_ONSET\",\"CHILDHOOD_ONSET\",\"JUVENILE_ONSET\",\"YOUNG_ADULT_ONSET\",\"LATE_ONSET\",\"MIDDLE_AGE_ONSET\"]}],\"doc\":\"Age of onset in months\"}]}}],\"doc\":\"Clinical Data (HPO terms)\"},{\"name\":\"ancestries\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Ancestries\",\"doc\":\"Ancestries, defined as Ethnic category(ies) and Chi-square test\",\"fields\":[{\"name\":\"mothersEthnicOrigin\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"EthnicCategory\",\"doc\":\"This is the list of ethnicities in ONS16\\n\\n    * `D`:  Mixed: White and Black Caribbean\\n    * `E`:  Mixed: White and Black African\\n    * `F`:  Mixed: White and Asian\\n    * `G`:  Mixed: Any other mixed background\\n    * `A`:  White: British\\n    * `B`:  White: Irish\\n    * `C`:  White: Any other White background\\n    * `L`:  Asian or Asian British: Any other Asian background\\n    * `M`:  Black or Black British: Caribbean\\n    * `N`:  Black or Black British: African\\n    * `H`:  Asian or Asian British: Indian\\n    * `J`:  Asian or Asian British: Pakistani\\n    * `K`:  Asian or Asian British: Bangladeshi\\n    * `P`:  Black or Black British: Any other Black background\\n    * `S`:  Other Ethnic Groups: Any other ethnic group\\n    * `R`:  Other Ethnic Groups: Chinese\\n    * `Z`:  Not stated\",\"symbols\":[\"D\",\"E\",\"F\",\"G\",\"A\",\"B\",\"C\",\"L\",\"M\",\"N\",\"H\",\"J\",\"K\",\"P\",\"S\",\"R\",\"Z\"]}],\"doc\":\"Mother's Ethnic Origin\"},{\"name\":\"mothersOtherRelevantAncestry\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Mother's Ethnic Origin Description\"},{\"name\":\"fathersEthnicOrigin\",\"type\":[\"null\",\"EthnicCategory\"],\"doc\":\"Father's Ethnic Origin\"},{\"name\":\"fathersOtherRelevantAncestry\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Father's Ethnic Origin Description\"},{\"name\":\"chiSquare1KGenomesPhase3Pop\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ChiSquare1KGenomesPhase3Pop\",\"doc\":\"Chi-square test for goodness of fit of this sample to 1000 Genomes Phase 3 populations\",\"fields\":[{\"name\":\"kgSuperPopCategory\",\"type\":{\"type\":\"enum\",\"name\":\"KgSuperPopCategory\",\"doc\":\"1K Genomes project super populations\",\"symbols\":[\"AFR\",\"AMR\",\"EAS\",\"EUR\",\"SAS\"]},\"doc\":\"1K Super Population\"},{\"name\":\"kgPopCategory\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"KgPopCategory\",\"doc\":\"1K Genomes project populations\",\"symbols\":[\"ACB\",\"ASW\",\"BEB\",\"CDX\",\"CEU\",\"CHB\",\"CHS\",\"CLM\",\"ESN\",\"FIN\",\"GBR\",\"GIH\",\"GWD\",\"IBS\",\"ITU\",\"JPT\",\"KHV\",\"LWK\",\"MSL\",\"MXL\",\"PEL\",\"PJL\",\"PUR\",\"STU\",\"TSI\",\"YRI\"]}],\"doc\":\"1K Population\"},{\"name\":\"chiSquare\",\"type\":\"double\",\"doc\":\"Chi-square test for goodness of fit of this sample to this 1000 Genomes Phase 3 population\"}]}}],\"doc\":\"Chi-square test for goodness of fit of this sample to 1000 Genomes Phase 3 populations\"}]}],\"doc\":\"Participant's ancestries, defined as Mother's/Father's Ethnic Origin and Chi-square test for goodness of fit of this sample to 1000 Genomes Phase 3 populations.\\n        NOTE: for GMS this field has been deprecated in favour of clinicalEthnicities\"},{\"name\":\"consentStatus\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"ConsentStatus\",\"doc\":\"Consent Status for 100k program\",\"fields\":[{\"name\":\"programmeConsent\",\"type\":\"boolean\",\"doc\":\"Is this individual consented to the programme?\\n        It could simply be a family member that is not consented but for whom affection status is known\",\"default\":false},{\"name\":\"primaryFindingConsent\",\"type\":\"boolean\",\"doc\":\"Consent for feedback of primary findings?\",\"default\":false},{\"name\":\"secondaryFindingConsent\",\"type\":\"boolean\",\"doc\":\"Consent for secondary finding lookup\",\"default\":false},{\"name\":\"carrierStatusConsent\",\"type\":\"boolean\",\"doc\":\"Consent for carrier status check?\",\"default\":false}]}],\"doc\":\"What has this participant consented to?\\n        A participant that has been consented to the programme should also have sequence data associated with them; however\\n        this needs to be programmatically checked\"},{\"name\":\"testConsentStatus\",\"type\":[\"nul","l\",{\"type\":\"record\",\"name\":\"GmsConsentStatus\",\"doc\":\"Consent Status for GMS\",\"fields\":[{\"name\":\"programmeConsent\",\"type\":{\"type\":\"enum\",\"name\":\"GenericConsent\",\"doc\":\"clinicalEthnicities supersedes Ancestries in GMS\",\"symbols\":[\"yes\",\"no\",\"undefined\",\"not_applicable\"]},\"doc\":\"Is this individual consented to the programme? It could simply be a family member that is not consented\\n        but for whom affection status is known\"},{\"name\":\"primaryFindingConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for feedback of primary findings?\\n        RD: Primary Findings\\n        Cancer: PrimaryFindings is somatic + pertinent germline findings\"},{\"name\":\"researchConsent\",\"type\":\"GenericConsent\",\"doc\":\"Research Consent\"},{\"name\":\"healthRelatedFindingConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for secondary health related findings?\"},{\"name\":\"carrierStatusConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for carrier status check?\"},{\"name\":\"pharmacogenomicsFindingConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for pharmacogenomics consent as secondary findings?\"}]}],\"doc\":\"What has this participant consented in the context of a Genomic Test?\"},{\"name\":\"samples\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"GermlineSample\",\"doc\":\"A germline sample\",\"fields\":[{\"name\":\"sampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Sample identifier (e.g, LP00012645_5GH))\"},{\"name\":\"labSampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Lab sample identifier\"},{\"name\":\"LDPCode\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"LDP Code (Local Delivery Partner)\"},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"SampleSource\",\"doc\":\"The source of the sample\\n    NOTE: IN GMS, BONE_MARROW_ASPIRATE_TUMOUR_CELLS and BONE_MARROW_ASPIRATE_TUMOUR_SORTED_CELLS are deprecated as they have been separated into their respective biotypes\",\"symbols\":[\"AMNIOTIC_FLUID\",\"BLOOD\",\"BONE_MARROW\",\"BONE_MARROW_ASPIRATE_TUMOUR_CELLS\",\"BONE_MARROW_ASPIRATE_TUMOUR_SORTED_CELLS\",\"BUCCAL_SWAB\",\"CHORIONIC_VILLUS_SAMPLE\",\"FIBROBLAST\",\"FLUID\",\"FRESH_TISSUE_IN_CULTURE_MEDIUM\",\"OTHER\",\"SALIVA\",\"TISSUE\",\"TUMOUR\",\"URINE\"]}],\"doc\":\"Source of the sample\"},{\"name\":\"product\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Product\",\"symbols\":[\"DNA\",\"RNA\"]}],\"doc\":\"Product of the sample\"},{\"name\":\"preparationMethod\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"PreparationMethod\",\"doc\":\"In 100K, preparation Method of sample\\n    NOTE: In GMS, this field is deprecated in favour of StorageMedium and Method\",\"symbols\":[\"ASPIRATE\",\"CD128_SORTED_CELLS\",\"CD138_SORTED_CELLS\",\"EDTA\",\"FF\",\"FFPE\",\"LI_HEP\",\"ORAGENE\"]}],\"doc\":\"Preparation method\\n        NOTE: In GMS, this has been deprecated in favour of Method and storageMedium\"},{\"name\":\"programmePhase\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"ProgrammePhase\",\"symbols\":[\"CRUK\",\"OXFORD\",\"CLL\",\"IIP\",\"MAIN\",\"EXPT\"]}],\"doc\":\"Genomics England programme phase\"},{\"name\":\"clinicalSampleDateTime\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The time when the sample was received. In the format YYYY-MM-DDTHH:MM:SS+0000\"},{\"name\":\"participantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"participantUid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Participant UId of the sample\"},{\"name\":\"sampleUid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"maskedPid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"method\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Method\",\"doc\":\"In GMS, Method is defined as how the sample was taken directly from the patient\",\"symbols\":[\"ASPIRATE\",\"BIOPSY\",\"NOT_APPLICABLE\",\"RESECTION\",\"SORTED_OTHER\",\"UNKNOWN\",\"UNSORTED\",\"CD138_SORTED\"]}],\"doc\":\"In GMS, this is how the sample was extracted from the participant\"},{\"name\":\"storageMedium\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"StorageMedium\",\"doc\":\"In GMS, storage medium of sample\",\"symbols\":[\"EDTA\",\"FF\",\"LI_HEP\",\"ORAGENE\",\"FFPE\"]}],\"doc\":\"In GMS, this is what solvent/medium the sample was stored in\"},{\"name\":\"sampleType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"In GMS, this is the sampleType as entered by the clinician in TOMs\"},{\"name\":\"sampleState\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"In GMS, this is the sampleState as entered by the clinician in TOMs\"}]}}],\"doc\":\"This is an array containing all the samples that belong to this individual, e.g [\\\"LP00002255_GA4\\\"]\"},{\"name\":\"inbreedingCoefficient\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"InbreedingCoefficient\",\"doc\":\"Inbreeding coefficient\",\"fields\":[{\"name\":\"sampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"This is the sample id against which the coefficient was estimated\"},{\"name\":\"program\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Name of program used to calculate the coefficient\"},{\"name\":\"version\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Version of the programme\"},{\"name\":\"estimationMethod\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Where various methods for estimation exist, which method was used.\"},{\"name\":\"coefficient\",\"type\":\"double\",\"doc\":\"Inbreeding coefficient ideally a real number in [0,1]\"},{\"name\":\"standardError\",\"type\":[\"null\",\"double\"],\"doc\":\"Standard error of the Inbreeding coefficient\"}]}],\"doc\":\"Inbreeding Coefficient Estimation\"},{\"name\":\"additionalInformation\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"We could add a map here to store additional information for example URIs to images, ECGs, etc\\n        Null by default\"},{\"name\":\"lastMenstrualPeriod\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Last Menstrual Period\"},{\"name\":\"diagnosticDetails\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DiagnosticDetail\",\"fields\":[{\"name\":\"diagnosisCodingSystem\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Diagnosis coding system\"},{\"name\":\"diagnosisCodingSystemVersion\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis Coding System Version\"},{\"name\":\"diagnosisCode\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Diagnosis Code\"},{\"name\":\"diagnosisDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis description\"},{\"name\":\"diagnosisCertainty\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis Certainty\"},{\"name\":\"ageAtOnsetInYears\",\"type\":[\"null\",\"float\"],\"doc\":\"Age at diagnosis\"}]}}],\"doc\":\"Additional set of diagnostic ontology terms\"},{\"name\":\"participantUid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"ParticipantGUID in GMS\"},{\"name\":\"clinicalEthnicities\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"enum\",\"name\":\"ClinicalEthnicity\",\"doc\":\"* A     British, Mixed British\\n    * B     Irish\\n    * C     Any other White background\\n    * C2    Northern Irish\\n    * C3    Other white, white unspecified\\n    * CA    English\\n    * CB    Scottish\\n    * CC    Welsh\\n    * CD    Cornish\\n    * CE    Cypriot (part not stated)\\n    * CF    Greek\\n    * CG    Greek Cypriot\\n    * CH    Turkish\\n    * CJ    Turkish Cypriot\\n    * CK    Italian\\n    * CL    Irish Traveller\\n    * CM    Traveller\\n    * CN    Gypsy/Romany\\n    * CP    Polish\\n    * CQ    All republics which made up the former USSR\\n    * CR    Kosovan\\n    * CS    Albanian\\n    * CT    Bosnian\\n    * CU    Croatian\\n    * CV    Serbian\\n    * CW    Other republics which made up the former Yugoslavia\\n    * CX    Mixed white\\n    * CY    Other white European, European unspecified, European mixed\\n    * D     White and Black Caribbean\\n    * E     White and Black African\\n    * F     White and Asian\\n    * G     Any other mixed background\\n    * GA    Black and Asian\\n    * GB    Black and Chinese\\n    * GC    Black and White\\n    * GD    Chinese and White\\n    * GE    Asian and Chinese\\n    * GF    Other Mixed, Mixed Unspecified\\n    * H     Indian or British Indian\\n    * J     Pakistani or British Pakistani\\n    * K     ","Bangladeshi or British Bangladeshi\\n    * L     Any other Asian background\\n    * LA    Mixed Asian\\n    * LB    Punjabi\\n    * LC    Kashmiri\\n    * LD    East African Asian\\n    * LE    Sri Lanka\\n    * LF    Tamil\\n    * LG    Sinhalese\\n    * LH    British Asian\\n    * LJ    Caribbean Asian\\n    * LK    Other Asian, Asian unspecified\\n    * M     Caribbean\\n    * N     African\\n    * P     Any other Black background\\n    * PA    Somali\\n    * PB    Mixed Black\\n    * PC    Nigerian\\n    * PD    Black British\\n    * PE    Other Black, Black unspecified\\n    * R     Chinese\\n    * S     Any other ethnic group\\n    * S1    Ashkenazi\\n    * S2    Sephardi\\n    * SA    Vietnamese\\n    * SB    Japanese\\n    * SC    Filipino\\n    * SD    Malaysian\\n    * SE    Any Other Group\\n    * Z     Not stated\",\"symbols\":[\"A\",\"B\",\"C\",\"C2\",\"C3\",\"CA\",\"CB\",\"CC\",\"CD\",\"CE\",\"CF\",\"CG\",\"CH\",\"CJ\",\"CK\",\"CL\",\"CM\",\"CN\",\"CP\",\"CQ\",\"CR\",\"CS\",\"CT\",\"CU\",\"CV\",\"CW\",\"CX\",\"CY\",\"D\",\"E\",\"F\",\"G\",\"GA\",\"GB\",\"GC\",\"GD\",\"GE\",\"GF\",\"H\",\"J\",\"K\",\"L\",\"LA\",\"LB\",\"LC\",\"LD\",\"LE\",\"LF\",\"LG\",\"LH\",\"LJ\",\"LK\",\"M\",\"N\",\"P\",\"PA\",\"PB\",\"PC\",\"PD\",\"PE\",\"R\",\"S\",\"S1\",\"S2\",\"SA\",\"SB\",\"SC\",\"SD\",\"SE\",\"Z\"]}}],\"doc\":\"ClinicalEthnicity as defined for GMS\"},{\"name\":\"previousTreatment\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"PreviousTreatment\",\"doc\":\"In GMS, Previous Treatment of Patient\",\"fields\":[{\"name\":\"previousTreatmentType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"previousTreatmentName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"previousTreatmentDate\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Date\",\"doc\":\"This defines a date record\",\"fields\":[{\"name\":\"year\",\"type\":\"int\",\"doc\":\"Format YYYY\"},{\"name\":\"month\",\"type\":[\"null\",\"int\"],\"doc\":\"Format MM. e.g June is 06\"},{\"name\":\"day\",\"type\":[\"null\",\"int\"],\"doc\":\"Format DD e.g. 12th of October is 12\"}]}]}]}}],\"doc\":\"For GMS cases, previous Treatment History\"}]}},\"doc\":\"List of members of a pedigree\"},{\"name\":\"analysisPanels\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AnalysisPanel\",\"doc\":\"An analysis panel\",\"fields\":[{\"name\":\"specificDisease\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The specific disease that a panel tests\"},{\"name\":\"panelName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The name of the panel\"},{\"name\":\"panelId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Id of the panel\"},{\"name\":\"panelVersion\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The version of the panel\"},{\"name\":\"reviewOutcome\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Deprecated\"},{\"name\":\"multipleGeneticOrigins\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Deprecated\"}]}}],\"doc\":\"List of panels\"},{\"name\":\"diseasePenetrances\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DiseasePenetrance\",\"doc\":\"A disease penetrance definition\",\"fields\":[{\"name\":\"specificDisease\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The disease to which the penetrance applies\"},{\"name\":\"penetrance\",\"type\":{\"type\":\"enum\",\"name\":\"Penetrance\",\"doc\":\"Penetrance assumed in the analysis\",\"symbols\":[\"complete\",\"incomplete\"]},\"doc\":\"The penetrance\"}]}}],\"doc\":\"List of disease penetrances. Moved to referralTest for GMS\"},{\"name\":\"readyForAnalysis\",\"type\":\"boolean\",\"doc\":\"Flag indicating if a pedigree is ready for analysis\"},{\"name\":\"familyQCState\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"FamilyQCState\",\"doc\":\"FamilyQCState\",\"symbols\":[\"noState\",\"passedMedicalReviewReadyForInterpretation\",\"passedMedicalReviewNotReadyForInterpretation\",\"queryToGel\",\"queryToGMC\",\"failed\"]}],\"doc\":\"The family quality control status\"}]}],\"doc\":\"Pedigree of the family.\"},{\"name\":\"otherFamilyHistory\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"OtherFamilyHistory\",\"doc\":\"Family history for secondary findings.\\n    Arrays of strings describing discrete family history phenotypes.\\n    Usually: `EndocrineTumours`, `colorectal`, `BreastOvarian` and `HDOrStroke` but can be others\",\"fields\":[{\"name\":\"maternalFamilyHistory\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Relevant Maternal family history\"},{\"name\":\"paternalFamilyHistory\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Relevant Maternal family history\"}]}],\"doc\":\"It is paternal or maternal with reference to the participant.\"},{\"name\":\"genePanelsCoverage\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"map\",\"values\":{\"type\":\"map\",\"values\":\"float\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"This map of key: panel_name, value: (map of key: gene, value: (map of metrics of key: metric name, value: float))\\n        That is: a map of tables of genes and metrics\"},{\"name\":\"interpretationFlags\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"InterpretationFlag\",\"doc\":\"A given interpretation flag together with an optional description\",\"fields\":[{\"name\":\"interpretationFlag\",\"type\":{\"type\":\"enum\",\"name\":\"InterpretationFlags\",\"doc\":\"Some flags relevant to the interpretation of a case\",\"symbols\":[\"mixed_chemistries\",\"mixedLab_preparation\",\"low_tumour_purity\",\"uniparental_isodisomy\",\"uniparental_heterodisomy\",\"unusual_karyotype\",\"high_cnv_count\",\"high_estimate_human_contamination_fraction\",\"mixed_recruiting_gmc\",\"suspected_mosaicism\",\"low_quality_sample\",\"ffpe_tumour_sample\",\"ff_nano_tumour_sample\",\"missing_values_for_proband_in_reported_variant\",\"reissued\",\"supplementary_report_errors\",\"internal_use_only\",\"high_priority\",\"suspected_increased_number_of_false_positive_heterozygous_loss_calls\",\"suspected_poor_quality_cnv_calls\",\"cnv_calls_assumed_xx_karyo\",\"cnv_calls_assumed_xy_karyo\",\"other\"]},\"doc\":\"The interpretation flag\"},{\"name\":\"additionalDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The description for the flag\"}]}}],\"doc\":\"Flags for this case relevant for interpretation\"},{\"name\":\"participantsInterpretationFlags\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ParticipantInterpretationFlags\",\"doc\":\"Interpretation flags at the participant level\",\"fields\":[{\"name\":\"interpretationFlag\",\"type\":\"InterpretationFlags\",\"doc\":\"The interpretation flag\"},{\"name\":\"additionalDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The description for the flag\"},{\"name\":\"SampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Sample Id fron which this flag was reported\"},{\"name\":\"participantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Participant Id as appeared in the pedigree\"}]}}],\"doc\":\"Flags for this case relevant for interpretation per participant\"},{\"name\":\"additionalInfo\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"Additional information\"}]}}]}},{\"name\":\"interpreted_genome\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DummyInterpretedGenome\",\"fields\":[{\"name\":\"status\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"DummyStatus\"}]},{\"name\":\"cva_variants_status\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"created_at\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"interpreted_genome_data\",\"type\":{\"type\":\"record\",\"name\":\"InterpretedGenome\",\"namespace\":\"org.gel.models.report.avro\",\"doc\":\"A interpreted genome for the rare disease program. This holds the list of candidate variants reported by an\\n    interpretation service together with all the relevant information that identify the case and how these conclusions were reached.\",\"fields\":[{\"name\":\"versionControl\",\"type\":\"ReportVersionControl\",\"doc\":\"Model version number\"},{\"name\":\"interpretationRequestId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Identifier for this interpretation request\"},{\"name\":\"interpretationRequestVersion\",\"type\":\"int\",\"doc\":\"Version for this interpretation request\"},{\"name\":\"interpretationService\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Name of the inter","pretation service\"},{\"name\":\"reportUrl\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"URL where the results can be accessed in the company's web interface\"},{\"name\":\"variants\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"SmallVariant\",\"doc\":\"A reported variant\",\"fields\":[{\"name\":\"variantCoordinates\",\"type\":\"VariantCoordinates\",\"doc\":\"The variant coordinates. Chromosome is either 1-22, X, Y, MT or any other contif in the reference genome,\\n        no \\\"chr\\\" prefix is expected. Position is 1- based. Reference and alternate should never be empty or any\\n        character representing emptiness (e.g.: . or -), a VCF-like indel representation is expected.\"},{\"name\":\"variantCalls\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"VariantCall\",\"doc\":\"This is intended to hold the genotypes for the family members. This assumes that varinats have been split before.\\n    In principle it is a phased zygosity as in VCF spec and called by the analysis provider if further phasing is conducted\",\"fields\":[{\"name\":\"participantId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Participant id\"},{\"name\":\"sampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Sample Id\"},{\"name\":\"zygosity\",\"type\":{\"type\":\"enum\",\"name\":\"Zygosity\",\"doc\":\"It is a representation of the zygosity\\n\\n* `reference_homozygous`: 0/0, 0|0\\n* `heterozygous`: 0/1, 1/0, 1|0, 0|1\\n* `alternate_homozygous`: 1/1, 1|1\\n* `missing`: ./., .|.\\n* `half_missing_reference`: ./0, 0/., 0|., .|0\\n* `half_missing_alternate`: ./1, 1/., 1|., .|1\\n* `alternate_hemizigous`: 1\\n* `reference_hemizigous`: 0\\n* `unk`: Anything unexpected\",\"symbols\":[\"reference_homozygous\",\"heterozygous\",\"alternate_homozygous\",\"missing\",\"half_missing_reference\",\"half_missing_alternate\",\"alternate_hemizigous\",\"reference_hemizigous\",\"unk\",\"na\"]},\"doc\":\"Zygosity. For somatic variants, or variants without zygosity use `na`\"},{\"name\":\"phaseGenotype\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"PhaseGenotype\",\"fields\":[{\"name\":\"sortedAlleles\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}},{\"name\":\"phaseSet\",\"type\":\"int\"}]}],\"doc\":\"phase alleles for those in phase\"},{\"name\":\"sampleVariantAlleleFrequency\",\"type\":[\"null\",\"double\"],\"doc\":\"Sample Variant Allele Frequency\"},{\"name\":\"depthReference\",\"type\":[\"null\",\"int\"],\"doc\":\"Depth for Reference Allele\"},{\"name\":\"depthAlternate\",\"type\":[\"null\",\"int\"],\"doc\":\"Depth for Alternate Allele\"},{\"name\":\"numberOfCopies\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"NumberOfCopies\",\"fields\":[{\"name\":\"numberOfCopies\",\"type\":\"int\",\"doc\":\"Number of copies given by the caller in one of the allele\"},{\"name\":\"confidenceIntervalMaximum\",\"type\":[\"null\",\"int\"]},{\"name\":\"confidenceIntervalMinimum\",\"type\":[\"null\",\"int\"]}]}}],\"doc\":\"Alleles for copy number variation - add doc\"},{\"name\":\"alleleOrigins\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"enum\",\"name\":\"AlleleOrigin\",\"doc\":\"Allele origin.\\n\\n* `SO_0001781`: de novo variant. http://purl.obolibrary.org/obo/SO_0001781\\n* `SO_0001778`: germline variant. http://purl.obolibrary.org/obo/SO_0001778\\n* `SO_0001775`: maternal variant. http://purl.obolibrary.org/obo/SO_0001775\\n* `SO_0001776`: paternal variant. http://purl.obolibrary.org/obo/SO_0001776\\n* `SO_0001779`: pedigree specific variant. http://purl.obolibrary.org/obo/SO_0001779\\n* `SO_0001780`: population specific variant. http://purl.obolibrary.org/obo/SO_0001780\\n* `SO_0001777`: somatic variant. http://purl.obolibrary.org/obo/SO_0001777\",\"symbols\":[\"de_novo_variant\",\"germline_variant\",\"maternal_variant\",\"paternal_variant\",\"pedigree_specific_variant\",\"population_specific_variant\",\"somatic_variant\"]}}],\"doc\":\"Describe whether this is a somatic or Germline variant\"},{\"name\":\"supportingReadTypes\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"enum\",\"name\":\"SupportingReadType\",\"symbols\":[\"spanning\",\"flanking\",\"inrepeat\"]}}]}]}},\"doc\":\"List of variant calls across all samples under analysis for this variant\"},{\"name\":\"reportEvents\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ReportEvent\",\"doc\":\"A report event holds all the information about why a given variant is relevant to report. The same variant may have\\n    several report events. For instance, we may have two report events from the tiering process when two panels are\\n    analysed, a positive report from a Genomic Medicine Centre (GMC) will correspond to an additional report event.\",\"fields\":[{\"name\":\"reportEventId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Unique identifier for each report event, this is unique across the whole report. A report having more than one\\n        report event with the same identifier is invalid. Repeating report event identifiers between different reports\\n        is valid. The uniqueness of this field will be checked in report validation\"},{\"name\":\"phenotypes\",\"type\":{\"type\":\"record\",\"name\":\"Phenotypes\",\"doc\":\"Oontology term based on the OBO format (see an example here http://snapshot.geneontology.org/ontology/go-basic.obo)\",\"fields\":[{\"name\":\"nonStandardPhenotype\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"The non standardised phenotypes (i.e.: may be free text)\"},{\"name\":\"standardPhenotypes\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"StandardPhenotype\",\"doc\":\"Standard phenotype term based on the OBO format (see an example here http://snapshot.geneontology.org/ontology/go-basic.obo)\",\"fields\":[{\"name\":\"id\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"name\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"namespace\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"definition\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"comment\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"alternativeIds\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"synonyms\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"isA\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"ontology\",\"type\":{\"type\":\"record\",\"name\":\"Ontology\",\"doc\":\"The ontology to which a standard term belongs\",\"fields\":[{\"name\":\"name\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"version\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]},\"doc\":\"The ontology (e.g.: HPO, OMIM, SNOMED CT)\"},{\"name\":\"matchScore\",\"type\":[\"null\",\"float\"],\"doc\":\"The match between the non standard phenotype and this term when in silico\"}]}}],\"doc\":\"The standardised phenotypes (i.e.: controlled terminology)\"}]},\"doc\":\"The list of phenotypes\"},{\"name\":\"variantConsequences\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"VariantConsequence\",\"doc\":\"A variant consequence as defined by the Sequence Ontology (SO) (e.g.: id = SO:0001816 ; name = non synonymous)\\n    NOTE: this record is equivalent to OpenCB's `ConsequenceType`, but we want to avoid naming collisions\",\"fields\":[{\"name\":\"id\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The SO term identifier (e.g.: SO:0001816)\"},{\"name\":\"name\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The SO term name (e.g.: non synonymous)\"}]}},\"doc\":\"Sequence Ontology terms for relevant consequence types for this report event\"},{\"name\":\"genePanel\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"GenePanel\",\"doc\":\"A panel of genes\",\"fields\":[{\"name\":\"panelIdentifier\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Panel name used\"},{\"name\":\"panelName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Panel name used\"},{\"name\":\"panelVersion\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Panel version\"},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"source i.e, PanelApp\"}]}],\"doc\":\"The panel of genes to which this report corresponds\"},{\"name\":\"modeOfInheritance\",\"type\":{\"type\":\"enum\",\"name\":\"ModeOfInheritance\",\"doc\":\"An enumeration for the different mode of inheritances:\\n\\n* `monoallelic_not_imprinted`: MONOALLELIC",", autosomal or pseudoautosomal, not imprinted\\n* `monoallelic_maternally_imprinted`: MONOALLELIC, autosomal or pseudoautosomal, maternally imprinted (paternal allele expressed)\\n* `monoallelic_paternally_imprinted`: MONOALLELIC, autosomal or pseudoautosomal, paternally imprinted (maternal allele expressed)\\n* `monoallelic`: MONOALLELIC, autosomal or pseudoautosomal, imprinted status unknown\\n* `biallelic`: BIALLELIC, autosomal or pseudoautosomal\\n* `monoallelic_and_biallelic`: BOTH monoallelic and biallelic, autosomal or pseudoautosomal\\n* `monoallelic_and_more_severe_biallelic`: BOTH monoallelic and biallelic, autosomal or pseudoautosomal (but BIALLELIC mutations cause a more SEVERE disease form), autosomal or pseudoautosomal\\n* `xlinked_biallelic`: X-LINKED: hemizygous mutation in males, biallelic mutations in females\\n* `xlinked_monoallelic`: X linked: hemizygous mutation in males, monoallelic mutations in females may cause disease (may be less severe, later onset than males)\\n* `mitochondrial`: MITOCHONDRIAL\\n* `unknown`: Unknown\",\"symbols\":[\"monoallelic\",\"monoallelic_not_imprinted\",\"monoallelic_maternally_imprinted\",\"monoallelic_paternally_imprinted\",\"biallelic\",\"monoallelic_and_biallelic\",\"monoallelic_and_more_severe_biallelic\",\"xlinked_biallelic\",\"xlinked_monoallelic\",\"mitochondrial\",\"unknown\",\"na\"]},\"doc\":\"Mode of inheritance used to analyse the family\"},{\"name\":\"genomicEntities\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"GenomicEntity\",\"doc\":\"A genomic feature\",\"fields\":[{\"name\":\"type\",\"type\":{\"type\":\"enum\",\"name\":\"GenomicEntityType\",\"doc\":\"Types of genomic features:\\n\\n* `regulatory_region`: a regulatory region\\n* `gene`: a gene\\n* `transcript`: a transcript\\n* `intergenic`: an intergenic region\",\"symbols\":[\"regulatory_region\",\"gene\",\"transcript\",\"intergenic\",\"gene_fusion\",\"genomic_region\",\"cytobands\"]},\"doc\":\"The type of the genomic entity\"},{\"name\":\"ensemblId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Ensembl identifier for the feature (e.g, ENST00000544455)\"},{\"name\":\"geneSymbol\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The HGNC gene symbol. This field is optional, BUT it should be filled if possible\"},{\"name\":\"otherIds\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Identifier\",\"fields\":[{\"name\":\"source\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Source i.e, esenmbl\"},{\"name\":\"identifier\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"identifier\"}]}}],\"doc\":\"Others identifiers for this genomic feature\"}]}},\"doc\":\"The list of genomic features of interest for this report event. Please note that one variant can overlap more\\n        that one gene/transcript. If more than one gene/transcript is considered interesting for this particular\\n        variant, should be reported in two different ReportEvents\"},{\"name\":\"segregationPattern\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"SegregationPattern\",\"symbols\":[\"UniparentalIsodisomy\",\"SimpleRecessive\",\"CompoundHeterozygous\",\"deNovo\",\"InheritedAutosomalDominant\",\"InheritedAutosomalDominantMaternallyImprinted\",\"InheritedAutosomalDominantPaternallyImprinted\",\"XLinkedCompoundHeterozygous\",\"XLinkedSimpleRecessive\",\"XLinkedMonoallelic\",\"MitochondrialGenome\"]}],\"doc\":\"Segregation pattern if any calculated using the genotypes information of a family\"},{\"name\":\"penetrance\",\"type\":[\"null\",\"org.gel.models.participant.avro.Penetrance\"],\"doc\":\"This is the penetrance assumed for scoring or classifying this variant\"},{\"name\":\"deNovoQualityScore\",\"type\":[\"null\",\"float\"],\"doc\":\"Likelihood of being a de novo variant\"},{\"name\":\"fullyExplainsPhenotype\",\"type\":[\"null\",\"boolean\"],\"doc\":\"Flag to indicate if this variant using this mode of inheritance can fully explain the phenotype\"},{\"name\":\"groupOfVariants\",\"type\":[\"null\",\"int\"],\"doc\":\"This value groups variants that together could explain the phenotype according to the mode of inheritance used.\\n        (e.g.: compound heterozygous). All the variants in the same report sharing the same value will be considered in\\n        the same group (i.e.: reported together). This value is an integer unique in the whole report.\\n        These values are only relevant within the same report.\"},{\"name\":\"eventJustification\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"This is the description of why this variant would be reported, for example that it affects the protein in this way\\n        and that this gene has been implicated in this disorder in these publications. Publications should be provided as PMIDs\\n        using the format [PMID:8075643]. Other sources can be used in the same manner, e.g. [OMIM:163500]. Brackets need to be included.\"},{\"name\":\"roleInCancer\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"enum\",\"name\":\"RoleInCancer\",\"doc\":\"The role of a given genomic feature in cancer\\n\\n* `NCIT_C16936`: oncogene. A gene that is a mutated (changed) form of a gene involved in normal cell growth. Oncogenes may cause the growth of cancer cells. Mutations in genes that become oncogenes can be inherited or caused by being exposed to substances in the environment that cause cancer. http://purl.obolibrary.org/obo/NCIT_C16936\\n* `NCIT_C17362`: tumor_suppressor_gene. A type of gene that makes a protein called a tumor suppressor protein that helps control cell growth. Mutations (changes in DNA) in antioncogenes may lead to cancer. http://purl.obolibrary.org/obo/NCIT_C17362\",\"symbols\":[\"oncogene\",\"tumor_suppressor_gene\",\"both\"]}}],\"doc\":\"The role of this variant in cancer if any\"},{\"name\":\"actions\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Actions\",\"doc\":\"Clinical actions\",\"fields\":[{\"name\":\"trials\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Trial\",\"fields\":[{\"name\":\"studyUrl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"URL where reference information for this trail can be found\"},{\"name\":\"studyIdentifier\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Trail/Study indetifier\"},{\"name\":\"startDate\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Start date of the study\"},{\"name\":\"estimateCompletionDate\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Completion date of the study\"},{\"name\":\"title\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Title of the study\"},{\"name\":\"phase\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"StudyPhase\",\"doc\":\"N/A: Trials without phases (for example, studies of devices or behavioural interventions).\\n    Early Phase 1 (Formerly listed as \\\"Phase 0\\\"): Exploratory trials, involving very limited human exposure, with no therapeutic or diagnostic intent (e.g., screening studies, microdose studies). See FDA guidance on exploratory IND studies for more information.\\n    Phase 1: Includes initial studies to determine the metabolism and pharmacologic actions of drugs in humans, the side effects associated with increasing doses, and to gain early evidence of effectiveness; may include healthy participants and/or patients.\\n    Phase 1/Phase 2: Trials that are a combination of phases 1 and 2.\\n    Phase 2: Includes controlled clinical studies conducted to evaluate the effectiveness of the drug for a particular indication or indications in participants with the disease or condition under study and to determine the common short-term side effects and risks.\\n    Phase 2/Phase 3: Trials that are a combination of phases 2 and 3.\\n    Phase 3: Includes trials conducted after preliminary evidence suggesting effectiveness of the drug has been obtained, and are intended to gather additional information to evaluate the overall benefit-risk relationship of the drug.\\n    Phase 4: Studies of FDA-approved drugs to delineate additional information including the drug's risks, benefits, and optimal use.\",\"symbols\":[\"na\",\"early_phase1\",\"phase1\",\"phase1_phase2\",\"phase2\",\"phase2_phase3\",\"phase3\",\"phase4\"]}],\"doc\":\"Study Phase\"},{\"name\":\"interventions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Intervention\",\"doc\":\"A process or action that is the focus of a clinical study.\\n    Ref. https://prsinfo.clinicaltrials.g","ov/definitions.html\",\"fields\":[{\"name\":\"interventionType\",\"type\":{\"type\":\"enum\",\"name\":\"InterventionType\",\"doc\":\"For each intervention studied in the clinical study, the general type of intervention\\n\\n* `drug`: Including placebo\\n* `device`: Including sham\\n* `biological`: Vaccine\\n* `procedure`: Surgery\\n* `radiation`\\n* `behavioral`: For example, psychotherapy, lifestyle counselling\\n* `genetic`: Including gene transfer, stem cell and recombinant DNA\\n* `dietary_supplement`: For example, vitamins, minerals\\n* `combination_product`: Combining a drug and device, a biological product and device; a drug and biological product; or a drug, biological product, and device\\n* `diagnostic_test`: For example, imaging, in-vitro\\n* `other`\\n\\n    Ref. https://prsinfo.clinicaltrials.gov/definitions.htm\",\"symbols\":[\"drug\",\"device\",\"procedure\",\"biological\",\"radiation\",\"behavioral\",\"genetic\",\"dietary_supplement\",\"combination_product\",\"diagnostic_test\",\"other\"]},\"doc\":\"Intervention type, i.e drug\"},{\"name\":\"interventionName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Intervention name: Placebo\"}]}}],\"doc\":\"Interventions\"},{\"name\":\"conditions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Conditions\"},{\"name\":\"primaryPurpose\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"PrimaryPurpose\",\"doc\":\"Treatment: One or more interventions are being evaluated for treating a disease, syndrome, or condition.\\n    Prevention: One or more interventions are being assessed for preventing the development of a specific disease or health condition.\\n    Diagnostic: One or more interventions are being evaluated for identifying a disease or health condition.\\n    Supportive Care: One or more interventions are evaluated for maximizing comfort, minimizing side effects, or mitigating against a decline in the participant's health or function.\\n    Screening: One or more interventions are assessed or examined for identifying a condition, or risk factors for a condition, in people who are not yet known to have the condition or risk factor.\\n    Health Services Research: One or more interventions for evaluating the delivery, processes, management, organization, or financing of healthcare.\\n    Basic Science: One or more interventions for examining the basic mechanism of action (for example, physiology or biomechanics of an intervention).\\n    Device Feasibility: An intervention of a device product is being evaluated in a small clinical trial (generally fewer than 10 participants) to determine the feasibility of the product; or a clinical trial to test a prototype device for feasibility and not health outcomes. Such studies are conducted to confirm the design and operating specifications of a device before beginning a full clinical trial.\\n    Other: None of the other options applies.\\n\\n    Ref. https://prsinfo.clinicaltrials.gov/definitions.htm\",\"symbols\":[\"treatment\",\"prevention\",\"diagnostic\",\"supportive_care\",\"screening\",\"health_services_research\",\"basic_science\",\"device_feasibility\",\"other\"]}],\"doc\":\"Primary Purpose of the study\"},{\"name\":\"studyType\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"StudyType\",\"doc\":\"* `Interventional (clinical trial)`: Participants are assigned prospectively to an intervention or interventions\\naccording to a protocol to evaluate the effect of the intervention(s) on biomedical or other health related outcomes.\\n* `Observational`: Studies in human beings in which biomedical and/or health outcomes are assessed in pre-defined groups\\nof individuals. Participants in the study may receive diagnostic, therapeutic, or other interventions, but the\\ninvestigator does not assign specific interventions to the study participants. This includes when participants\\nreceive interventions as part of routine medical care, and a researcher studies the effect of the intervention.\\n* `Expanded Access`: An investigational drug product (including biological product)\\navailable through expanded access for patients who do not qualify for enrollment in a clinical trial.\\nExpanded Access includes all expanded access types under section 561 of the Federal Food, Drug, and\\nCosmetic Act: (1) for individual patients, including emergency use; (2) for intermediate-size patient populations;\\nand (3) under a treatment IND or treatment protocol. (For more information on data requirements for this Study Type,\\nsee Expanded Access Data Element Definitions).\",\"symbols\":[\"interventional\",\"observational\",\"patient_registry\",\"expanded_access\"]}],\"doc\":\"Study Type\"},{\"name\":\"demogrphicElegibilityCriteria\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"DemographicElegibilityCriteria\",\"fields\":[{\"name\":\"sex\",\"type\":\"org.gel.models.participant.avro.Sex\"},{\"name\":\"ageRange\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"AgeRange\",\"fields\":[{\"name\":\"minimumAge\",\"type\":\"int\"},{\"name\":\"maximumAge\",\"type\":\"int\"},{\"name\":\"timeunit\",\"type\":{\"type\":\"enum\",\"name\":\"TimeUnit\",\"symbols\":[\"years\",\"months\",\"weeks\",\"days\",\"hours\",\"minutes\",\"na\"]}}]}]}]}],\"doc\":\"Elegigility Criteria based on Age and Sex\"},{\"name\":\"locations\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"TrialLocation\",\"fields\":[{\"name\":\"name\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"city\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"country\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"zip\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}],\"doc\":\"List with all of the locations where participant can enrolle\"},{\"name\":\"variantActionable\",\"type\":\"boolean\",\"doc\":\"If true, the association was made using the variant information,\\n        if not the association was made at Genomic Entity level\"}]}}]},{\"name\":\"prognosis\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Prognosis\",\"fields\":[{\"name\":\"referenceUrl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"URL where reference information for this prognosis can be found\"},{\"name\":\"prognosis\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"PrognosisClassification\",\"symbols\":[\"altered_prognosis\",\"favourable_prognosis\",\"unfavourable_prognosis\"]}],\"doc\":\"Prognosis classification (defined as favourable or unfavourable),\\n        in the case that the direction of the prognosis is not known altered_prognosis should be used\"},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Source if known\"},{\"name\":\"references\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"References\"},{\"name\":\"conditions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Conditions\"},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Full description of the associated prognosis\"},{\"name\":\"variantActionable\",\"type\":\"boolean\",\"doc\":\"If true, the association was made at the variant level, if not the association was made at Genomic Entity level\"}]}}]},{\"name\":\"therapies\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Therapy\",\"fields\":[{\"name\":\"referenceUrl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"URL where reference information for this therapy association can be found\"},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Source\"},{\"name\":\"references\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"References\"},{\"name\":\"conditions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Conditions\"},{\"name\":\"drugResponse\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DrugResponse\",\"fields\":[{\"name\":\"TreatmentAgent\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Treatment agent\"},{\"name\":\"drugResponseClassification\",\"type\":{\"type\":\"enum\",\"name\":\"DrugResponseClassification\",\"symbols\":[\"altered_sensitivity\",\"reduced_sensitivity\",\"increased_sensitivity\",\"altered_resistance\",\"increased_resistance\",\"reduced_resistance\",\"increased_risk_of_toxicity\",\"reduced_risk_of_toxicity\",\"altered_toxicity\",\"adverse_drug_reaction\",\"indication","\",\"contraindication\",\"dosing_alteration\",\"increased_dose\",\"reduced_dose\",\"increased_monitoring\",\"increased_efficacy\",\"reduced_efficacy\",\"altered_efficacy\"]},\"doc\":\"associated effect of the drug\"}]}}],\"doc\":\"Drug responses\"},{\"name\":\"otherInterventions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"Intervention\"}],\"doc\":\"Any other clinical intervention\"},{\"name\":\"variantActionable\",\"type\":\"boolean\",\"doc\":\"If true, the association was made at the variant level, if not the association was made at Genomic Entity level\"}]}}]},{\"name\":\"diagnostic\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Diagnostic\",\"fields\":[{\"name\":\"referenceUrl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"URL where reference information for this prognosis can be found\"},{\"name\":\"sources\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Sources if known\"},{\"name\":\"references\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"References\"},{\"name\":\"biomarkers\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Biomarkers\"},{\"name\":\"conditions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Associated conditions\"},{\"name\":\"diagnosis\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis\"},{\"name\":\"diagnosisStatus\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis status\"},{\"name\":\"otherCondition\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Other condition\"},{\"name\":\"variantActionable\",\"type\":\"boolean\",\"doc\":\"If true, the association was made at the variant level, if not the association was made at Genomic Entity level\"}]}}]},{\"name\":\"otherAction\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"OtherAction\",\"fields\":[{\"name\":\"referenceUrl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"URL where reference information for this action can be found\"},{\"name\":\"identifier\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Action identifier\"},{\"name\":\"sources\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Sources if known\"},{\"name\":\"actionType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Action type\"},{\"name\":\"conditions\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Associated conditions\"},{\"name\":\"otherAttributes\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"Other attributes\"},{\"name\":\"variantActionable\",\"type\":\"boolean\",\"doc\":\"If true, the association was made at the variant level, if not the association was made at Genomic Entity level\"}]}}]}]}],\"doc\":\"Actions can be taken on the variant if any\"},{\"name\":\"score\",\"type\":[\"null\",\"float\"],\"doc\":\"This is the score provided to reflect a variant's likelihood of explaining the phenotype using a specific\\n        mode of Inheritance. This may be the result of different scoring systems\"},{\"name\":\"vendorSpecificScores\",\"type\":[\"null\",{\"type\":\"map\",\"values\":\"float\",\"avro.java.string\":\"String\"}],\"doc\":\"Other scores that the interpretation provider may add (for example phenotypically informed or family\\n        informed scores)\"},{\"name\":\"variantClassification\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VariantClassification\",\"doc\":\"The variant classification according to different properties.\",\"fields\":[{\"name\":\"clinicalSignificance\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"ClinicalSignificance\",\"symbols\":[\"benign\",\"likely_benign\",\"likely_pathogenic\",\"pathogenic\",\"uncertain_significance\",\"excluded\"]}],\"doc\":\"The variant's clinical significance.\"},{\"name\":\"drugResponseClassification\",\"type\":[\"null\",\"DrugResponseClassification\"],\"doc\":\"The variant's pharmacogenomics classification.\"},{\"name\":\"traitAssociation\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"TraitAssociation\",\"symbols\":[\"established_risk_allele\",\"likely_risk_allele\",\"uncertain_risk_allele\",\"protective\"]}],\"doc\":\"The variant's trait association.\"},{\"name\":\"tumorigenesisClassification\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"TumorigenesisClassification\",\"symbols\":[\"driver\",\"passenger\",\"modifier\"]}],\"doc\":\"The variant's tumorigenesis classification.\"},{\"name\":\"functionalEffect\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"VariantFunctionalEffect\",\"symbols\":[\"dominant_negative_variant\",\"gain_of_function_variant\",\"lethal_variant\",\"loss_of_function_variant\",\"loss_of_heterozygosity\",\"null_variant\"]}],\"doc\":\"The variant functional effect\"}]}],\"doc\":\"Variant classification\"},{\"name\":\"guidelineBasedVariantClassification\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"GuidelineBasedVariantClassification\",\"doc\":\"Variant classification based on guidlines, AMP and ACMG are supported\",\"fields\":[{\"name\":\"acmgVariantClassification\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"AcmgVariantClassification\",\"doc\":\"Full record for the ACMG variant clasiffication, including all selectedd evidences and the final classification.\",\"fields\":[{\"name\":\"acmgEvidences\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AcmgEvidence\",\"doc\":\"AcmgEvidence. This should be buit for each one of the evidences assing to a variants following the ACMG guidelines.\\n    An AcmgEvidence, should map with one of the criteria defined, i.e, PVS1, BA1, PM1...\",\"fields\":[{\"name\":\"category\",\"type\":{\"type\":\"enum\",\"name\":\"AcmgEvidenceCategory\",\"doc\":\"Each ACMG criterion is classified in one of these categories\",\"symbols\":[\"population_data\",\"computational_and_predictive_data\",\"functional_data\",\"segregation_data\",\"de_novo_data\",\"allelic_data\",\"other_database\",\"other_data\"]},\"doc\":\"Evidence category as defined in ACMG guidelines\"},{\"name\":\"type\",\"type\":{\"type\":\"enum\",\"name\":\"AcmgEvidenceType\",\"doc\":\"Each ACMG cirterion will be classifed as benign or pathogenic\",\"symbols\":[\"benign\",\"pathogenic\"]},\"doc\":\"Evidence type: benign or pathogenic\"},{\"name\":\"weight\",\"type\":{\"type\":\"enum\",\"name\":\"AcmgEvidenceWeight\",\"doc\":\"Each ACMG criterion is weighted using the following terms:\\n\\n* `stand_alone`: `A`, stand-alone applied for benign variant critieria `(BA1)`\\n* `supporting`: `P`, supporting applied for benign variant critieria `(BP1-6)` and pathogenic variant criteria `(PP1-5)`\\n* `moderate`: `M`, moderate applied for pathogenic variant critieria (PM1-6)\\n* `strong`: `S`, strong applied for pathogenic variant critieria (PS1-4)\\n* `very_strong`: `S`, Very Stong applied for pathogenic variant critieria (PVS1)\",\"symbols\":[\"stand_alone\",\"supporting\",\"moderate\",\"strong\",\"very_strong\"]},\"doc\":\"Default strength for criterion as defined in Table 3 of ACMG guidelines (Richards et al 2015). e.g. PM2 would be \\\"moderate\\\"\"},{\"name\":\"modifier\",\"type\":\"int\",\"doc\":\"The number suffix at the end of the ACMG criteria code e.g PM2 would be 2\"},{\"name\":\"activationStrength\",\"type\":{\"type\":\"enum\",\"name\":\"ActivationStrength\",\"doc\":\"Activation Strength enumeration:\\n* `strong`\\n* `moderate`\\n* `supporting`\\n* `very_strong`\\n* `stand_alone`\",\"symbols\":[\"strong\",\"moderate\",\"supporting\",\"very_strong\",\"stand_alone\"]},\"doc\":\"The strength this criterion has been used at in this interpretation. e.g. if PM2 was only used at \\\"supporting\\\" rather than \\\"moderate\\\", the activation strength would be \\\"supporting\\\"\"},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The description of the evidence as described in ACMG guidelines e.g. for PM2 the description would be \\\"Absent from controls (or at extremely low frequency if recessive) in Exome Sequencing Project, 1000 Genomes Project, or Exome Aggregation Consortium\\\"\"},{\"name\":\"comments\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"UserComment\",\"fields\":[{\"name\":\"comment\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Comment text\"},{\"name\":\"user\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"User\",\"fields\":[{\"name\":\"userid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Azure Active Directory immutable user OID\"},{\"name\":\"email\",\"type\":{\"type\":","\"string\",\"avro.java.string\":\"String\"},\"doc\":\"User email address\"},{\"name\":\"username\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Username\"},{\"name\":\"role\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"groups\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}]}]}],\"doc\":\"User who created comment\"},{\"name\":\"timestamp\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Date and time comment was created (ISO 8601 datetime with seconds and timezone e.g. 2020-11-23T15:52:36+00:00)\"}]}}],\"doc\":\"User comments attached to this ACMG criteria in this case\"}]}},\"doc\":\"Details of ACMG criteria used to score this variant\"},{\"name\":\"clinicalSignificance\",\"type\":\"ClinicalSignificance\",\"doc\":\"Final classification selected by user\"},{\"name\":\"calculatedClinicalSignificance\",\"type\":[\"null\",\"ClinicalSignificance\"],\"doc\":\"Classification computed from ACMG scores\"},{\"name\":\"assessment\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"noAcmgEvidence\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"AcmgEvidenceCategory\"}],\"doc\":\"ACMG evidence categories for which the user has indicated there is no evidence available\"}]}],\"doc\":\"Variant classification using ACMG framework\"},{\"name\":\"ampVariantClassification\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"AmpVariantClassification\",\"doc\":\"Full Variant classification acording to AMP guideline, including all supporting evidences and the final\\n    assessment\",\"fields\":[{\"name\":\"ampEvidences\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AmpEvidence\",\"doc\":\"Evidences as defined in AMP guidelines, they are composed by a evidence type (first column in the evidence table of\\n    the guidlines) and a assessment of the evicence, this last one will define the streght of the evidence, supporting\\n    the variant to be classified as TierI-IV\",\"fields\":[{\"name\":\"type\",\"type\":{\"type\":\"enum\",\"name\":\"AmpEvidenceType\",\"doc\":\"Type of evidence in the AMP guideline\",\"symbols\":[\"mutation_type\",\"therapies\",\"variant_frequencies\",\"potential_germline\",\"population_database_presence\",\"germline_database_presence\",\"somatic_database_presence\",\"impact_predictive_software\",\"pathway_involvement\",\"publications\"]},\"doc\":\"AMP evidence type according to Guidlines, i.e germline_database_presence\"},{\"name\":\"evidenceAssessment\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Assessment for AMP evidence, i.e Present in ClinVar\"}]}},\"doc\":\"List of AMP evidences\"},{\"name\":\"ampTier\",\"type\":{\"type\":\"enum\",\"name\":\"AmpTier\",\"doc\":\"AMP tier:\\n* `TierI`: Variants of Strong Clinical Significance\\n* `TierII`: Variants of Potential Clinical Significance\\n* `TierIII`: Variants of Unknown Clinical Significance\\n* `TierIV`: Benign or Likely Benign Variants\",\"symbols\":[\"tierI\",\"tierII\",\"tierIII\",\"tierIV\"]},\"doc\":\"Final Clasification taken in account the evidences\"},{\"name\":\"ampClincialOrExperimentalEvidence\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AmpClincialOrExperimentalEvidence\",\"doc\":\"Amp Clinical or Experimental Evidence, the level will define the overal clasification of the variant together with\\n    the tiering.\",\"fields\":[{\"name\":\"category\",\"type\":{\"type\":\"enum\",\"name\":\"AmpClinicalOrExperimentalEvidenceCategory\",\"doc\":\"Categories of Clinical and/or Experimental Evidence as defined in AMP guidelines\",\"symbols\":[\"therapeutic\",\"diagnosis\",\"prognosis\"]},\"doc\":\"As denined in AMP guidelines: therapeutic, diagnosis or prognosis\"},{\"name\":\"level\",\"type\":{\"type\":\"enum\",\"name\":\"AmpClinicalOrExperimentalEvidenceLevel\",\"doc\":\"Levels for categories of Clinical and/or Experimental Evidence as defined in AMP guidelines\",\"symbols\":[\"levelA\",\"levelB\",\"levelC\",\"levelD\"]},\"doc\":\"As denined in AMP guidelines: levelA, levelB, levelC, levelD\"},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Description of the evidence\"}]}}],\"doc\":\"Clinical or Experimental evicence\"},{\"name\":\"assessment\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Final Assessment\"}]}],\"doc\":\"Variant classification using AMP framework\"}]}],\"doc\":\"Guidelines based Variant classification\"},{\"name\":\"algorithmBasedVariantClassifications\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AlgorithmBasedVariantClassification\",\"fields\":[{\"name\":\"algorithmName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Name of the applied algorithm\"},{\"name\":\"classification\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"classification\"},{\"name\":\"rank\",\"type\":[\"null\",\"int\"],\"doc\":\"rank\"},{\"name\":\"score\",\"type\":[\"null\",\"int\"],\"doc\":\"Score\"}]}}],\"doc\":\"Algorithm based variant classifications\"},{\"name\":\"tier\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Tier\",\"doc\":\"Variant tiers as defined by Genomics England\",\"symbols\":[\"NONE\",\"TIER1\",\"TIER2\",\"TIER3\",\"TIER4\",\"TIER5\",\"TIERA\",\"TIERB\"]}],\"doc\":\"The tier where this variant has been classified. Tier is a property of the model of inheritance and therefore\\n        is subject to change depending on the inheritance assumptions\"},{\"name\":\"domain\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Domain\",\"symbols\":[\"DOMAIN1\",\"DOMAIN2\",\"DOMAIN3\",\"DOMAIN4\",\"NONE\"]}],\"doc\":\"The Domain where this variant has been classified.\"}]}},\"doc\":\"The list of report events for this variant across multiple modes of inheritance and panels\"},{\"name\":\"variantAttributes\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VariantAttributes\",\"doc\":\"Some additional variant attributes\",\"fields\":[{\"name\":\"genomicChanges\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"gDNA change, HGVS nomenclature (e.g.: g.476A>T)\"},{\"name\":\"cdnaChanges\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"cDNA change, HGVS nomenclature (e.g.: c.76A>T)\"},{\"name\":\"proteinChanges\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Protein change, HGVS nomenclature (e.g.: p.Lys76Asn)\"},{\"name\":\"additionalTextualVariantAnnotations\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"Any additional information in a free text field. For example a quote from a paper\"},{\"name\":\"references\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"Additional references for ths variant. For example HGMD ID or Pubmed Id\"},{\"name\":\"variantIdentifiers\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VariantIdentifiers\",\"fields\":[{\"name\":\"dbSnpId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Variant identifier in dbSNP\"},{\"name\":\"cosmicIds\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Variant identifier in Cosmic\"},{\"name\":\"clinVarIds\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Variant identifier in ClinVar\"},{\"name\":\"otherIds\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"Identifier\"}]}]}]},{\"name\":\"alleleFrequencies\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AlleleFrequency\",\"doc\":\"The population allele frequency of a given variant in a given study and optionally population\",\"fields\":[{\"name\":\"study\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The study from where this data comes from\"},{\"name\":\"population\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The specific population where this allele frequency belongs\"},{\"name\":\"alternateFrequency\",\"type\":\"float\",\"doc\":\"The frequency of the alternate allele\"}]}}],\"doc\":\"A list of population allele frequencies\"},{\"name\":\"additionalNumericVariantAnnotations\",\"type\":[\"null\",{\"type\":\"map\",\"values\":\"float\",\"avro.java.string\":\"String\"}],\"doc\":\"Additional numeric variant annotations for this variant. For Example (Allele Frequency, sift, polyphen,\\n        mutationTaster, CADD. ..)\"},{\"name\":\"comments\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Comments\"},{\"name\":\"alleleOrigins\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"Allele","Origin\"}],\"doc\":\"List of allele origins for this variant in this report\"},{\"name\":\"ihp\",\"type\":[\"null\",\"int\"],\"doc\":\"Largest reference interrupted homopolymer length intersecting with the indel\"},{\"name\":\"recurrentlyReported\",\"type\":[\"null\",\"boolean\"],\"doc\":\"Flag indicating if the variant is recurrently reported\"},{\"name\":\"fdp50\",\"type\":[\"null\",\"float\"],\"doc\":\"Average tier1 number of basecalls filtered from original read depth within 50 bases\"},{\"name\":\"others\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"Map of other attributes where keys are the attribute names and values are the attributes\"}]}]}]}}],\"doc\":\"List of small reported variants\"},{\"name\":\"structuralVariants\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"StructuralVariant\",\"fields\":[{\"name\":\"variantType\",\"type\":\"StructuralVariantType\",\"doc\":\"Structural Variant Type as would appear in VCF\"},{\"name\":\"coordinates\",\"type\":\"Coordinates\"},{\"name\":\"leftInsSeq\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"rightInsSeq\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"reportEvents\",\"type\":{\"type\":\"array\",\"items\":\"ReportEvent\"}},{\"name\":\"variantCalls\",\"type\":{\"type\":\"array\",\"items\":\"VariantCall\"},\"doc\":\"array of genotypes for the samples\"},{\"name\":\"variantAttributes\",\"type\":[\"null\",\"VariantAttributes\"]}]}}],\"doc\":\"List of simple structural reported variants (duplications, deletions, insertions, inversions, CNVs)\"},{\"name\":\"chromosomalRearrangements\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ChromosomalRearrangement\",\"fields\":[{\"name\":\"breakPoints\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"BreakPoint\",\"fields\":[{\"name\":\"coordinates\",\"type\":\"Coordinates\"},{\"name\":\"reference\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"alternate\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"info\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}]}]}}]},{\"name\":\"rearrangements\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Rearrangement\",\"fields\":[{\"name\":\"leftCoordinates\",\"type\":\"Coordinates\"},{\"name\":\"rightCoordinates\",\"type\":\"Coordinates\"},{\"name\":\"orientation\",\"type\":{\"type\":\"enum\",\"name\":\"Orientation\",\"symbols\":[\"start_start\",\"start_end\",\"end_end\"]}},{\"name\":\"leftInsSeq\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"rightInsSeq\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}},{\"name\":\"reportEvents\",\"type\":{\"type\":\"array\",\"items\":\"ReportEvent\"}},{\"name\":\"variantCalls\",\"type\":{\"type\":\"array\",\"items\":\"VariantCall\"},\"doc\":\"array of genotypes for the samples\"},{\"name\":\"variantAttributes\",\"type\":[\"null\",\"VariantAttributes\"]}]}}],\"doc\":\"List of complex structural reported variants (chomosomal rearrangement)\"},{\"name\":\"shortTandemRepeats\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ShortTandemRepeat\",\"fields\":[{\"name\":\"coordinates\",\"type\":\"Coordinates\"},{\"name\":\"reportEvents\",\"type\":{\"type\":\"array\",\"items\":\"ReportEvent\"}},{\"name\":\"variantCalls\",\"type\":{\"type\":\"array\",\"items\":\"VariantCall\"},\"doc\":\"array of genotypes for the samples\"},{\"name\":\"variantAttributes\",\"type\":[\"null\",\"VariantAttributes\"]},{\"name\":\"shortTandemRepeatReferenceData\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"ShortTandemRepeatReferenceData\",\"fields\":[{\"name\":\"repeatedSequence\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"pathogenic_number_of_repeats_threshold\",\"type\":\"int\"},{\"name\":\"normal_number_of_repeats_threshold\",\"type\":\"int\"}]}]}]}}],\"doc\":\"List of short tandem repeat variants\"},{\"name\":\"uniparentalDisomies\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"UniparentalDisomy\",\"fields\":[{\"name\":\"assembly\",\"type\":\"Assembly\",\"doc\":\"The assembly\"},{\"name\":\"chromosome\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Chromosome where two homologues were inherited from one parent\"},{\"name\":\"complete\",\"type\":[\"null\",\"boolean\"],\"doc\":\"indicates Whether the UPD event involves an entire chromosome or part of a chromosome\"},{\"name\":\"origin\",\"type\":{\"type\":\"enum\",\"name\":\"UniparentalDisomyOrigin\",\"symbols\":[\"paternal\",\"maternal\",\"unknown\"]},\"doc\":\"The parent who contributed two chromosomes was the mother (maternal) or the father (paternal)\"},{\"name\":\"uniparentalDisomyFragments\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"UniparentalDisomyFragment\",\"fields\":[{\"name\":\"coordinates\",\"type\":[\"null\",\"Coordinates\"],\"doc\":\"Coordinates can be specified to indicate the part of the chromosome affected\"},{\"name\":\"uniparentalDisomyType\",\"type\":{\"type\":\"enum\",\"name\":\"UniparentalDisomyType\",\"symbols\":[\"isodisomy\",\"heterodisomy\",\"both\"]},\"doc\":\"indicates whether the UPD event involves `isodisomy`, `heterodisomy` or `both`\"}]}}],\"doc\":\"List of all of the UPD fragments for this UPD event\"},{\"name\":\"participantId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Participant affected by this UPD\"},{\"name\":\"uniparentalDisomyEvidences\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"UniparentalDisomyEvidences\",\"fields\":[{\"name\":\"ibds\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"IdentityByDescent\",\"fields\":[{\"name\":\"relatedSample\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},{\"name\":\"ibd0\",\"type\":\"float\"},{\"name\":\"ibd1\",\"type\":\"float\"},{\"name\":\"ibd2\",\"type\":\"float\"},{\"name\":\"pihat\",\"type\":\"float\"}]}}]}]}],\"doc\":\"Evidences for the UPD call\"}]}}],\"doc\":\"List of uniparental disomies across all the individuals in this report\"},{\"name\":\"karyotypes\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Karyotype\",\"fields\":[{\"name\":\"iscn\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"International System for Human Cytogenetic Nomenclature (e.g.: \\\"46,XY\\\", \\\"46,XY,-5,-12,+t(5pl2p),+t(5ql2q)\\\")\"},{\"name\":\"description\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Full description of the karyotype\"},{\"name\":\"aneuploidies\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Aneuploidy\",\"fields\":[{\"name\":\"iscn\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"International System for Human Cytogenetic Nomenclature (.e.g: \\\"+14p+\\\", \\\"+t(13q21 q)\\\")\"},{\"name\":\"assembly\",\"type\":\"Assembly\",\"doc\":\"The assembly\"},{\"name\":\"chromosome\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Chromosome involved in the aneuploidy\"},{\"name\":\"complete\",\"type\":\"boolean\",\"doc\":\"Wheter the aneuploidy is for the whole chromosme or just a fragment\"},{\"name\":\"coordinates\",\"type\":[\"null\",\"Coordinates\"],\"doc\":\"Coordinates can be specified to indicate the part of the chromosome affected\"},{\"name\":\"numberOfCopies\",\"type\":\"int\",\"doc\":\"Number of copies\"}]}}],\"doc\":\"List of aneuploidies\"},{\"name\":\"numberOfChromosomes\",\"type\":\"int\",\"doc\":\"Total number of chromosomes\"},{\"name\":\"personKaryotipicSex\",\"type\":\"org.gel.models.participant.avro.PersonKaryotipicSex\",\"doc\":\"Kariotypic sex\"},{\"name\":\"participantId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Participant identifier\"}]}}],\"doc\":\"List of inferred karyotypes across all the individuals in this report\"},{\"name\":\"referenceDatabasesVersions\",\"type\":{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"},\"doc\":\"This map contains the versions of the different databases used in the process, being the database names the\\n        keys and the versions the values.\"},{\"name\":\"softwareVersions\",\"type\":{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"},\"doc\":\"This map contains the versions of the different software systems used in the process, being the software\\n        names the keys and the versions the values.\"},{\"name\":\"comments\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Comments about the report\"}]}}]}}}]}");
  public static org.apache.avro.Schema getClassSchema() { return SCHEMA$; }
   private java.util.List<DummyFile> files;
   private int number_of_samples;
   private String program;
   private String family_id;
   private java.util.List<String> labkey_links;
   private String created_at;
   private String last_modified;
   private String proband;
   private java.util.List<DummyStatus> status;
   private String case_priority;
   private String assembly;
   private String case_id;
   private int interpretation_request_id;
   private String cohort_id;
   private java.util.List<DummyClinicalReport> clinical_report;
   private DummyInterpretationRequest interpretation_request_data;
   private java.util.List<DummyInterpretedGenome> interpreted_genome;

  /**
   * Default constructor.  Note that this does not initialize fields
   * to their default values from the schema.  If that is desired then
   * one should use <code>newBuilder()</code>. 
   */
  public CipapiFacade() {}

  /**
   * All-args constructor.
   */
  public CipapiFacade(java.util.List<DummyFile> files, Integer number_of_samples, String program, String family_id, java.util.List<String> labkey_links, String created_at, String last_modified, String proband, java.util.List<DummyStatus> status, String case_priority, String assembly, String case_id, Integer interpretation_request_id, String cohort_id, java.util.List<DummyClinicalReport> clinical_report, DummyInterpretationRequest interpretation_request_data, java.util.List<DummyInterpretedGenome> interpreted_genome) {
    this.files = files;
    this.number_of_samples = number_of_samples;
    this.program = program;
    this.family_id = family_id;
    this.labkey_links = labkey_links;
    this.created_at = created_at;
    this.last_modified = last_modified;
    this.proband = proband;
    this.status = status;
    this.case_priority = case_priority;
    this.assembly = assembly;
    this.case_id = case_id;
    this.interpretation_request_id = interpretation_request_id;
    this.cohort_id = cohort_id;
    this.clinical_report = clinical_report;
    this.interpretation_request_data = interpretation_request_data;
    this.interpreted_genome = interpreted_genome;
  }

  public org.apache.avro.Schema getSchema() { return SCHEMA$; }
  // Used by DatumWriter.  Applications should not call. 
  public Object get(int field$) {
    switch (field$) {
    case 0: return files;
    case 1: return number_of_samples;
    case 2: return program;
    case 3: return family_id;
    case 4: return labkey_links;
    case 5: return created_at;
    case 6: return last_modified;
    case 7: return proband;
    case 8: return status;
    case 9: return case_priority;
    case 10: return assembly;
    case 11: return case_id;
    case 12: return interpretation_request_id;
    case 13: return cohort_id;
    case 14: return clinical_report;
    case 15: return interpretation_request_data;
    case 16: return interpreted_genome;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }
  // Used by DatumReader.  Applications should not call. 
  @SuppressWarnings(value="unchecked")
  public void put(int field$, Object value$) {
    switch (field$) {
    case 0: files = (java.util.List<DummyFile>)value$; break;
    case 1: number_of_samples = (Integer)value$; break;
    case 2: program = (String)value$; break;
    case 3: family_id = (String)value$; break;
    case 4: labkey_links = (java.util.List<String>)value$; break;
    case 5: created_at = (String)value$; break;
    case 6: last_modified = (String)value$; break;
    case 7: proband = (String)value$; break;
    case 8: status = (java.util.List<DummyStatus>)value$; break;
    case 9: case_priority = (String)value$; break;
    case 10: assembly = (String)value$; break;
    case 11: case_id = (String)value$; break;
    case 12: interpretation_request_id = (Integer)value$; break;
    case 13: cohort_id = (String)value$; break;
    case 14: clinical_report = (java.util.List<DummyClinicalReport>)value$; break;
    case 15: interpretation_request_data = (DummyInterpretationRequest)value$; break;
    case 16: interpreted_genome = (java.util.List<DummyInterpretedGenome>)value$; break;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }

  /**
   * Gets the value of the 'files' field.
   */
  public java.util.List<DummyFile> getFiles() {
    return files;
  }

  /**
   * Sets the value of the 'files' field.
   * @param value the value to set.
   */
  public void setFiles(java.util.List<DummyFile> value) {
    this.files = value;
  }

  /**
   * Gets the value of the 'number_of_samples' field.
   */
  public Integer getNumberOfSamples() {
    return number_of_samples;
  }

  /**
   * Sets the value of the 'number_of_samples' field.
   * @param value the value to set.
   */
  public void setNumberOfSamples(Integer value) {
    this.number_of_samples = value;
  }

  /**
   * Gets the value of the 'program' field.
   */
  public String getProgram() {
    return program;
  }

  /**
   * Sets the value of the 'program' field.
   * @param value the value to set.
   */
  public void setProgram(String value) {
    this.program = value;
  }

  /**
   * Gets the value of the 'family_id' field.
   */
  public String getFamilyId() {
    return family_id;
  }

  /**
   * Sets the value of the 'family_id' field.
   * @param value the value to set.
   */
  public void setFamilyId(String value) {
    this.family_id = value;
  }

  /**
   * Gets the value of the 'labkey_links' field.
   */
  public java.util.List<String> getLabkeyLinks() {
    return labkey_links;
  }

  /**
   * Sets the value of the 'labkey_links' field.
   * @param value the value to set.
   */
  public void setLabkeyLinks(java.util.List<String> value) {
    this.labkey_links = value;
  }

  /**
   * Gets the value of the 'created_at' field.
   */
  public String getCreatedAt() {
    return created_at;
  }

  /**
   * Sets the value of the 'created_at' field.
   * @param value the value to set.
   */
  public void setCreatedAt(String value) {
    this.created_at = value;
  }

  /**
   * Gets the value of the 'last_modified' field.
   */
  public String getLastModified() {
    return last_modified;
  }

  /**
   * Sets the value of the 'last_modified' field.
   * @param value the value to set.
   */
  public void setLastModified(String value) {
    this.last_modified = value;
  }

  /**
   * Gets the value of the 'proband' field.
   */
  public String getProband() {
    return proband;
  }

  /**
   * Sets the value of the 'proband' field.
   * @param value the value to set.
   */
  public void setProband(String value) {
    this.proband = value;
  }

  /**
   * Gets the value of the 'status' field.
   */
  public java.util.List<DummyStatus> getStatus() {
    return status;
  }

  /**
   * Sets the value of the 'status' field.
   * @param value the value to set.
   */
  public void setStatus(java.util.List<DummyStatus> value) {
    this.status = value;
  }

  /**
   * Gets the value of the 'case_priority' field.
   */
  public String getCasePriority() {
    return case_priority;
  }

  /**
   * Sets the value of the 'case_priority' field.
   * @param value the value to set.
   */
  public void setCasePriority(String value) {
    this.case_priority = value;
  }

  /**
   * Gets the value of the 'assembly' field.
   */
  public String getAssembly() {
    return assembly;
  }

  /**
   * Sets the value of the 'assembly' field.
   * @param value the value to set.
   */
  public void setAssembly(String value) {
    this.assembly = value;
  }

  /**
   * Gets the value of the 'case_id' field.
   */
  public String getCaseId() {
    return case_id;
  }

  /**
   * Sets the value of the 'case_id' field.
   * @param value the value to set.
   */
  public void setCaseId(String value) {
    this.case_id = value;
  }

  /**
   * Gets the value of the 'interpretation_request_id' field.
   */
  public Integer getInterpretationRequestId() {
    return interpretation_request_id;
  }

  /**
   * Sets the value of the 'interpretation_request_id' field.
   * @param value the value to set.
   */
  public void setInterpretationRequestId(Integer value) {
    this.interpretation_request_id = value;
  }

  /**
   * Gets the value of the 'cohort_id' field.
   */
  public String getCohortId() {
    return cohort_id;
  }

  /**
   * Sets the value of the 'cohort_id' field.
   * @param value the value to set.
   */
  public void setCohortId(String value) {
    this.cohort_id = value;
  }

  /**
   * Gets the value of the 'clinical_report' field.
   */
  public java.util.List<DummyClinicalReport> getClinicalReport() {
    return clinical_report;
  }

  /**
   * Sets the value of the 'clinical_report' field.
   * @param value the value to set.
   */
  public void setClinicalReport(java.util.List<DummyClinicalReport> value) {
    this.clinical_report = value;
  }

  /**
   * Gets the value of the 'interpretation_request_data' field.
   */
  public DummyInterpretationRequest getInterpretationRequestData() {
    return interpretation_request_data;
  }

  /**
   * Sets the value of the 'interpretation_request_data' field.
   * @param value the value to set.
   */
  public void setInterpretationRequestData(DummyInterpretationRequest value) {
    this.interpretation_request_data = value;
  }

  /**
   * Gets the value of the 'interpreted_genome' field.
   */
  public java.util.List<DummyInterpretedGenome> getInterpretedGenome() {
    return interpreted_genome;
  }

  /**
   * Sets the value of the 'interpreted_genome' field.
   * @param value the value to set.
   */
  public void setInterpretedGenome(java.util.List<DummyInterpretedGenome> value) {
    this.interpreted_genome = value;
  }

  /** Creates a new CipapiFacade RecordBuilder */
  public static Builder newBuilder() {
    return new Builder();
  }
  
  /** Creates a new CipapiFacade RecordBuilder by copying an existing Builder */
  public static Builder newBuilder(Builder other) {
    return new Builder(other);
  }
  
  /** Creates a new CipapiFacade RecordBuilder by copying an existing CipapiFacade instance */
  public static Builder newBuilder(CipapiFacade other) {
    return new Builder(other);
  }
  
  /**
   * RecordBuilder for CipapiFacade instances.
   */
  public static class Builder extends org.apache.avro.specific.SpecificRecordBuilderBase<CipapiFacade>
    implements org.apache.avro.data.RecordBuilder<CipapiFacade> {

    private java.util.List<DummyFile> files;
    private int number_of_samples;
    private String program;
    private String family_id;
    private java.util.List<String> labkey_links;
    private String created_at;
    private String last_modified;
    private String proband;
    private java.util.List<DummyStatus> status;
    private String case_priority;
    private String assembly;
    private String case_id;
    private int interpretation_request_id;
    private String cohort_id;
    private java.util.List<DummyClinicalReport> clinical_report;
    private DummyInterpretationRequest interpretation_request_data;
    private java.util.List<DummyInterpretedGenome> interpreted_genome;

    /** Creates a new Builder */
    private Builder() {
      super(CipapiFacade.SCHEMA$);
    }
    
    /** Creates a Builder by copying an existing Builder */
    private Builder(Builder other) {
      super(other);
      if (isValidValue(fields()[0], other.files)) {
        this.files = data().deepCopy(fields()[0].schema(), other.files);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.number_of_samples)) {
        this.number_of_samples = data().deepCopy(fields()[1].schema(), other.number_of_samples);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.program)) {
        this.program = data().deepCopy(fields()[2].schema(), other.program);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.family_id)) {
        this.family_id = data().deepCopy(fields()[3].schema(), other.family_id);
        fieldSetFlags()[3] = true;
      }
      if (isValidValue(fields()[4], other.labkey_links)) {
        this.labkey_links = data().deepCopy(fields()[4].schema(), other.labkey_links);
        fieldSetFlags()[4] = true;
      }
      if (isValidValue(fields()[5], other.created_at)) {
        this.created_at = data().deepCopy(fields()[5].schema(), other.created_at);
        fieldSetFlags()[5] = true;
      }
      if (isValidValue(fields()[6], other.last_modified)) {
        this.last_modified = data().deepCopy(fields()[6].schema(), other.last_modified);
        fieldSetFlags()[6] = true;
      }
      if (isValidValue(fields()[7], other.proband)) {
        this.proband = data().deepCopy(fields()[7].schema(), other.proband);
        fieldSetFlags()[7] = true;
      }
      if (isValidValue(fields()[8], other.status)) {
        this.status = data().deepCopy(fields()[8].schema(), other.status);
        fieldSetFlags()[8] = true;
      }
      if (isValidValue(fields()[9], other.case_priority)) {
        this.case_priority = data().deepCopy(fields()[9].schema(), other.case_priority);
        fieldSetFlags()[9] = true;
      }
      if (isValidValue(fields()[10], other.assembly)) {
        this.assembly = data().deepCopy(fields()[10].schema(), other.assembly);
        fieldSetFlags()[10] = true;
      }
      if (isValidValue(fields()[11], other.case_id)) {
        this.case_id = data().deepCopy(fields()[11].schema(), other.case_id);
        fieldSetFlags()[11] = true;
      }
      if (isValidValue(fields()[12], other.interpretation_request_id)) {
        this.interpretation_request_id = data().deepCopy(fields()[12].schema(), other.interpretation_request_id);
        fieldSetFlags()[12] = true;
      }
      if (isValidValue(fields()[13], other.cohort_id)) {
        this.cohort_id = data().deepCopy(fields()[13].schema(), other.cohort_id);
        fieldSetFlags()[13] = true;
      }
      if (isValidValue(fields()[14], other.clinical_report)) {
        this.clinical_report = data().deepCopy(fields()[14].schema(), other.clinical_report);
        fieldSetFlags()[14] = true;
      }
      if (isValidValue(fields()[15], other.interpretation_request_data)) {
        this.interpretation_request_data = data().deepCopy(fields()[15].schema(), other.interpretation_request_data);
        fieldSetFlags()[15] = true;
      }
      if (isValidValue(fields()[16], other.interpreted_genome)) {
        this.interpreted_genome = data().deepCopy(fields()[16].schema(), other.interpreted_genome);
        fieldSetFlags()[16] = true;
      }
    }
    
    /** Creates a Builder by copying an existing CipapiFacade instance */
    private Builder(CipapiFacade other) {
            super(CipapiFacade.SCHEMA$);
      if (isValidValue(fields()[0], other.files)) {
        this.files = data().deepCopy(fields()[0].schema(), other.files);
        fieldSetFlags()[0] = true;
      }
      if (isValidValue(fields()[1], other.number_of_samples)) {
        this.number_of_samples = data().deepCopy(fields()[1].schema(), other.number_of_samples);
        fieldSetFlags()[1] = true;
      }
      if (isValidValue(fields()[2], other.program)) {
        this.program = data().deepCopy(fields()[2].schema(), other.program);
        fieldSetFlags()[2] = true;
      }
      if (isValidValue(fields()[3], other.family_id)) {
        this.family_id = data().deepCopy(fields()[3].schema(), other.family_id);
        fieldSetFlags()[3] = true;
      }
      if (isValidValue(fields()[4], other.labkey_links)) {
        this.labkey_links = data().deepCopy(fields()[4].schema(), other.labkey_links);
        fieldSetFlags()[4] = true;
      }
      if (isValidValue(fields()[5], other.created_at)) {
        this.created_at = data().deepCopy(fields()[5].schema(), other.created_at);
        fieldSetFlags()[5] = true;
      }
      if (isValidValue(fields()[6], other.last_modified)) {
        this.last_modified = data().deepCopy(fields()[6].schema(), other.last_modified);
        fieldSetFlags()[6] = true;
      }
      if (isValidValue(fields()[7], other.proband)) {
        this.proband = data().deepCopy(fields()[7].schema(), other.proband);
        fieldSetFlags()[7] = true;
      }
      if (isValidValue(fields()[8], other.status)) {
        this.status = data().deepCopy(fields()[8].schema(), other.status);
        fieldSetFlags()[8] = true;
      }
      if (isValidValue(fields()[9], other.case_priority)) {
        this.case_priority = data().deepCopy(fields()[9].schema(), other.case_priority);
        fieldSetFlags()[9] = true;
      }
      if (isValidValue(fields()[10], other.assembly)) {
        this.assembly = data().deepCopy(fields()[10].schema(), other.assembly);
        fieldSetFlags()[10] = true;
      }
      if (isValidValue(fields()[11], other.case_id)) {
        this.case_id = data().deepCopy(fields()[11].schema(), other.case_id);
        fieldSetFlags()[11] = true;
      }
      if (isValidValue(fields()[12], other.interpretation_request_id)) {
        this.interpretation_request_id = data().deepCopy(fields()[12].schema(), other.interpretation_request_id);
        fieldSetFlags()[12] = true;
      }
      if (isValidValue(fields()[13], other.cohort_id)) {
        this.cohort_id = data().deepCopy(fields()[13].schema(), other.cohort_id);
        fieldSetFlags()[13] = true;
      }
      if (isValidValue(fields()[14], other.clinical_report)) {
        this.clinical_report = data().deepCopy(fields()[14].schema(), other.clinical_report);
        fieldSetFlags()[14] = true;
      }
      if (isValidValue(fields()[15], other.interpretation_request_data)) {
        this.interpretation_request_data = data().deepCopy(fields()[15].schema(), other.interpretation_request_data);
        fieldSetFlags()[15] = true;
      }
      if (isValidValue(fields()[16], other.interpreted_genome)) {
        this.interpreted_genome = data().deepCopy(fields()[16].schema(), other.interpreted_genome);
        fieldSetFlags()[16] = true;
      }
    }

    /** Gets the value of the 'files' field */
    public java.util.List<DummyFile> getFiles() {
      return files;
    }
    
    /** Sets the value of the 'files' field */
    public Builder setFiles(java.util.List<DummyFile> value) {
      validate(fields()[0], value);
      this.files = value;
      fieldSetFlags()[0] = true;
      return this; 
    }
    
    /** Checks whether the 'files' field has been set */
    public boolean hasFiles() {
      return fieldSetFlags()[0];
    }
    
    /** Clears the value of the 'files' field */
    public Builder clearFiles() {
      files = null;
      fieldSetFlags()[0] = false;
      return this;
    }

    /** Gets the value of the 'number_of_samples' field */
    public Integer getNumberOfSamples() {
      return number_of_samples;
    }
    
    /** Sets the value of the 'number_of_samples' field */
    public Builder setNumberOfSamples(int value) {
      validate(fields()[1], value);
      this.number_of_samples = value;
      fieldSetFlags()[1] = true;
      return this; 
    }
    
    /** Checks whether the 'number_of_samples' field has been set */
    public boolean hasNumberOfSamples() {
      return fieldSetFlags()[1];
    }
    
    /** Clears the value of the 'number_of_samples' field */
    public Builder clearNumberOfSamples() {
      fieldSetFlags()[1] = false;
      return this;
    }

    /** Gets the value of the 'program' field */
    public String getProgram() {
      return program;
    }
    
    /** Sets the value of the 'program' field */
    public Builder setProgram(String value) {
      validate(fields()[2], value);
      this.program = value;
      fieldSetFlags()[2] = true;
      return this; 
    }
    
    /** Checks whether the 'program' field has been set */
    public boolean hasProgram() {
      return fieldSetFlags()[2];
    }
    
    /** Clears the value of the 'program' field */
    public Builder clearProgram() {
      program = null;
      fieldSetFlags()[2] = false;
      return this;
    }

    /** Gets the value of the 'family_id' field */
    public String getFamilyId() {
      return family_id;
    }
    
    /** Sets the value of the 'family_id' field */
    public Builder setFamilyId(String value) {
      validate(fields()[3], value);
      this.family_id = value;
      fieldSetFlags()[3] = true;
      return this; 
    }
    
    /** Checks whether the 'family_id' field has been set */
    public boolean hasFamilyId() {
      return fieldSetFlags()[3];
    }
    
    /** Clears the value of the 'family_id' field */
    public Builder clearFamilyId() {
      family_id = null;
      fieldSetFlags()[3] = false;
      return this;
    }

    /** Gets the value of the 'labkey_links' field */
    public java.util.List<String> getLabkeyLinks() {
      return labkey_links;
    }
    
    /** Sets the value of the 'labkey_links' field */
    public Builder setLabkeyLinks(java.util.List<String> value) {
      validate(fields()[4], value);
      this.labkey_links = value;
      fieldSetFlags()[4] = true;
      return this; 
    }
    
    /** Checks whether the 'labkey_links' field has been set */
    public boolean hasLabkeyLinks() {
      return fieldSetFlags()[4];
    }
    
    /** Clears the value of the 'labkey_links' field */
    public Builder clearLabkeyLinks() {
      labkey_links = null;
      fieldSetFlags()[4] = false;
      return this;
    }

    /** Gets the value of the 'created_at' field */
    public String getCreatedAt() {
      return created_at;
    }
    
    /** Sets the value of the 'created_at' field */
    public Builder setCreatedAt(String value) {
      validate(fields()[5], value);
      this.created_at = value;
      fieldSetFlags()[5] = true;
      return this; 
    }
    
    /** Checks whether the 'created_at' field has been set */
    public boolean hasCreatedAt() {
      return fieldSetFlags()[5];
    }
    
    /** Clears the value of the 'created_at' field */
    public Builder clearCreatedAt() {
      created_at = null;
      fieldSetFlags()[5] = false;
      return this;
    }

    /** Gets the value of the 'last_modified' field */
    public String getLastModified() {
      return last_modified;
    }
    
    /** Sets the value of the 'last_modified' field */
    public Builder setLastModified(String value) {
      validate(fields()[6], value);
      this.last_modified = value;
      fieldSetFlags()[6] = true;
      return this; 
    }
    
    /** Checks whether the 'last_modified' field has been set */
    public boolean hasLastModified() {
      return fieldSetFlags()[6];
    }
    
    /** Clears the value of the 'last_modified' field */
    public Builder clearLastModified() {
      last_modified = null;
      fieldSetFlags()[6] = false;
      return this;
    }

    /** Gets the value of the 'proband' field */
    public String getProband() {
      return proband;
    }
    
    /** Sets the value of the 'proband' field */
    public Builder setProband(String value) {
      validate(fields()[7], value);
      this.proband = value;
      fieldSetFlags()[7] = true;
      return this; 
    }
    
    /** Checks whether the 'proband' field has been set */
    public boolean hasProband() {
      return fieldSetFlags()[7];
    }
    
    /** Clears the value of the 'proband' field */
    public Builder clearProband() {
      proband = null;
      fieldSetFlags()[7] = false;
      return this;
    }

    /** Gets the value of the 'status' field */
    public java.util.List<DummyStatus> getStatus() {
      return status;
    }
    
    /** Sets the value of the 'status' field */
    public Builder setStatus(java.util.List<DummyStatus> value) {
      validate(fields()[8], value);
      this.status = value;
      fieldSetFlags()[8] = true;
      return this; 
    }
    
    /** Checks whether the 'status' field has been set */
    public boolean hasStatus() {
      return fieldSetFlags()[8];
    }
    
    /** Clears the value of the 'status' field */
    public Builder clearStatus() {
      status = null;
      fieldSetFlags()[8] = false;
      return this;
    }

    /** Gets the value of the 'case_priority' field */
    public String getCasePriority() {
      return case_priority;
    }
    
    /** Sets the value of the 'case_priority' field */
    public Builder setCasePriority(String value) {
      validate(fields()[9], value);
      this.case_priority = value;
      fieldSetFlags()[9] = true;
      return this; 
    }
    
    /** Checks whether the 'case_priority' field has been set */
    public boolean hasCasePriority() {
      return fieldSetFlags()[9];
    }
    
    /** Clears the value of the 'case_priority' field */
    public Builder clearCasePriority() {
      case_priority = null;
      fieldSetFlags()[9] = false;
      return this;
    }

    /** Gets the value of the 'assembly' field */
    public String getAssembly() {
      return assembly;
    }
    
    /** Sets the value of the 'assembly' field */
    public Builder setAssembly(String value) {
      validate(fields()[10], value);
      this.assembly = value;
      fieldSetFlags()[10] = true;
      return this; 
    }
    
    /** Checks whether the 'assembly' field has been set */
    public boolean hasAssembly() {
      return fieldSetFlags()[10];
    }
    
    /** Clears the value of the 'assembly' field */
    public Builder clearAssembly() {
      assembly = null;
      fieldSetFlags()[10] = false;
      return this;
    }

    /** Gets the value of the 'case_id' field */
    public String getCaseId() {
      return case_id;
    }
    
    /** Sets the value of the 'case_id' field */
    public Builder setCaseId(String value) {
      validate(fields()[11], value);
      this.case_id = value;
      fieldSetFlags()[11] = true;
      return this; 
    }
    
    /** Checks whether the 'case_id' field has been set */
    public boolean hasCaseId() {
      return fieldSetFlags()[11];
    }
    
    /** Clears the value of the 'case_id' field */
    public Builder clearCaseId() {
      case_id = null;
      fieldSetFlags()[11] = false;
      return this;
    }

    /** Gets the value of the 'interpretation_request_id' field */
    public Integer getInterpretationRequestId() {
      return interpretation_request_id;
    }
    
    /** Sets the value of the 'interpretation_request_id' field */
    public Builder setInterpretationRequestId(int value) {
      validate(fields()[12], value);
      this.interpretation_request_id = value;
      fieldSetFlags()[12] = true;
      return this; 
    }
    
    /** Checks whether the 'interpretation_request_id' field has been set */
    public boolean hasInterpretationRequestId() {
      return fieldSetFlags()[12];
    }
    
    /** Clears the value of the 'interpretation_request_id' field */
    public Builder clearInterpretationRequestId() {
      fieldSetFlags()[12] = false;
      return this;
    }

    /** Gets the value of the 'cohort_id' field */
    public String getCohortId() {
      return cohort_id;
    }
    
    /** Sets the value of the 'cohort_id' field */
    public Builder setCohortId(String value) {
      validate(fields()[13], value);
      this.cohort_id = value;
      fieldSetFlags()[13] = true;
      return this; 
    }
    
    /** Checks whether the 'cohort_id' field has been set */
    public boolean hasCohortId() {
      return fieldSetFlags()[13];
    }
    
    /** Clears the value of the 'cohort_id' field */
    public Builder clearCohortId() {
      cohort_id = null;
      fieldSetFlags()[13] = false;
      return this;
    }

    /** Gets the value of the 'clinical_report' field */
    public java.util.List<DummyClinicalReport> getClinicalReport() {
      return clinical_report;
    }
    
    /** Sets the value of the 'clinical_report' field */
    public Builder setClinicalReport(java.util.List<DummyClinicalReport> value) {
      validate(fields()[14], value);
      this.clinical_report = value;
      fieldSetFlags()[14] = true;
      return this; 
    }
    
    /** Checks whether the 'clinical_report' field has been set */
    public boolean hasClinicalReport() {
      return fieldSetFlags()[14];
    }
    
    /** Clears the value of the 'clinical_report' field */
    public Builder clearClinicalReport() {
      clinical_report = null;
      fieldSetFlags()[14] = false;
      return this;
    }

    /** Gets the value of the 'interpretation_request_data' field */
    public DummyInterpretationRequest getInterpretationRequestData() {
      return interpretation_request_data;
    }
    
    /** Sets the value of the 'interpretation_request_data' field */
    public Builder setInterpretationRequestData(DummyInterpretationRequest value) {
      validate(fields()[15], value);
      this.interpretation_request_data = value;
      fieldSetFlags()[15] = true;
      return this; 
    }
    
    /** Checks whether the 'interpretation_request_data' field has been set */
    public boolean hasInterpretationRequestData() {
      return fieldSetFlags()[15];
    }
    
    /** Clears the value of the 'interpretation_request_data' field */
    public Builder clearInterpretationRequestData() {
      interpretation_request_data = null;
      fieldSetFlags()[15] = false;
      return this;
    }

    /** Gets the value of the 'interpreted_genome' field */
    public java.util.List<DummyInterpretedGenome> getInterpretedGenome() {
      return interpreted_genome;
    }
    
    /** Sets the value of the 'interpreted_genome' field */
    public Builder setInterpretedGenome(java.util.List<DummyInterpretedGenome> value) {
      validate(fields()[16], value);
      this.interpreted_genome = value;
      fieldSetFlags()[16] = true;
      return this; 
    }
    
    /** Checks whether the 'interpreted_genome' field has been set */
    public boolean hasInterpretedGenome() {
      return fieldSetFlags()[16];
    }
    
    /** Clears the value of the 'interpreted_genome' field */
    public Builder clearInterpretedGenome() {
      interpreted_genome = null;
      fieldSetFlags()[16] = false;
      return this;
    }

    @Override
    public CipapiFacade build() {
      try {
        CipapiFacade record = new CipapiFacade();
        record.files = fieldSetFlags()[0] ? this.files : (java.util.List<DummyFile>) defaultValue(fields()[0]);
        record.number_of_samples = fieldSetFlags()[1] ? this.number_of_samples : (Integer) defaultValue(fields()[1]);
        record.program = fieldSetFlags()[2] ? this.program : (String) defaultValue(fields()[2]);
        record.family_id = fieldSetFlags()[3] ? this.family_id : (String) defaultValue(fields()[3]);
        record.labkey_links = fieldSetFlags()[4] ? this.labkey_links : (java.util.List<String>) defaultValue(fields()[4]);
        record.created_at = fieldSetFlags()[5] ? this.created_at : (String) defaultValue(fields()[5]);
        record.last_modified = fieldSetFlags()[6] ? this.last_modified : (String) defaultValue(fields()[6]);
        record.proband = fieldSetFlags()[7] ? this.proband : (String) defaultValue(fields()[7]);
        record.status = fieldSetFlags()[8] ? this.status : (java.util.List<DummyStatus>) defaultValue(fields()[8]);
        record.case_priority = fieldSetFlags()[9] ? this.case_priority : (String) defaultValue(fields()[9]);
        record.assembly = fieldSetFlags()[10] ? this.assembly : (String) defaultValue(fields()[10]);
        record.case_id = fieldSetFlags()[11] ? this.case_id : (String) defaultValue(fields()[11]);
        record.interpretation_request_id = fieldSetFlags()[12] ? this.interpretation_request_id : (Integer) defaultValue(fields()[12]);
        record.cohort_id = fieldSetFlags()[13] ? this.cohort_id : (String) defaultValue(fields()[13]);
        record.clinical_report = fieldSetFlags()[14] ? this.clinical_report : (java.util.List<DummyClinicalReport>) defaultValue(fields()[14]);
        record.interpretation_request_data = fieldSetFlags()[15] ? this.interpretation_request_data : (DummyInterpretationRequest) defaultValue(fields()[15]);
        record.interpreted_genome = fieldSetFlags()[16] ? this.interpreted_genome : (java.util.List<DummyInterpretedGenome>) defaultValue(fields()[16]);
        return record;
      } catch (Exception e) {
        throw new org.apache.avro.AvroRuntimeException(e);
      }
    }
  }
}
