/**
 * Autogenerated by Avro
 * 
 * DO NOT EDIT DIRECTLY
 */
package org.opencb.biodata.external.gel_participant_6_2_0;

import com.fasterxml.jackson.databind.PropertyNamingStrategy;
import com.fasterxml.jackson.databind.annotation.JsonNaming;

@SuppressWarnings("all")
@org.apache.avro.specific.AvroGenerated
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
public class DummyInterpretationRequest extends org.apache.avro.specific.SpecificRecordBase implements org.apache.avro.specific.SpecificRecord {
  public static final org.apache.avro.Schema SCHEMA$ = new org.apache.avro.Schema.Parser().parse("{\"type\":\"record\",\"name\":\"DummyInterpretationRequest\",\"namespace\":\"org.dummy.avro\",\"fields\":[{\"name\":\"json_request\",\"type\":{\"type\":\"record\",\"name\":\"InterpretationRequestRD\",\"namespace\":\"org.gel.models.report.avro\",\"doc\":\"This record represents basic information for this report\",\"fields\":[{\"name\":\"versionControl\",\"type\":{\"type\":\"record\",\"name\":\"ReportVersionControl\",\"fields\":[{\"name\":\"gitVersionControl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"This is the version for the entire set of data models as referred to the Git release tag\",\"default\":\"6.1.1\"}]},\"doc\":\"Model version number\"},{\"name\":\"interpretationRequestId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Identifier for this interpretation request\"},{\"name\":\"interpretationRequestVersion\",\"type\":\"int\",\"doc\":\"Version for this interpretation request\"},{\"name\":\"internalStudyId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Internal study identifier\"},{\"name\":\"familyInternalId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Family internal identifier\"},{\"name\":\"genomeAssembly\",\"type\":{\"type\":\"enum\",\"name\":\"Assembly\",\"doc\":\"The reference genome assembly\",\"symbols\":[\"GRCh38\",\"GRCh37\"]},\"doc\":\"This is the version of the assembly used to align the reads\"},{\"name\":\"workspace\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}},\"doc\":\"The genome shall be assigned to the workspaces(projects or domains with a predefined set of users) to control user access\"},{\"name\":\"bams\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"File\",\"doc\":\"This defines a file\\n    This record is uniquely defined by the sample identfier and an URI\\n    Currently sample identifier can be a single string or a list of strings if multiple samples are associated with the same file\\n    *\",\"fields\":[{\"name\":\"sampleId\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Unique identifier(s) of the sample. For example in a multisample vcf this would have an array of all the sample identifiers\"},{\"name\":\"uriFile\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"URI path of the file\"},{\"name\":\"fileType\",\"type\":{\"type\":\"enum\",\"name\":\"FileType\",\"symbols\":[\"BAM\",\"gVCF\",\"VCF_small\",\"VCF_somatic_small\",\"VCF_CNV\",\"VCF_somatic_CNV\",\"VCF_SV\",\"VCF_somatic_SV\",\"VCF_SV_CNV\",\"SVG\",\"ANN\",\"BigWig\",\"MD5Sum\",\"ROH\",\"OTHER\",\"PARTITION\",\"VARIANT_FREQUENCIES\",\"COVERAGE\"]},\"doc\":\"The type of the file\"},{\"name\":\"md5Sum\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The MD5 checksum\"}]}}],\"doc\":\"BAMs Files\"},{\"name\":\"vcfs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"File\"}],\"doc\":\"VCFs Files where SVs and CNVs are represented\"},{\"name\":\"bigWigs\",\"type\":[\"null\",{\"type\":\"array\",\"items\":\"File\"}],\"doc\":\"BigWig Files\"},{\"name\":\"pedigreeDiagram\",\"type\":[\"null\",\"File\"],\"doc\":\"Pedigree Diagram Files as an SGV\"},{\"name\":\"annotationFile\",\"type\":[\"null\",\"File\"],\"doc\":\"Variant Annotation File\"},{\"name\":\"otherFiles\",\"type\":[\"null\",{\"type\":\"map\",\"values\":\"File\",\"avro.java.string\":\"String\"}],\"doc\":\"Other files that may be vendor specific\\n        map of key: type of file, value: record of type File\"},{\"name\":\"pedigree\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Pedigree\",\"namespace\":\"org.gel.models.participant.avro\",\"doc\":\"This is the concept of a family with associated phenotypes as present in the record RDParticipant\",\"fields\":[{\"name\":\"versionControl\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"VersionControl\",\"fields\":[{\"name\":\"GitVersionControl\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"This is the version for the entire set of data models as referred to the Git release tag\",\"default\":\"1.3.0\"}]}],\"doc\":\"Model version number\"},{\"name\":\"LDPCode\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"LDP Code (Local Delivery Partner)\"},{\"name\":\"familyId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Family identifier which internally translate to a sample set\"},{\"name\":\"members\",\"type\":{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"PedigreeMember\",\"doc\":\"This defines a RD Participant (demographics and pedigree information)\",\"fields\":[{\"name\":\"pedigreeId\",\"type\":[\"null\",\"int\"],\"doc\":\"Numbering used to refer to each member of the pedigree\"},{\"name\":\"isProband\",\"type\":[\"null\",\"boolean\"],\"doc\":\"If this field is true, the member should be considered the proband of this family\"},{\"name\":\"participantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"participantId.  This is the human readable ID in GMS\"},{\"name\":\"participantQCState\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"ParticipantQCState\",\"doc\":\"QCState Status\",\"symbols\":[\"noState\",\"passedMedicalReviewReadyForInterpretation\",\"passedMedicalReviewNotReadyForInterpretation\",\"queryToGel\",\"queryToGMC\",\"failed\"]}],\"doc\":\"participantQCState\"},{\"name\":\"gelSuperFamilyId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"superFamily id, this id is built as a concatenation of all\\n        families id in this superfamily i.e, fam10024_fam100457.\\n        NOTE: Retired for GMS\"},{\"name\":\"sex\",\"type\":{\"type\":\"enum\",\"name\":\"Sex\",\"doc\":\"Sex\",\"symbols\":[\"MALE\",\"FEMALE\",\"UNKNOWN\"]},\"doc\":\"Sex of the Participant\"},{\"name\":\"personKaryotypicSex\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"PersonKaryotipicSex\",\"doc\":\"Karyotipic Sex\",\"symbols\":[\"UNKNOWN\",\"XX\",\"XY\",\"XO\",\"XXY\",\"XXX\",\"XXYY\",\"XXXY\",\"XXXX\",\"XYY\",\"OTHER\"]}],\"doc\":\"Karyotypic sex of the participant as previously established or by looking at the GEL genome\"},{\"name\":\"yearOfBirth\",\"type\":[\"null\",\"int\"],\"doc\":\"Year of Birth\"},{\"name\":\"fatherId\",\"type\":[\"null\",\"int\"],\"doc\":\"refers to the pedigreeId of the father\\n        Id of the parent, if unknown then no parent is referenced. Parents may need to be entered even if no data is known\\n        about them in order to unambiguously reconstruct the pedigree.\"},{\"name\":\"motherId\",\"type\":[\"null\",\"int\"],\"doc\":\"refers to the pedigreeId of the mother\\n        Id of the parent, if unknown then no parent is referenced. Parents may need to be entered even if no data is known\\n        about them in order to unambiguously reconstruct the pedigree.\"},{\"name\":\"superFatherId\",\"type\":[\"null\",\"int\"],\"doc\":\"this id is built using the original familyId and the original pedigreeId of the father\\n        NOTE: retired for GMS\"},{\"name\":\"superMotherId\",\"type\":[\"null\",\"int\"],\"doc\":\"this id is built using the original familyId and the original pedigreeId of the mother\\n        NOTE: Retired for GMS\"},{\"name\":\"twinGroup\",\"type\":[\"null\",\"int\"],\"doc\":\"Each twin group is numbered, i.e. all members of a group of multiparous births receive the same number\"},{\"name\":\"monozygotic\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"TernaryOption\",\"doc\":\"This defines a yes/no/unknown case\",\"symbols\":[\"yes\",\"no\",\"unknown\"]}],\"doc\":\"A property of the twinning group but should be entered for all members\"},{\"name\":\"adoptedStatus\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"AdoptedStatus\",\"doc\":\"adoptedin means adopted into the family\\n    adoptedout means child belonged to the family and was adopted out\",\"symbols\":[\"notadopted\",\"adoptedin\",\"adoptedout\"]}],\"doc\":\"Adopted Status\"},{\"name\":\"lifeStatus\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"LifeStatus\",\"doc\":\"Life Status\",\"symbols\":[\"ALIVE\",\"ABORTED\",\"DECEASED\",\"UNBORN\",\"STILLBORN\",\"MISCARRIAGE\"]}],\"doc\":\"Life Status\"},{\"name\":\"consanguineousParents\",\"type\":[\"null\",\"TernaryOption\"],\"doc\":\"The parents of this participant has a consanguineous relationship\"},{\"name\":\"affectionStatus\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"AffectionStatus\",\"doc\":\"Affection Status\",\"symbols\":[\"UNAFFECTED\",\"AFFECTED\",\"UNCERTAIN\"]}],\"doc\":\"Affection Status\"},{\"name\":\"disorderList\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"Disorder\",\"doc\":\"This is quite GEL specific. This is the way is stored in ModelCatalogue and PanelApp.\\n    Currently all specific disease titles are assigned to a disease subgroup so really only specificDisease needs to be\\n    completed but we add the others for generality\",\"fields\":[{\"name\":\"diseaseGroup\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"This i","s Level2 Title for this disorder\"},{\"name\":\"diseaseSubGroup\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"This is Level3 Title for this disorder\"},{\"name\":\"specificDisease\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"This is Level4 Title for this disorder.\\n        In GMS, this is the clinicalIndicationFullName/clinicalIndicationCode.\"},{\"name\":\"ageOfOnset\",\"type\":[\"null\",\"float\"],\"doc\":\"Age of onset in years\"}]}}],\"doc\":\"Clinical Data (disorders). If the family member is unaffected as per affectionStatus then this list is empty.\"},{\"name\":\"hpoTermList\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"HpoTerm\",\"doc\":\"This defines an HPO term and its modifiers (possibly multiple)\\n    If HPO term presence is unknown we don't have a entry on the list\",\"fields\":[{\"name\":\"term\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Identifier of the HPO term\"},{\"name\":\"termPresence\",\"type\":[\"null\",\"TernaryOption\"],\"doc\":\"This is whether the term is present in the participant (default is unknown) yes=term is present in participant,\\n        no=term is not present\"},{\"name\":\"hpoBuildNumber\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"hpoBuildNumber\"},{\"name\":\"modifiers\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"HpoTermModifiers\",\"doc\":\"HPO Modifiers\\n    For GMS, hpoModifierCode and hpoModifierVersion will be used\",\"fields\":[{\"name\":\"laterality\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Laterality\",\"symbols\":[\"RIGHT\",\"UNILATERAL\",\"BILATERAL\",\"LEFT\"]}]},{\"name\":\"progression\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Progression\",\"symbols\":[\"PROGRESSIVE\",\"NONPROGRESSIVE\"]}]},{\"name\":\"severity\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Severity\",\"symbols\":[\"BORDERLINE\",\"MILD\",\"MODERATE\",\"SEVERE\",\"PROFOUND\"]}]},{\"name\":\"spatialPattern\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"SpatialPattern\",\"symbols\":[\"DISTAL\",\"GENERALIZED\",\"LOCALIZED\",\"PROXIMAL\"]}]},{\"name\":\"hpoModifierCode\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"hpoModifierVersion\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]}]}}],\"doc\":\"Modifier associated with the HPO term\"},{\"name\":\"ageOfOnset\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"AgeOfOnset\",\"symbols\":[\"EMBRYONAL_ONSET\",\"FETAL_ONSET\",\"NEONATAL_ONSET\",\"INFANTILE_ONSET\",\"CHILDHOOD_ONSET\",\"JUVENILE_ONSET\",\"YOUNG_ADULT_ONSET\",\"LATE_ONSET\",\"MIDDLE_AGE_ONSET\"]}],\"doc\":\"Age of onset in months\"}]}}],\"doc\":\"Clinical Data (HPO terms)\"},{\"name\":\"ancestries\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Ancestries\",\"doc\":\"Ancestries, defined as Ethnic category(ies) and Chi-square test\",\"fields\":[{\"name\":\"mothersEthnicOrigin\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"EthnicCategory\",\"doc\":\"This is the list of ethnicities in ONS16\\n\\n    * `D`:  Mixed: White and Black Caribbean\\n    * `E`:  Mixed: White and Black African\\n    * `F`:  Mixed: White and Asian\\n    * `G`:  Mixed: Any other mixed background\\n    * `A`:  White: British\\n    * `B`:  White: Irish\\n    * `C`:  White: Any other White background\\n    * `L`:  Asian or Asian British: Any other Asian background\\n    * `M`:  Black or Black British: Caribbean\\n    * `N`:  Black or Black British: African\\n    * `H`:  Asian or Asian British: Indian\\n    * `J`:  Asian or Asian British: Pakistani\\n    * `K`:  Asian or Asian British: Bangladeshi\\n    * `P`:  Black or Black British: Any other Black background\\n    * `S`:  Other Ethnic Groups: Any other ethnic group\\n    * `R`:  Other Ethnic Groups: Chinese\\n    * `Z`:  Not stated\",\"symbols\":[\"D\",\"E\",\"F\",\"G\",\"A\",\"B\",\"C\",\"L\",\"M\",\"N\",\"H\",\"J\",\"K\",\"P\",\"S\",\"R\",\"Z\"]}],\"doc\":\"Mother's Ethnic Origin\"},{\"name\":\"mothersOtherRelevantAncestry\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Mother's Ethnic Origin Description\"},{\"name\":\"fathersEthnicOrigin\",\"type\":[\"null\",\"EthnicCategory\"],\"doc\":\"Father's Ethnic Origin\"},{\"name\":\"fathersOtherRelevantAncestry\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Father's Ethnic Origin Description\"},{\"name\":\"chiSquare1KGenomesPhase3Pop\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ChiSquare1KGenomesPhase3Pop\",\"doc\":\"Chi-square test for goodness of fit of this sample to 1000 Genomes Phase 3 populations\",\"fields\":[{\"name\":\"kgSuperPopCategory\",\"type\":{\"type\":\"enum\",\"name\":\"KgSuperPopCategory\",\"doc\":\"1K Genomes project super populations\",\"symbols\":[\"AFR\",\"AMR\",\"EAS\",\"EUR\",\"SAS\"]},\"doc\":\"1K Super Population\"},{\"name\":\"kgPopCategory\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"KgPopCategory\",\"doc\":\"1K Genomes project populations\",\"symbols\":[\"ACB\",\"ASW\",\"BEB\",\"CDX\",\"CEU\",\"CHB\",\"CHS\",\"CLM\",\"ESN\",\"FIN\",\"GBR\",\"GIH\",\"GWD\",\"IBS\",\"ITU\",\"JPT\",\"KHV\",\"LWK\",\"MSL\",\"MXL\",\"PEL\",\"PJL\",\"PUR\",\"STU\",\"TSI\",\"YRI\"]}],\"doc\":\"1K Population\"},{\"name\":\"chiSquare\",\"type\":\"double\",\"doc\":\"Chi-square test for goodness of fit of this sample to this 1000 Genomes Phase 3 population\"}]}}],\"doc\":\"Chi-square test for goodness of fit of this sample to 1000 Genomes Phase 3 populations\"}]}],\"doc\":\"Participant's ancestries, defined as Mother's/Father's Ethnic Origin and Chi-square test for goodness of fit of this sample to 1000 Genomes Phase 3 populations.\\n        NOTE: for GMS this field has been deprecated in favour of clinicalEthnicities\"},{\"name\":\"consentStatus\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"ConsentStatus\",\"doc\":\"Consent Status for 100k program\",\"fields\":[{\"name\":\"programmeConsent\",\"type\":\"boolean\",\"doc\":\"Is this individual consented to the programme?\\n        It could simply be a family member that is not consented but for whom affection status is known\",\"default\":false},{\"name\":\"primaryFindingConsent\",\"type\":\"boolean\",\"doc\":\"Consent for feedback of primary findings?\",\"default\":false},{\"name\":\"secondaryFindingConsent\",\"type\":\"boolean\",\"doc\":\"Consent for secondary finding lookup\",\"default\":false},{\"name\":\"carrierStatusConsent\",\"type\":\"boolean\",\"doc\":\"Consent for carrier status check?\",\"default\":false}]}],\"doc\":\"What has this participant consented to?\\n        A participant that has been consented to the programme should also have sequence data associated with them; however\\n        this needs to be programmatically checked\"},{\"name\":\"testConsentStatus\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"GmsConsentStatus\",\"doc\":\"Consent Status for GMS\",\"fields\":[{\"name\":\"programmeConsent\",\"type\":{\"type\":\"enum\",\"name\":\"GenericConsent\",\"doc\":\"clinicalEthnicities supersedes Ancestries in GMS\",\"symbols\":[\"yes\",\"no\",\"undefined\",\"not_applicable\"]},\"doc\":\"Is this individual consented to the programme? It could simply be a family member that is not consented\\n        but for whom affection status is known\"},{\"name\":\"primaryFindingConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for feedback of primary findings?\\n        RD: Primary Findings\\n        Cancer: PrimaryFindings is somatic + pertinent germline findings\"},{\"name\":\"researchConsent\",\"type\":\"GenericConsent\",\"doc\":\"Research Consent\"},{\"name\":\"healthRelatedFindingConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for secondary health related findings?\"},{\"name\":\"carrierStatusConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for carrier status check?\"},{\"name\":\"pharmacogenomicsFindingConsent\",\"type\":\"GenericConsent\",\"doc\":\"Consent for pharmacogenomics consent as secondary findings?\"}]}],\"doc\":\"What has this participant consented in the context of a Genomic Test?\"},{\"name\":\"samples\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"GermlineSample\",\"doc\":\"A germline sample\",\"fields\":[{\"name\":\"sampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Sample identifier (e.g, LP00012645_5GH))\"},{\"name\":\"labSampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Lab sample identifier\"},{\"name\":\"LDPCode\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"LDP Code (Local Delivery Partner)\"},{\"name\":\"source\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"SampleSource\",\"doc\":\"The source of the sample\\n    NOTE: IN GMS, BONE_MARROW_ASPIRATE_TUMOUR_CELLS and BONE_MARROW_ASPIRATE_TUMOUR_SORTED_CELLS are deprecated as they have been separated into their respective biotypes\",\"symbols\":[\"AMNIOTIC_FLUID\",\"","BLOOD\",\"BONE_MARROW\",\"BONE_MARROW_ASPIRATE_TUMOUR_CELLS\",\"BONE_MARROW_ASPIRATE_TUMOUR_SORTED_CELLS\",\"BUCCAL_SWAB\",\"CHORIONIC_VILLUS_SAMPLE\",\"FIBROBLAST\",\"FLUID\",\"FRESH_TISSUE_IN_CULTURE_MEDIUM\",\"OTHER\",\"SALIVA\",\"TISSUE\",\"TUMOUR\",\"URINE\"]}],\"doc\":\"Source of the sample\"},{\"name\":\"product\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Product\",\"symbols\":[\"DNA\",\"RNA\"]}],\"doc\":\"Product of the sample\"},{\"name\":\"preparationMethod\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"PreparationMethod\",\"doc\":\"In 100K, preparation Method of sample\\n    NOTE: In GMS, this field is deprecated in favour of StorageMedium and Method\",\"symbols\":[\"ASPIRATE\",\"CD128_SORTED_CELLS\",\"CD138_SORTED_CELLS\",\"EDTA\",\"FF\",\"FFPE\",\"LI_HEP\",\"ORAGENE\"]}],\"doc\":\"Preparation method\\n        NOTE: In GMS, this has been deprecated in favour of Method and storageMedium\"},{\"name\":\"programmePhase\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"ProgrammePhase\",\"symbols\":[\"CRUK\",\"OXFORD\",\"CLL\",\"IIP\",\"MAIN\",\"EXPT\"]}],\"doc\":\"Genomics England programme phase\"},{\"name\":\"clinicalSampleDateTime\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The time when the sample was received. In the format YYYY-MM-DDTHH:MM:SS+0000\"},{\"name\":\"participantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"participantUid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Participant UId of the sample\"},{\"name\":\"sampleUid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"maskedPid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"method\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"Method\",\"doc\":\"In GMS, Method is defined as how the sample was taken directly from the patient\",\"symbols\":[\"ASPIRATE\",\"BIOPSY\",\"NOT_APPLICABLE\",\"RESECTION\",\"SORTED_OTHER\",\"UNKNOWN\",\"UNSORTED\",\"CD138_SORTED\"]}],\"doc\":\"In GMS, this is how the sample was extracted from the participant\"},{\"name\":\"storageMedium\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"StorageMedium\",\"doc\":\"In GMS, storage medium of sample\",\"symbols\":[\"EDTA\",\"FF\",\"LI_HEP\",\"ORAGENE\",\"FFPE\"]}],\"doc\":\"In GMS, this is what solvent/medium the sample was stored in\"},{\"name\":\"sampleType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"In GMS, this is the sampleType as entered by the clinician in TOMs\"},{\"name\":\"sampleState\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"In GMS, this is the sampleState as entered by the clinician in TOMs\"}]}}],\"doc\":\"This is an array containing all the samples that belong to this individual, e.g [\\\"LP00002255_GA4\\\"]\"},{\"name\":\"inbreedingCoefficient\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"InbreedingCoefficient\",\"doc\":\"Inbreeding coefficient\",\"fields\":[{\"name\":\"sampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"This is the sample id against which the coefficient was estimated\"},{\"name\":\"program\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Name of program used to calculate the coefficient\"},{\"name\":\"version\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Version of the programme\"},{\"name\":\"estimationMethod\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Where various methods for estimation exist, which method was used.\"},{\"name\":\"coefficient\",\"type\":\"double\",\"doc\":\"Inbreeding coefficient ideally a real number in [0,1]\"},{\"name\":\"standardError\",\"type\":[\"null\",\"double\"],\"doc\":\"Standard error of the Inbreeding coefficient\"}]}],\"doc\":\"Inbreeding Coefficient Estimation\"},{\"name\":\"additionalInformation\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"We could add a map here to store additional information for example URIs to images, ECGs, etc\\n        Null by default\"},{\"name\":\"lastMenstrualPeriod\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Last Menstrual Period\"},{\"name\":\"diagnosticDetails\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DiagnosticDetail\",\"fields\":[{\"name\":\"diagnosisCodingSystem\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Diagnosis coding system\"},{\"name\":\"diagnosisCodingSystemVersion\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis Coding System Version\"},{\"name\":\"diagnosisCode\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Diagnosis Code\"},{\"name\":\"diagnosisDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis description\"},{\"name\":\"diagnosisCertainty\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Diagnosis Certainty\"},{\"name\":\"ageAtOnsetInYears\",\"type\":[\"null\",\"float\"],\"doc\":\"Age at diagnosis\"}]}}],\"doc\":\"Additional set of diagnostic ontology terms\"},{\"name\":\"participantUid\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"ParticipantGUID in GMS\"},{\"name\":\"clinicalEthnicities\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"enum\",\"name\":\"ClinicalEthnicity\",\"doc\":\"* A     British, Mixed British\\n    * B     Irish\\n    * C     Any other White background\\n    * C2    Northern Irish\\n    * C3    Other white, white unspecified\\n    * CA    English\\n    * CB    Scottish\\n    * CC    Welsh\\n    * CD    Cornish\\n    * CE    Cypriot (part not stated)\\n    * CF    Greek\\n    * CG    Greek Cypriot\\n    * CH    Turkish\\n    * CJ    Turkish Cypriot\\n    * CK    Italian\\n    * CL    Irish Traveller\\n    * CM    Traveller\\n    * CN    Gypsy/Romany\\n    * CP    Polish\\n    * CQ    All republics which made up the former USSR\\n    * CR    Kosovan\\n    * CS    Albanian\\n    * CT    Bosnian\\n    * CU    Croatian\\n    * CV    Serbian\\n    * CW    Other republics which made up the former Yugoslavia\\n    * CX    Mixed white\\n    * CY    Other white European, European unspecified, European mixed\\n    * D     White and Black Caribbean\\n    * E     White and Black African\\n    * F     White and Asian\\n    * G     Any other mixed background\\n    * GA    Black and Asian\\n    * GB    Black and Chinese\\n    * GC    Black and White\\n    * GD    Chinese and White\\n    * GE    Asian and Chinese\\n    * GF    Other Mixed, Mixed Unspecified\\n    * H     Indian or British Indian\\n    * J     Pakistani or British Pakistani\\n    * K     Bangladeshi or British Bangladeshi\\n    * L     Any other Asian background\\n    * LA    Mixed Asian\\n    * LB    Punjabi\\n    * LC    Kashmiri\\n    * LD    East African Asian\\n    * LE    Sri Lanka\\n    * LF    Tamil\\n    * LG    Sinhalese\\n    * LH    British Asian\\n    * LJ    Caribbean Asian\\n    * LK    Other Asian, Asian unspecified\\n    * M     Caribbean\\n    * N     African\\n    * P     Any other Black background\\n    * PA    Somali\\n    * PB    Mixed Black\\n    * PC    Nigerian\\n    * PD    Black British\\n    * PE    Other Black, Black unspecified\\n    * R     Chinese\\n    * S     Any other ethnic group\\n    * S1    Ashkenazi\\n    * S2    Sephardi\\n    * SA    Vietnamese\\n    * SB    Japanese\\n    * SC    Filipino\\n    * SD    Malaysian\\n    * SE    Any Other Group\\n    * Z     Not stated\",\"symbols\":[\"A\",\"B\",\"C\",\"C2\",\"C3\",\"CA\",\"CB\",\"CC\",\"CD\",\"CE\",\"CF\",\"CG\",\"CH\",\"CJ\",\"CK\",\"CL\",\"CM\",\"CN\",\"CP\",\"CQ\",\"CR\",\"CS\",\"CT\",\"CU\",\"CV\",\"CW\",\"CX\",\"CY\",\"D\",\"E\",\"F\",\"G\",\"GA\",\"GB\",\"GC\",\"GD\",\"GE\",\"GF\",\"H\",\"J\",\"K\",\"L\",\"LA\",\"LB\",\"LC\",\"LD\",\"LE\",\"LF\",\"LG\",\"LH\",\"LJ\",\"LK\",\"M\",\"N\",\"P\",\"PA\",\"PB\",\"PC\",\"PD\",\"PE\",\"R\",\"S\",\"S1\",\"S2\",\"SA\",\"SB\",\"SC\",\"SD\",\"SE\",\"Z\"]}}],\"doc\":\"ClinicalEthnicity as defined for GMS\"},{\"name\":\"previousTreatment\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"PreviousTreatment\",\"doc\":\"In GMS, Previous Treatment of Patient\",\"fields\":[{\"name\":\"previousTreatmentType\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"previousTreatmentName\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}]},{\"name\":\"previousTreatmentDate\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"Date\",\"doc\":\"This defines a date record\",\"fields\":[{\"name\":\"year\",\"type\":\"int\",\"doc\":\"Format YYYY\"},{\"name\":\"month\",\"type\":[\"null\",\"int\"],\"doc\":\"Format MM. e.g June is 06\"},{\"name\":\"day\",\"type\":[\"null\",\"int\"],\"doc\":\"Format DD e.g. 12th of October is 12\"}]}]}]}}],\"doc\":\"For GMS cases, previous Treatment Hi","story\"}]}},\"doc\":\"List of members of a pedigree\"},{\"name\":\"analysisPanels\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"AnalysisPanel\",\"doc\":\"An analysis panel\",\"fields\":[{\"name\":\"specificDisease\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The specific disease that a panel tests\"},{\"name\":\"panelName\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The name of the panel\"},{\"name\":\"panelId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Id of the panel\"},{\"name\":\"panelVersion\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The version of the panel\"},{\"name\":\"reviewOutcome\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Deprecated\"},{\"name\":\"multipleGeneticOrigins\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Deprecated\"}]}}],\"doc\":\"List of panels\"},{\"name\":\"diseasePenetrances\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"DiseasePenetrance\",\"doc\":\"A disease penetrance definition\",\"fields\":[{\"name\":\"specificDisease\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"The disease to which the penetrance applies\"},{\"name\":\"penetrance\",\"type\":{\"type\":\"enum\",\"name\":\"Penetrance\",\"doc\":\"Penetrance assumed in the analysis\",\"symbols\":[\"complete\",\"incomplete\"]},\"doc\":\"The penetrance\"}]}}],\"doc\":\"List of disease penetrances. Moved to referralTest for GMS\"},{\"name\":\"readyForAnalysis\",\"type\":\"boolean\",\"doc\":\"Flag indicating if a pedigree is ready for analysis\"},{\"name\":\"familyQCState\",\"type\":[\"null\",{\"type\":\"enum\",\"name\":\"FamilyQCState\",\"doc\":\"FamilyQCState\",\"symbols\":[\"noState\",\"passedMedicalReviewReadyForInterpretation\",\"passedMedicalReviewNotReadyForInterpretation\",\"queryToGel\",\"queryToGMC\",\"failed\"]}],\"doc\":\"The family quality control status\"}]}],\"doc\":\"Pedigree of the family.\"},{\"name\":\"otherFamilyHistory\",\"type\":[\"null\",{\"type\":\"record\",\"name\":\"OtherFamilyHistory\",\"doc\":\"Family history for secondary findings.\\n    Arrays of strings describing discrete family history phenotypes.\\n    Usually: `EndocrineTumours`, `colorectal`, `BreastOvarian` and `HDOrStroke` but can be others\",\"fields\":[{\"name\":\"maternalFamilyHistory\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Relevant Maternal family history\"},{\"name\":\"paternalFamilyHistory\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"string\",\"avro.java.string\":\"String\"}}],\"doc\":\"Relevant Maternal family history\"}]}],\"doc\":\"It is paternal or maternal with reference to the participant.\"},{\"name\":\"genePanelsCoverage\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"map\",\"values\":{\"type\":\"map\",\"values\":\"float\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"This map of key: panel_name, value: (map of key: gene, value: (map of metrics of key: metric name, value: float))\\n        That is: a map of tables of genes and metrics\"},{\"name\":\"interpretationFlags\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"InterpretationFlag\",\"doc\":\"A given interpretation flag together with an optional description\",\"fields\":[{\"name\":\"interpretationFlag\",\"type\":{\"type\":\"enum\",\"name\":\"InterpretationFlags\",\"doc\":\"Some flags relevant to the interpretation of a case\",\"symbols\":[\"mixed_chemistries\",\"mixedLab_preparation\",\"low_tumour_purity\",\"uniparental_isodisomy\",\"uniparental_heterodisomy\",\"unusual_karyotype\",\"high_cnv_count\",\"high_estimate_human_contamination_fraction\",\"mixed_recruiting_gmc\",\"suspected_mosaicism\",\"low_quality_sample\",\"ffpe_tumour_sample\",\"ff_nano_tumour_sample\",\"missing_values_for_proband_in_reported_variant\",\"reissued\",\"supplementary_report_errors\",\"internal_use_only\",\"high_priority\",\"suspected_increased_number_of_false_positive_heterozygous_loss_calls\",\"suspected_poor_quality_cnv_calls\",\"cnv_calls_assumed_xx_karyo\",\"cnv_calls_assumed_xy_karyo\",\"other\"]},\"doc\":\"The interpretation flag\"},{\"name\":\"additionalDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The description for the flag\"}]}}],\"doc\":\"Flags for this case relevant for interpretation\"},{\"name\":\"participantsInterpretationFlags\",\"type\":[\"null\",{\"type\":\"array\",\"items\":{\"type\":\"record\",\"name\":\"ParticipantInterpretationFlags\",\"doc\":\"Interpretation flags at the participant level\",\"fields\":[{\"name\":\"interpretationFlag\",\"type\":\"InterpretationFlags\",\"doc\":\"The interpretation flag\"},{\"name\":\"additionalDescription\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"The description for the flag\"},{\"name\":\"SampleId\",\"type\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"doc\":\"Sample Id fron which this flag was reported\"},{\"name\":\"participantId\",\"type\":[\"null\",{\"type\":\"string\",\"avro.java.string\":\"String\"}],\"doc\":\"Participant Id as appeared in the pedigree\"}]}}],\"doc\":\"Flags for this case relevant for interpretation per participant\"},{\"name\":\"additionalInfo\",\"type\":[\"null\",{\"type\":\"map\",\"values\":{\"type\":\"string\",\"avro.java.string\":\"String\"},\"avro.java.string\":\"String\"}],\"doc\":\"Additional information\"}]}}]}");
  public static org.apache.avro.Schema getClassSchema() { return SCHEMA$; }
   private org.gel.models.report.avro.InterpretationRequestRD json_request;

  /**
   * Default constructor.  Note that this does not initialize fields
   * to their default values from the schema.  If that is desired then
   * one should use <code>newBuilder()</code>. 
   */
  public DummyInterpretationRequest() {}

  /**
   * All-args constructor.
   */
  public DummyInterpretationRequest(org.gel.models.report.avro.InterpretationRequestRD json_request) {
    this.json_request = json_request;
  }

  public org.apache.avro.Schema getSchema() { return SCHEMA$; }
  // Used by DatumWriter.  Applications should not call. 
  public Object get(int field$) {
    switch (field$) {
    case 0: return json_request;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }
  // Used by DatumReader.  Applications should not call. 
  @SuppressWarnings(value="unchecked")
  public void put(int field$, Object value$) {
    switch (field$) {
    case 0: json_request = (org.gel.models.report.avro.InterpretationRequestRD)value$; break;
    default: throw new org.apache.avro.AvroRuntimeException("Bad index");
    }
  }

  /**
   * Gets the value of the 'json_request' field.
   */
  public org.gel.models.report.avro.InterpretationRequestRD getJsonRequest() {
    return json_request;
  }

  /**
   * Sets the value of the 'json_request' field.
   * @param value the value to set.
   */
  public void setJsonRequest(org.gel.models.report.avro.InterpretationRequestRD value) {
    this.json_request = value;
  }

  /** Creates a new DummyInterpretationRequest RecordBuilder */
  public static Builder newBuilder() {
    return new Builder();
  }
  
  /** Creates a new DummyInterpretationRequest RecordBuilder by copying an existing Builder */
  public static Builder newBuilder(Builder other) {
    return new Builder(other);
  }
  
  /** Creates a new DummyInterpretationRequest RecordBuilder by copying an existing DummyInterpretationRequest instance */
  public static Builder newBuilder(DummyInterpretationRequest other) {
    return new Builder(other);
  }
  
  /**
   * RecordBuilder for DummyInterpretationRequest instances.
   */
  public static class Builder extends org.apache.avro.specific.SpecificRecordBuilderBase<DummyInterpretationRequest>
    implements org.apache.avro.data.RecordBuilder<DummyInterpretationRequest> {

    private org.gel.models.report.avro.InterpretationRequestRD json_request;

    /** Creates a new Builder */
    private Builder() {
      super(DummyInterpretationRequest.SCHEMA$);
    }
    
    /** Creates a Builder by copying an existing Builder */
    private Builder(Builder other) {
      super(other);
      if (isValidValue(fields()[0], other.json_request)) {
        this.json_request = data().deepCopy(fields()[0].schema(), other.json_request);
        fieldSetFlags()[0] = true;
      }
    }
    
    /** Creates a Builder by copying an existing DummyInterpretationRequest instance */
    private Builder(DummyInterpretationRequest other) {
            super(DummyInterpretationRequest.SCHEMA$);
      if (isValidValue(fields()[0], other.json_request)) {
        this.json_request = data().deepCopy(fields()[0].schema(), other.json_request);
        fieldSetFlags()[0] = true;
      }
    }

    /** Gets the value of the 'json_request' field */
    public org.gel.models.report.avro.InterpretationRequestRD getJsonRequest() {
      return json_request;
    }
    
    /** Sets the value of the 'json_request' field */
    public Builder setJsonRequest(org.gel.models.report.avro.InterpretationRequestRD value) {
      validate(fields()[0], value);
      this.json_request = value;
      fieldSetFlags()[0] = true;
      return this; 
    }
    
    /** Checks whether the 'json_request' field has been set */
    public boolean hasJsonRequest() {
      return fieldSetFlags()[0];
    }
    
    /** Clears the value of the 'json_request' field */
    public Builder clearJsonRequest() {
      json_request = null;
      fieldSetFlags()[0] = false;
      return this;
    }

    @Override
    public DummyInterpretationRequest build() {
      try {
        DummyInterpretationRequest record = new DummyInterpretationRequest();
        record.json_request = fieldSetFlags()[0] ? this.json_request : (org.gel.models.report.avro.InterpretationRequestRD) defaultValue(fields()[0]);
        return record;
      } catch (Exception e) {
        throw new org.apache.avro.AvroRuntimeException(e);
      }
    }
  }
}
